{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file   = \"../../target/harmonized-table.db\"\n",
    "\n",
    "# biosample_sqlite_file   = \"../../biosample-analysis/target/harmonized-table.db\"\n",
    "\n",
    "# ncbitaxon_sqlite_file   = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "\n",
    "# where do we require a single ontology and where can we use multiple?\n",
    "target_onto_prefix    = \"PO\"\n",
    "# 'gaz', \n",
    "first_pass_ontologies = ['envo', 'ncbitaxon']\n",
    "\n",
    "chars_to_whiteout  = \"._-\"\n",
    "my_query_fields    = \"\"  # OLS weighted default\n",
    "my_row_req         = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "salvage_max_dist   = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What Biosample field should be mapped to ontology classes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_medium\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How should the mapping effort be scoped?\n",
    "Additional taxonomic filters may be applied below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "scoping_col   = \"env_package_normalized.EnvPackage\"\n",
    "scoping_value = \"soil\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ols_review_file = \"../local/ols_review.tsv\"\n",
    "# ols_review_seperator = \"\\t\"\n",
    "\n",
    "strategy_col = \"strategy\"\n",
    "include_col  = \"include\"\n",
    "\n",
    "first_pass_include_val  = True\n",
    "# first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val     = False\n",
    "# salvage_strategy_val    = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = biosample_col_to_map + \"_\" + scoping_col + \"_\" + scoping_value\n",
    "sssom_file = sssom_subject_prefix + \"_SSSOM.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# per_biosample_mapping_results_file = \"per_biosample_scoped_mapping_results.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## It's nice to see everything accounted for\n",
    "(In terms of manual mappings for env_package.) But we can prioritize NMDC for now:\n",
    "\n",
    "- Soil 15,777\n",
    "- Sediment 7,147\n",
    "- Plant-associated 3,142\n",
    "\n",
    "Could some of these \"no environmental package\" mappings be losing important granularity?\n",
    "\n",
    "Map `None` and '' to \"no environmental package\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_overrides = {\n",
    "    \"built environment\": \"built\",\n",
    "    \"misc environment\": \"miscellaneous\",\n",
    "    \"missing\": \"no environmental package\",\n",
    "    \"unknown\": \"no environmental package\",\n",
    "    \"default\": \"no environmental package\",\n",
    "    \"unspecified\": \"no environmental package\",\n",
    "    \"not available\": \"no environmental package\",\n",
    "    \"not collected\": \"no environmental package\",\n",
    "    \"miscellaneous natural or artificial environment\": \"miscellaneous\",\n",
    "    \"not applicable\": \"no environmental package\",\n",
    "    \"soil-associated\": \"soil\",\n",
    "    \"soil associated\": \"soil\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "# ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "# primary_ontology_sqlite_file = \"../semantic-sql/db/\" + target_onto_prefix.lower() + \".db\"\n",
    "# primary_ontology_cnx         = sqlite3.connect(primary_ontology_sqlite_file)\n",
    "first_pass_ontologies.insert(0, target_onto_prefix.lower())\n",
    "first_pass_ontologies = ','.join(first_pass_ontologies) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Print a sample of the data we're working with\n",
    "Specifically, INSDC/NCBI Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package</th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445970</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445972</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>449673</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445974</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>411461</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td>451639</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td>451638</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>451640</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>470145</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>15</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>471870</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id env_package      package                                  package_name  \\\n",
       "0   2     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "1   3     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "2   4     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "3   7     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "4   8     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "5  10              Generic.1.0                                       Generic   \n",
       "6  11              Generic.1.0                                       Generic   \n",
       "7  12     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "8  14     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "9  15     missing  MIGS.ba.5.0  MIGS: cultured bacteria/archaea; version 5.0   \n",
       "\n",
       "  host_taxid taxonomy_id                    env_broad_scale  \\\n",
       "0       9606      445970  terrestrial biome [ENVO:00000446]   \n",
       "1       9606      445972  terrestrial biome [ENVO:00000446]   \n",
       "2       9606      449673  terrestrial biome [ENVO:00000446]   \n",
       "3       9606      445974  terrestrial biome [ENVO:00000446]   \n",
       "4       9606      411461  terrestrial biome [ENVO:00000446]   \n",
       "5                 451639                                      \n",
       "6                 451638                                      \n",
       "7       9606      451640  terrestrial biome [ENVO:00000446]   \n",
       "8       9606      470145  terrestrial biome [ENVO:00000446]   \n",
       "9       9606      471870  terrestrial biome [ENVO:00000446]   \n",
       "\n",
       "                            env_local_scale  \\\n",
       "0  human-associated habitat [ENVO:00009003]   \n",
       "1  human-associated habitat [ENVO:00009003]   \n",
       "2  human-associated habitat [ENVO:00009003]   \n",
       "3  human-associated habitat [ENVO:00009003]   \n",
       "4  human-associated habitat [ENVO:00009003]   \n",
       "5                                             \n",
       "6                                             \n",
       "7  human-associated habitat [ENVO:00009003]   \n",
       "8  human-associated habitat [ENVO:00009003]   \n",
       "9  human-associated habitat [ENVO:00009003]   \n",
       "\n",
       "                           env_medium  \n",
       "0  biological product [ENVO:02000043]  \n",
       "1  biological product [ENVO:02000043]  \n",
       "2  biological product [ENVO:02000043]  \n",
       "3  biological product [ENVO:02000043]  \n",
       "4  biological product [ENVO:02000043]  \n",
       "5                                      \n",
       "6                                      \n",
       "7  biological product [ENVO:02000043]  \n",
       "8  biological product [ENVO:02000043]  \n",
       "9  biological product [ENVO:02000043]  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "biosample_first_ten"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get the canonical checklist and package terms from NCBI\n",
    "\n",
    "### Could (should?) also examine mixs.yaml (from where?)\n",
    "\n",
    "Unfortunately it doesn't do a very good job of differentiating checklists (MIMAG, MIMARKS, etc.) from packages (soil, water, etc.)\n",
    "\n",
    "_What about .ba, .euk, etc?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2.wwsurv.1.0</td>\n",
       "      <td>SARS-CoV-2: wastewater surveillance; version 1.0</td>\n",
       "      <td>SARS-CoV-2: wastewater surveillance</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 wastewater surveillance sam...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2          SARS-CoV-2.wwsurv.1.0   \n",
       "3                Pathogen.cl.1.0   \n",
       "4               Pathogen.env.1.0   \n",
       "..                           ...   \n",
       "150  MIUVIG.plant-associated.5.0   \n",
       "151          MIUVIG.sediment.5.0   \n",
       "152              MIUVIG.soil.5.0   \n",
       "153        MIUVIG.wastewater.5.0   \n",
       "154             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2     SARS-CoV-2: wastewater surveillance; version 1.0   \n",
       "3    Pathogen: clinical or host-associated; version...   \n",
       "4      Pathogen: environmental/food/other; version 1.0   \n",
       "..                                                 ...   \n",
       "150  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "151  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "152  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "153  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "154  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2        SARS-CoV-2: wastewater surveillance                     \n",
       "3      Pathogen: clinical or host-associated                     \n",
       "4         Pathogen: environmental/food/other                     \n",
       "..                                       ...               ...   \n",
       "150         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "151         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "152         MIUVIG Uncultivated Virus Genome              soil   \n",
       "153         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "154         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                      wgs_single;wgs_batch;wgs_diploid   \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "150  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "151          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "152              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "153        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "154             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2    Use for SARS-CoV-2 wastewater surveillance sam...                \n",
       "3                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "4                Environmental, food or other pathogen  SAMN02730065  \n",
       "..                                                 ...           ...  \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "154  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[155 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_dictionary = scoped_mapping.get_package_dictionary(biosample_packages_file)\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do the Biosample checklist/package fields match any of the canonical values?\n",
    "\n",
    "Start by counting the Biosample rows/records. See XXX notes on extracting this *harmonized* database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.111714\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18068841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_row_count\n",
       "0             18068841"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=False\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "biosample_row_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "Specifically, joining `biosample.package_name` with `package_dictionary.DisplayName`\n",
    "\n",
    "_Note that indices are built as part of the makefile_\n",
    "\n",
    "\n",
    "- create index biosample_package_name_idx on biosample(package_name);\n",
    "- create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "- create index biosample_package_idx on biosample(package);\n",
    "- create index biosample_p_pn_idx on biosample(package, package_name);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-04 16:50:47.091119\n",
      "2021-07-04 16:50:48.073258\n",
      "0:00:00.982139\n",
      "0:00:00.982139\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>canonical_package_name_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18068841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   canonical_package_name_count\n",
       "0                      18068841"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as canonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[canonical_package_name_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "canonical_package_name_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combinations of `package` and `package_name` values in the Biosample dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-04 16:50:48.675359\n",
      "2021-07-04 16:50:50.711518\n",
      "0:00:02.036159\n",
      "0:00:02.036159\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-lactamase.1.0</td>\n",
       "      <td>Beta-lactamase; version 1.0</td>\n",
       "      <td>630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>12383104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Human.1.0</td>\n",
       "      <td>Human; version 1.0</td>\n",
       "      <td>448733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Invertebrate.1.0</td>\n",
       "      <td>Invertebrate; version 1.0</td>\n",
       "      <td>173962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>34259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>737186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>294681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Plant.1.0</td>\n",
       "      <td>Plant; version 1.0</td>\n",
       "      <td>513794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>115291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Virus.1.0</td>\n",
       "      <td>Virus; version 1.0</td>\n",
       "      <td>19774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>134 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                package                                       package_name  \\\n",
       "0    Beta-lactamase.1.0                        Beta-lactamase; version 1.0   \n",
       "1           Generic.1.0                                            Generic   \n",
       "2             Human.1.0                                 Human; version 1.0   \n",
       "3      Invertebrate.1.0                          Invertebrate; version 1.0   \n",
       "4           MIGS.ba.5.0       MIGS: cultured bacteria/archaea; version 5.0   \n",
       "..                  ...                                                ...   \n",
       "129     Pathogen.cl.1.0  Pathogen: clinical or host-associated; version...   \n",
       "130    Pathogen.env.1.0    Pathogen: environmental/food/other; version 1.0   \n",
       "131           Plant.1.0                                 Plant; version 1.0   \n",
       "132   SARS-CoV-2.cl.1.0  SARS-CoV-2: clinical or host-associated; versi...   \n",
       "133           Virus.1.0                                 Virus; version 1.0   \n",
       "\n",
       "        count  \n",
       "0         630  \n",
       "1    12383104  \n",
       "2      448733  \n",
       "3      173962  \n",
       "4       34259  \n",
       "..        ...  \n",
       "129    737186  \n",
       "130    294681  \n",
       "131    513794  \n",
       "132    115291  \n",
       "133     19774  \n",
       "\n",
       "[134 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "package_name_combos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What about the Biosample `env_package` values?\n",
    "Are they also a small, highly regular set, like the `package` and `package_name` combinations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.942203\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>17815491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>69896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>53685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>17554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>misc environment</td>\n",
       "      <td>14418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>ocean[ENVO:00000015]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         env_package     count\n",
       "0                                     17815491\n",
       "1                    host-associated     69896\n",
       "2                          human-gut     53685\n",
       "3                              water     17554\n",
       "4                   misc environment     14418\n",
       "..                               ...       ...\n",
       "100                          Unknown         1\n",
       "101  marine sediment (ENVO:00002113)         1\n",
       "102             ocean[ENVO:00000015]         1\n",
       "103                           saliva         1\n",
       "104       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[105 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `env_package` is a mixture of `ENVO` term ids and strings\n",
    "Those strings may or may not be the term's labels. There are many redundancies due to small spelling and punctuation variations.\n",
    "\n",
    "Start by tidying `env_package` etc. values recognized by INSDC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackage\", \"eptidy\"\n",
    ")\n",
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackageDisplay\", \"epdtidy\"\n",
    ")\n",
    "\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
