{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "621bbcd9",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "# User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file = \"../target/harmonized_table.db\"\n",
    "\n",
    "# where do we require a single ontology and where can we use multiple?\n",
    "target_onto_prefix = \"PO\"\n",
    "\n",
    "first_pass_ontologies = [\"envo\", \"ncbitaxon\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21778e4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0bcc99",
   "metadata": {},
   "source": [
    "## It's nice to see everything accounted for\n",
    "(In terms of manual mappings for env_package.) But we can prioritize NMDC for now:\n",
    "\n",
    "- Soil 15,777\n",
    "- Sediment 7,147\n",
    "- Plant-associated 3,142\n",
    "\n",
    "Could some of these \"no environmental package\" mappings be losing important granularity?\n",
    "\n",
    "Map `None` and '' to \"no environmental package\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "068fa8f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_overrides = {\n",
    "    \"built environment\": \"built\",\n",
    "    \"misc environment\": \"miscellaneous\",\n",
    "    \"missing\": \"no environmental package\",\n",
    "    \"unknown\": \"no environmental package\",\n",
    "    \"default\": \"no environmental package\",\n",
    "    \"unspecified\": \"no environmental package\",\n",
    "    \"not available\": \"no environmental package\",\n",
    "    \"not collected\": \"no environmental package\",\n",
    "    \"miscellaneous natural or artificial environment\": \"miscellaneous\",\n",
    "    \"not applicable\": \"no environmental package\",\n",
    "    \"soil-associated\": \"soil\",\n",
    "    \"soil associated\": \"soil\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4494ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "\n",
    "first_pass_ontologies.insert(0, target_onto_prefix.lower())\n",
    "first_pass_ontologies_str = \",\".join(first_pass_ontologies)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be9eb2",
   "metadata": {},
   "source": [
    "## Determine ID patterns for common ontologies, like `ENVO`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "73b56b75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['envo', 'po']\n",
      "envo\n",
      "0:00:00.460708\n",
      "po\n",
      "0:00:00.054062\n"
     ]
    }
   ],
   "source": [
    "# first_pass_ontologies, including primary but not ncbitaxon\n",
    "# doesn't check for exisitence of DB files\n",
    "\n",
    "temp = first_pass_ontologies\n",
    "temp = [each_string.lower() for each_string in temp]\n",
    "temp.sort()\n",
    "temp = [each_string for each_string in temp if each_string != \"ncbitaxon\"]\n",
    "\n",
    "print(temp)\n",
    "\n",
    "first_pass_id_frames = []\n",
    "for one_ontology in temp:\n",
    "    print(one_ontology)\n",
    "    one_db_file = \"../semantic-sql/db/\" + one_ontology.lower() + \".db\"\n",
    "    #     print(one_db_file)\n",
    "    one_con = sqlite3.connect(one_db_file)\n",
    "\n",
    "    # FIXED? this wont include a term unless it's a class with a label\n",
    "    # add obsolete tags?\n",
    "    # may want to make a local-label only frame for later tasks?\n",
    "\n",
    "    q = \"\"\"\n",
    "select\n",
    "\tdistinct s1.stanza,\n",
    "\ts2.value\n",
    "from\n",
    "\tstatements s1\n",
    "left join statements s2 on\n",
    "\ts2.subject = s1.subject\n",
    "where\n",
    "\ts1.predicate = 'rdf:type'\n",
    "\tand s1.object = 'owl:Class'\n",
    "\tand s1.stanza = s1.subject\n",
    "\tand s2.predicate = 'rdfs:label'\"\"\"\n",
    "    [ids_labs_selected_ontolgies, query_duration] = scoped_mapping.timed_query(\n",
    "        q, one_con\n",
    "    )\n",
    "    print(query_duration)\n",
    "    ids_labs_selected_ontolgies[\"ontology\"] = one_ontology\n",
    "    first_pass_id_frames.append(ids_labs_selected_ontolgies)\n",
    "    one_con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71e95f09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stanza</th>\n",
       "      <th>value</th>\n",
       "      <th>ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000001</td>\n",
       "      <td>entity</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000002</td>\n",
       "      <td>continuant</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFO:0000003</td>\n",
       "      <td>occurrent</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFO:0000004</td>\n",
       "      <td>independent continuant</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFO:0000006</td>\n",
       "      <td>spatial region</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1783</th>\n",
       "      <td>PO:0030132</td>\n",
       "      <td>cincinnus inflorescence</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1784</th>\n",
       "      <td>PO:0030133</td>\n",
       "      <td>compound cincinnus inflorescence</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1785</th>\n",
       "      <td>PO:0030134</td>\n",
       "      <td>bostryx inflorescence</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786</th>\n",
       "      <td>PO:0030135</td>\n",
       "      <td>drepanium inflorescence</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1787</th>\n",
       "      <td>PO:0030136</td>\n",
       "      <td>compound drepanium inflorescence</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           stanza                             value ontology\n",
       "0     BFO:0000001                            entity     envo\n",
       "1     BFO:0000002                        continuant     envo\n",
       "2     BFO:0000003                         occurrent     envo\n",
       "3     BFO:0000004            independent continuant     envo\n",
       "4     BFO:0000006                    spatial region     envo\n",
       "...           ...                               ...      ...\n",
       "1783   PO:0030132           cincinnus inflorescence       po\n",
       "1784   PO:0030133  compound cincinnus inflorescence       po\n",
       "1785   PO:0030134             bostryx inflorescence       po\n",
       "1786   PO:0030135           drepanium inflorescence       po\n",
       "1787   PO:0030136  compound drepanium inflorescence       po\n",
       "\n",
       "[8355 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids_labs_selected_ontolgies = pd.concat(first_pass_id_frames)\n",
    "ids_labs_selected_ontolgies.to_sql(\n",
    "    \"ids_labs_selected_ontolgies\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "ids_labs_selected_ontolgies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dcf3bb61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stanza</th>\n",
       "      <th>value</th>\n",
       "      <th>ontology</th>\n",
       "      <th>prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000001</td>\n",
       "      <td>entity</td>\n",
       "      <td>envo</td>\n",
       "      <td>BFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000002</td>\n",
       "      <td>continuant</td>\n",
       "      <td>envo</td>\n",
       "      <td>BFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFO:0000003</td>\n",
       "      <td>occurrent</td>\n",
       "      <td>envo</td>\n",
       "      <td>BFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFO:0000004</td>\n",
       "      <td>independent continuant</td>\n",
       "      <td>envo</td>\n",
       "      <td>BFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFO:0000006</td>\n",
       "      <td>spatial region</td>\n",
       "      <td>envo</td>\n",
       "      <td>BFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1783</th>\n",
       "      <td>PO:0030132</td>\n",
       "      <td>cincinnus inflorescence</td>\n",
       "      <td>po</td>\n",
       "      <td>PO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1784</th>\n",
       "      <td>PO:0030133</td>\n",
       "      <td>compound cincinnus inflorescence</td>\n",
       "      <td>po</td>\n",
       "      <td>PO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1785</th>\n",
       "      <td>PO:0030134</td>\n",
       "      <td>bostryx inflorescence</td>\n",
       "      <td>po</td>\n",
       "      <td>PO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786</th>\n",
       "      <td>PO:0030135</td>\n",
       "      <td>drepanium inflorescence</td>\n",
       "      <td>po</td>\n",
       "      <td>PO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1787</th>\n",
       "      <td>PO:0030136</td>\n",
       "      <td>compound drepanium inflorescence</td>\n",
       "      <td>po</td>\n",
       "      <td>PO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           stanza                             value ontology prefix\n",
       "0     BFO:0000001                            entity     envo    BFO\n",
       "1     BFO:0000002                        continuant     envo    BFO\n",
       "2     BFO:0000003                         occurrent     envo    BFO\n",
       "3     BFO:0000004            independent continuant     envo    BFO\n",
       "4     BFO:0000006                    spatial region     envo    BFO\n",
       "...           ...                               ...      ...    ...\n",
       "1783   PO:0030132           cincinnus inflorescence       po     PO\n",
       "1784   PO:0030133  compound cincinnus inflorescence       po     PO\n",
       "1785   PO:0030134             bostryx inflorescence       po     PO\n",
       "1786   PO:0030135           drepanium inflorescence       po     PO\n",
       "1787   PO:0030136  compound drepanium inflorescence       po     PO\n",
       "\n",
       "[8355 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids_labs_selected_ontolgies = scoped_mapping.add_prefix_col(\n",
    "    ids_labs_selected_ontolgies, \"stanza\", \"prefix\"\n",
    ")\n",
    "ids_labs_selected_ontolgies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8b9e5da2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BFO': 'BFO:\\\\d{7}',\n",
       " 'CARO': 'CARO:\\\\d{7}',\n",
       " 'CHEBI': 'CHEBI:\\\\d{4,6}',\n",
       " 'ENVO': 'ENVO:\\\\d{7,8}',\n",
       " 'FAO': 'FAO:0000001',\n",
       " 'FOODON': 'FOODON:\\\\d{8}',\n",
       " 'GO': 'GO:\\\\d{7}',\n",
       " 'IAO': 'IAO:\\\\d{7}',\n",
       " 'NCBITaxon': 'NCBITaxon:\\\\d+',\n",
       " 'OBI': 'OBI:\\\\d{7}',\n",
       " 'PATO': 'PATO:\\\\d{7}',\n",
       " 'PCO': 'PCO:\\\\d{7}',\n",
       " 'PO': 'PO:\\\\d{7}',\n",
       " 'RO': 'RO:0002577',\n",
       " 'UBERON': 'UBERON:\\\\d{7}'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_patterns = scoped_mapping.get_multi_term_patterns(\n",
    "    ids_labs_selected_ontolgies, \"stanza\", \"prefix\"\n",
    ")\n",
    "\n",
    "id_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a75abaf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ipf = pd.DataFrame(id_patterns.items(), columns=[\"ontology\", \"id_pattern\"])\n",
    "ipf.to_sql(\"id_patterns\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a86ea30e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # round trip\n",
    "# ipd = dict(zip(ipf.ontology, ipf.id_pattern))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e587086",
   "metadata": {},
   "source": [
    "## Retreive `env_package` values from Biosample table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1513523f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.100606\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count\n",
       "0                              None  14083847\n",
       "1                   host-associated     49254\n",
       "2                         human-gut     47921\n",
       "3                             water     16367\n",
       "4                        human-skin     13706\n",
       "..                              ...       ...\n",
       "87                      env_package         1\n",
       "88                   gut microbiome         1\n",
       "89  marine sediment (ENVO:00002113)         1\n",
       "90                           saliva         1\n",
       "91       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c56612",
   "metadata": {},
   "source": [
    "## Apply some normalization rules to the `env_package` values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d8f0613",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>marine sediment envo 00002113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>sea water envo 00002149</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs extract                 remaining_string  \\\n",
       "0                              None                                            \n",
       "1                   host-associated                          host-associated   \n",
       "2                         human-gut                                human-gut   \n",
       "3                             water                                    water   \n",
       "4                        human-skin                               human-skin   \n",
       "..                              ...     ...                              ...   \n",
       "87                      env_package                              env_package   \n",
       "88                   gut microbiome                           gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)          marine sediment (ENVO:00002113)   \n",
       "90                           saliva                                   saliva   \n",
       "91       sea water, [ENVO:00002149]               sea water, [ENVO:00002149]   \n",
       "\n",
       "                 remaining_tidied  \n",
       "0                                  \n",
       "1                 host associated  \n",
       "2                       human gut  \n",
       "3                           water  \n",
       "4                      human skin  \n",
       "..                            ...  \n",
       "87                    env package  \n",
       "88                 gut microbiome  \n",
       "89  marine sediment envo 00002113  \n",
       "90                         saliva  \n",
       "91        sea water envo 00002149  \n",
       "\n",
       "[92 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalization = scoped_mapping.env_package_nomralizastion(\n",
    "    env_package_count, \"env_package\", id_patterns[target_onto_prefix]\n",
    ")\n",
    "\n",
    "# getting rid of redundant? 'string' column\n",
    "env_package_normalization = env_package_normalization[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "env_package_normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "151e482a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microbe.1.0</td>\n",
       "      <td>Microbe; version 1.0</td>\n",
       "      <td>Microbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Use for bacteria or other unicellular microbes...</td>\n",
       "      <td>SAMN02911891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2                Pathogen.cl.1.0   \n",
       "3               Pathogen.env.1.0   \n",
       "4                    Microbe.1.0   \n",
       "..                           ...   \n",
       "149  MIUVIG.plant-associated.5.0   \n",
       "150          MIUVIG.sediment.5.0   \n",
       "151              MIUVIG.soil.5.0   \n",
       "152        MIUVIG.wastewater.5.0   \n",
       "153             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2    Pathogen: clinical or host-associated; version...   \n",
       "3      Pathogen: environmental/food/other; version 1.0   \n",
       "4                                 Microbe; version 1.0   \n",
       "..                                                 ...   \n",
       "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2      Pathogen: clinical or host-associated                     \n",
       "3         Pathogen: environmental/food/other                     \n",
       "4                                    Microbe                     \n",
       "..                                       ...               ...   \n",
       "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "151         MIUVIG Uncultivated Virus Genome              soil   \n",
       "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "153         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "3                Environmental, food or other pathogen  SAMN02730065  \n",
       "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
       "..                                                 ...           ...  \n",
       "149  Use for uncultivated virus genome identified i...                \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_dictionary = scoped_mapping.get_package_dictionary(biosample_packages_file)\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5f346dd",
   "metadata": {},
   "source": [
    "pick \"EnvPackage\"/\"eptidy\" from package dictionary as canonical, not \"EnvPackageDisplay\"/\"epdtidy\"\n",
    "\n",
    "\n",
    "But still want to support making XXX values from YYY canonical according to \"EnvPackageDisplay\"/\"epdtidy\", so make a mapping/override table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "48043728",
   "metadata": {},
   "outputs": [],
   "source": [
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackage\", \"eptidy\"\n",
    ")\n",
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackageDisplay\", \"epdtidy\"\n",
    ")\n",
    "\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f91fa7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eptidy</th>\n",
       "      <th>epdtidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous or artificial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           eptidy                      epdtidy\n",
       "21      microbial        microbial mat biofilm\n",
       "22  miscellaneous  miscellaneous or artificial\n",
       "26     wastewater            wastewater sludge"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epd_to_ep = package_dictionary[[\"eptidy\", \"epdtidy\"]]\n",
    "# drop duplicates\n",
    "epd_to_ep = epd_to_ep.drop_duplicates()\n",
    "\n",
    "# drop blank eptidy rows\n",
    "ep_blank_flag = epd_to_ep[\"eptidy\"].eq(\"\")\n",
    "epd_to_ep = epd_to_ep.loc[~ep_blank_flag]\n",
    "\n",
    "# drop rows where eptidy and epdtidy are the same\n",
    "identical_flag = epd_to_ep[\"eptidy\"] == epd_to_ep[\"epdtidy\"]\n",
    "epd_to_ep = epd_to_ep.loc[~identical_flag]\n",
    "\n",
    "epd_to_ep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d17acd46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# and add to manualy asserted overrides above\n",
    "overrides_supplement = dict(zip(epd_to_ep[\"epdtidy\"], epd_to_ep[\"eptidy\"]))\n",
    "\n",
    "overrides_supplement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77aca877",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'built environment': 'built',\n",
       " 'misc environment': 'miscellaneous',\n",
       " 'missing': 'no environmental package',\n",
       " 'unknown': 'no environmental package',\n",
       " 'default': 'no environmental package',\n",
       " 'unspecified': 'no environmental package',\n",
       " 'not available': 'no environmental package',\n",
       " 'not collected': 'no environmental package',\n",
       " 'miscellaneous natural or artificial environment': 'miscellaneous',\n",
       " 'not applicable': 'no environmental package',\n",
       " 'soil-associated': 'soil',\n",
       " 'soil associated': 'soil',\n",
       " 'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_overrides.update(overrides_supplement)\n",
    "\n",
    "env_package_overrides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c4c2c6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalization = scoped_mapping.add_overrides(\n",
    "    env_package_normalization, \"remaining_tidied\", \"rt_override\", env_package_overrides\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3b7eb038",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>eptidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>No environmental package</td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>air</td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>built</td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>human-associated</td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>human-oral</td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>human-vaginal</td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  EnvPackage                    eptidy\n",
       "0                                                     \n",
       "12  No environmental package  no environmental package\n",
       "13                       air                       air\n",
       "14                     built                     built\n",
       "15           host-associated           host associated\n",
       "16          human-associated          human associated\n",
       "17                 human-gut                 human gut\n",
       "18                human-oral                human oral\n",
       "19                human-skin                human skin\n",
       "20             human-vaginal             human vaginal\n",
       "21                 microbial                 microbial\n",
       "22             miscellaneous             miscellaneous\n",
       "23          plant-associated          plant associated\n",
       "24                  sediment                  sediment\n",
       "25                      soil                      soil\n",
       "26                wastewater                wastewater\n",
       "27                     water                     water"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "denorm_frame = package_dictionary[[\"EnvPackage\", \"eptidy\"]]\n",
    "denorm_frame = denorm_frame.drop_duplicates()\n",
    "denorm_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "73694ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalization = env_package_normalization.merge(\n",
    "    denorm_frame, how=\"left\", left_on=\"rt_override\", right_on=\"eptidy\"\n",
    ")\n",
    "\n",
    "env_package_normalization = env_package_normalization[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "        \"rt_override\",\n",
    "        \"EnvPackage\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "non_canonical_flag = env_package_normalization[\"EnvPackage\"].isna()\n",
    "env_package_normalization[\"is_canonical\"] = True\n",
    "env_package_normalization.loc[non_canonical_flag, \"is_canonical\"] = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db3fd267",
   "metadata": {},
   "source": [
    "- env_package = env_package annotation from NCBI Biosample file XXX\n",
    "- count = number of biosamples using that env_package annotation\n",
    "- lhs = checklist info\n",
    "- rhs = potential package info\n",
    "- extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "- remaining_string = rhs/string, with potential OBO IDs removed\n",
    "- remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "- rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "- EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "- is_canonical = false when EnvPackage is NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f2d82ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalization.to_sql(\n",
    "    \"env_package_normalization\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3766cbe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>rt_override</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>is_canonical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host-associated</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human-skin</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "      <td>env package</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>marine sediment envo 00002113</td>\n",
       "      <td>marine sediment envo 00002113</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>sea water envo 00002149</td>\n",
       "      <td>sea water envo 00002149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs extract                 remaining_string  \\\n",
       "0                              None                                            \n",
       "1                   host-associated                          host-associated   \n",
       "2                         human-gut                                human-gut   \n",
       "3                             water                                    water   \n",
       "4                        human-skin                               human-skin   \n",
       "..                              ...     ...                              ...   \n",
       "87                      env_package                              env_package   \n",
       "88                   gut microbiome                           gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)          marine sediment (ENVO:00002113)   \n",
       "90                           saliva                                   saliva   \n",
       "91       sea water, [ENVO:00002149]               sea water, [ENVO:00002149]   \n",
       "\n",
       "                 remaining_tidied                    rt_override  \\\n",
       "0                                                                  \n",
       "1                 host associated                host associated   \n",
       "2                       human gut                      human gut   \n",
       "3                           water                          water   \n",
       "4                      human skin                     human skin   \n",
       "..                            ...                            ...   \n",
       "87                    env package                    env package   \n",
       "88                 gut microbiome                 gut microbiome   \n",
       "89  marine sediment envo 00002113  marine sediment envo 00002113   \n",
       "90                         saliva                         saliva   \n",
       "91        sea water envo 00002149        sea water envo 00002149   \n",
       "\n",
       "         EnvPackage  is_canonical  \n",
       "0                            True  \n",
       "1   host-associated          True  \n",
       "2         human-gut          True  \n",
       "3             water          True  \n",
       "4        human-skin          True  \n",
       "..              ...           ...  \n",
       "87              NaN         False  \n",
       "88              NaN         False  \n",
       "89              NaN         False  \n",
       "90              NaN         False  \n",
       "91              NaN         False  \n",
       "\n",
       "[92 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a7fed8",
   "metadata": {},
   "source": [
    "## What do the successful normalizations look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7c1fbb75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.001661\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Water</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>not collected</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>host associated</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        env_package     count lhs extract                EnvPackage\n",
       "0              None  14083847                                      \n",
       "1   host-associated     49254                       host-associated\n",
       "2         human-gut     47921                             human-gut\n",
       "3             water     16367                                 water\n",
       "4        human-skin     13706                            human-skin\n",
       "..              ...       ...  ..     ...                       ...\n",
       "56            Water         3                                 water\n",
       "57    not collected         3              No environmental package\n",
       "58  host associated         2                       host-associated\n",
       "59         Sediment         1                              sediment\n",
       "60          Unknown         1              No environmental package\n",
       "\n",
       "[61 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalization\n",
    "where\n",
    "    is_canonical = 1\n",
    "\"\"\"\n",
    "[successful_normalizastions, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "successful_normalizastions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "740ffe3c",
   "metadata": {},
   "source": [
    "# Are there any normalization failures?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "17e08383",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.001545\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mimarks</td>\n",
       "      <td>479</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse-gut</td>\n",
       "      <td>406</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gut</td>\n",
       "      <td>172</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>biofilm</td>\n",
       "      <td>114</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-not providedsopharyngeal</td>\n",
       "      <td>107</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mice gut</td>\n",
       "      <td>87</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CV</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>home, outdoor environment</td>\n",
       "      <td>44</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fermentation-associated</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sterile water</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Mouse_Gut</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fresh water</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rock</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>arctic</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>freshwater sediment</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>seawater</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MIENS_16S</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MIMARKS_water</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cave</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>plant</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sea water</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Coastal water body (ENVO:02000049)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Deep-sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MIMARKS_16S</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           env_package  count lhs extract EnvPackage\n",
       "0                              mimarks    479                   None\n",
       "1                            mouse-gut    406                   None\n",
       "2                                  gut    172                   None\n",
       "3                              biofilm    114                   None\n",
       "4       human-not providedsopharyngeal    107                   None\n",
       "5                             mice gut     87                   None\n",
       "6                                   CV     60                   None\n",
       "7            home, outdoor environment     44                   None\n",
       "8              fermentation-associated     42                   None\n",
       "9                        sterile water     35                   None\n",
       "10                           Mouse_Gut     27                   None\n",
       "11                         fresh water     27                   None\n",
       "12                                rock     12                   None\n",
       "13                              arctic     10                   None\n",
       "14                 freshwater sediment      9                   None\n",
       "15                            seawater      9                   None\n",
       "16                           MIENS_16S      6                   None\n",
       "17                       MIMARKS_water      6                   None\n",
       "18                                cave      5                   None\n",
       "19                               plant      3                   None\n",
       "20                           sea water      2                   None\n",
       "21  Coastal water body (ENVO:02000049)      1                   None\n",
       "22                      Deep-sea water      1                   None\n",
       "23                       ENVO:00000016      1                   None\n",
       "24                         MIMARKS_16S      1                   None\n",
       "25                           Sea water      1                   None\n",
       "26                         env_package      1                   None\n",
       "27                      gut microbiome      1                   None\n",
       "28     marine sediment (ENVO:00002113)      1                   None\n",
       "29                              saliva      1                   None\n",
       "30          sea water, [ENVO:00002149]      1                   None"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalization\n",
    "where\n",
    "    is_canonical = 0\n",
    "\"\"\"\n",
    "[normalizastion_failures, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "normalizastion_failures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8382a3ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursor = biosample_cnx.cursor()\n",
    "statement = \"\"\"\n",
    "DROP TABLE if exists repaired_env_package ;\n",
    "\n",
    "CREATE TABLE repaired_env_package AS\n",
    "select\n",
    "\tb.id,\n",
    "\tepn.env_package as env_package_orig,\n",
    "\tepn.EnvPackage as env_package_rep\n",
    "from\n",
    "\tenv_package_normalization epn\n",
    "join biosample b on\n",
    "\tb.env_package = epn.env_package\n",
    "where\n",
    "\tis_canonical = 1\n",
    "\tand EnvPackage != '';\n",
    "\t\n",
    "select\n",
    "\tcount(*)\n",
    "from\n",
    "\trepaired_env_package rep\n",
    "\"\"\"\n",
    "\n",
    "cursor.executescript(statement)\n",
    "biosample_cnx.commit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
