{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import OGER output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])? y\n"
     ]
    }
   ],
   "source": [
    "%reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 49 µs, sys: 0 ns, total: 49 µs\n",
      "Wall time: 54.1 µs\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of rows:  2386121\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BiosampleId</th>\n",
       "      <th>Biolink</th>\n",
       "      <th>BeginTerm</th>\n",
       "      <th>EndTerm</th>\n",
       "      <th>TokenizedTerm</th>\n",
       "      <th>PreferredTerm</th>\n",
       "      <th>CURIE</th>\n",
       "      <th>S-Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN10586215</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "      <td>contains</td>\n",
       "      <td>contains</td>\n",
       "      <td>RO:0001019</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN10586215</td>\n",
       "      <td>biolink:OrganismalEntity</td>\n",
       "      <td>119</td>\n",
       "      <td>127</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN05720810</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>50</td>\n",
       "      <td>57</td>\n",
       "      <td>part of</td>\n",
       "      <td>part of</td>\n",
       "      <td>BFO:0000050</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN05720810</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>79</td>\n",
       "      <td>90</td>\n",
       "      <td>Agriculture</td>\n",
       "      <td>agricultural process</td>\n",
       "      <td>ENVO:01001246</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN05720810</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>79</td>\n",
       "      <td>90</td>\n",
       "      <td>Agriculture</td>\n",
       "      <td>animal husbandry</td>\n",
       "      <td>ENVO:01001248</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    BiosampleId                   Biolink  BeginTerm  EndTerm TokenizedTerm  \\\n",
       "0  SAMN10586215     biolink:OntologyClass         16       24      contains   \n",
       "1  SAMN10586215  biolink:OrganismalEntity        119      127      Bacteria   \n",
       "2  SAMN05720810     biolink:OntologyClass         50       57       part of   \n",
       "3  SAMN05720810     biolink:OntologyClass         79       90   Agriculture   \n",
       "4  SAMN05720810     biolink:OntologyClass         79       90   Agriculture   \n",
       "\n",
       "          PreferredTerm          CURIE S-Type  \n",
       "0              contains     RO:0001019     S1  \n",
       "1              Bacteria    NCBITaxon:2     S1  \n",
       "2               part of    BFO:0000050     S1  \n",
       "3  agricultural process  ENVO:01001246     S1  \n",
       "4      animal husbandry  ENVO:01001248     S1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = ['BiosampleId', 'Biolink', 'BeginTerm', 'EndTerm', 'TokenizedTerm', 'PreferredTerm', 'CURIE', 'NaN1', 'S-Type', 'NaN2', 'UMLS_CUI']\n",
    "ogerOutput = pd.read_csv('../output/FullNLPOutput.tsv', sep='\\t', names=cols )\n",
    "ogerOutput = ogerOutput.drop(['NaN1', 'NaN2', 'UMLS_CUI'], axis = 1)\n",
    "print('Total number of rows: ', str(len(ogerOutput)))\n",
    "ogerOutput.head()\n",
    "#pd.set_option(\"display.max_rows\", None)\n",
    "#print(ogerOutput.CURIE.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 948401\n",
      "Number of unique Ids: 938333\n",
      "Number of unique Desc: 59789\n",
      "******After duplicate row removal*****\n",
      "Number of rows: 947374\n",
      "Number of unique Ids: 938333\n",
      "Number of unique Description: 59789\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>BiosampleId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Treponema denticola ATCC 35405. Treponema dent...</td>\n",
       "      <td>SAMN02603967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Treponema pallidum subsp. pallidum. This organ...</td>\n",
       "      <td>SAMN02604348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Campylobacter fetus subsp. venerealis str. Azu...</td>\n",
       "      <td>SAMN02471365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Campylobacter jejuni strain NCTC11168. This st...</td>\n",
       "      <td>SAMEA1705929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Francisella tularensis subsp. tularensis strai...</td>\n",
       "      <td>SAMEA3138185</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Description   BiosampleId\n",
       "0  Treponema denticola ATCC 35405. Treponema dent...  SAMN02603967\n",
       "1  Treponema pallidum subsp. pallidum. This organ...  SAMN02604348\n",
       "2  Campylobacter fetus subsp. venerealis str. Azu...  SAMN02471365\n",
       "3  Campylobacter jejuni strain NCTC11168. This st...  SAMEA1705929\n",
       "4  Francisella tularensis subsp. tularensis strai...  SAMEA3138185"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ogerInput = pd.read_csv('../output/biosampleDescriptionDF.tsv', sep='\\t')\n",
    "ogerInput.BiosampleId = ogerInput.BiosampleId.map(lambda x: x.replace('BIOSAMPLE:',''))\n",
    "ogerInput = ogerInput.drop(['StudyId', 'Name', 'Title'], axis = 1)\n",
    "print('Number of rows: '+ str(len(ogerInput)))\n",
    "print('Number of unique Ids: '+ str(len(ogerInput.BiosampleId.unique())))\n",
    "print('Number of unique Desc: '+ str(len(ogerInput.Description.unique())))\n",
    "ogerInput = ogerInput.drop_duplicates()\n",
    "print('******After duplicate row removal*****')\n",
    "print('Number of rows: '+ str(len(ogerInput)))\n",
    "print('Number of unique Ids: '+ str(len(ogerInput.BiosampleId.unique())))\n",
    "print('Number of unique Description: '+ str(len(ogerInput.Description.unique())))\n",
    "ogerInput.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique BiosampleIds: 649677\n",
      "Unique CURIEs: 1686\n",
      "Unique Biolinks: 4\n",
      "Unique Standard Terms: 1677\n"
     ]
    }
   ],
   "source": [
    "print('Unique BiosampleIds: '+ str(len(ogerOutput.BiosampleId.unique())))\n",
    "print('Unique CURIEs: '+ str(len(ogerOutput.CURIE.unique())))\n",
    "print('Unique Biolinks: '+ str(len(ogerOutput.Biolink.unique())))\n",
    "print('Unique Standard Terms: '+ str(len(ogerOutput.PreferredTerm.unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curieOfInterest = ['ENVO', 'PATO', 'RO']\n",
    "envoPato = ogerOutput[ogerOutput.CURIE.str.contains('|'.join(curieOfInterest))]\n",
    "print('Number of rows: '+ str(len(envoPato)))\n",
    "envoPato.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "salineTerms = ['sali', 'salt']\n",
    "envPatSal = envoPato[envoPato.TokenizedTerm\t.str.contains('|'.join(salineTerms))]\n",
    "print('Number of rows: '+ str(len(envPatSal)))\n",
    "envPatSal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows:  18615\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tot_diss_nitro</th>\n",
       "      <th>salinity_meth</th>\n",
       "      <th>tidal_stage</th>\n",
       "      <th>num_replicons</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>rel_to_oxygen</th>\n",
       "      <th>potassium</th>\n",
       "      <th>sodium</th>\n",
       "      <th>env_package</th>\n",
       "      <th>biomass</th>\n",
       "      <th>...</th>\n",
       "      <th>depth</th>\n",
       "      <th>silicate</th>\n",
       "      <th>tot_part_carb</th>\n",
       "      <th>trophic_level</th>\n",
       "      <th>calcium</th>\n",
       "      <th>id</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>taxonomy_name</th>\n",
       "      <th>package</th>\n",
       "      <th>norm_env_package</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.66</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706329</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62.28</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706311</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.05</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706320</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>126.11</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706300</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>332.17</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706316</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  tot_diss_nitro salinity_meth tidal_stage num_replicons nitrate  \\\n",
       "0            NaN           NaN         NaN           NaN     NaN   \n",
       "1            NaN           NaN         NaN           NaN     NaN   \n",
       "2            NaN           NaN         NaN           NaN     NaN   \n",
       "3            NaN           NaN         NaN           NaN     NaN   \n",
       "4            NaN           NaN         NaN           NaN     NaN   \n",
       "\n",
       "  rel_to_oxygen potassium sodium env_package biomass  ... depth silicate  \\\n",
       "0           NaN      2.33    NaN       water     NaN  ...   NaN      NaN   \n",
       "1           NaN      1.33    NaN       water     NaN  ...   NaN      NaN   \n",
       "2           NaN      2.77    NaN       water     NaN  ...   NaN      NaN   \n",
       "3           NaN      2.52    NaN       water     NaN  ...   NaN      NaN   \n",
       "4           NaN       3.4    NaN       water     NaN  ...   NaN      NaN   \n",
       "\n",
       "  tot_part_carb trophic_level calcium                      id taxonomy_id  \\\n",
       "0           NaN           NaN   87.66  BIOSAMPLE:SAMEA1706329      449393   \n",
       "1           NaN           NaN   62.28  BIOSAMPLE:SAMEA1706311      449393   \n",
       "2           NaN           NaN   87.05  BIOSAMPLE:SAMEA1706320      449393   \n",
       "3           NaN           NaN  126.11  BIOSAMPLE:SAMEA1706300      449393   \n",
       "4           NaN           NaN  332.17  BIOSAMPLE:SAMEA1706316      449393   \n",
       "\n",
       "           taxonomy_name      package  norm_env_package  \n",
       "0  freshwater metagenome  Generic.1.0             water  \n",
       "1  freshwater metagenome  Generic.1.0             water  \n",
       "2  freshwater metagenome  Generic.1.0             water  \n",
       "3  freshwater metagenome  Generic.1.0             water  \n",
       "4  freshwater metagenome  Generic.1.0             water  \n",
       "\n",
       "[5 rows x 97 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get list of IDs from non-human-sample file\n",
    "#nonHumanDF = pd.read_csv('../input/non-human-samples.tsv.gz', sep='\\t', low_memory=False)\n",
    "structuredDataDF = pd.read_csv('../input/water-related-samples.tsv.gz', sep='\\t', low_memory=False)\n",
    "print('Number of rows: ', str(len(structuredDataDF)))\n",
    "structuredDataDF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get descriptions of non-human-samples only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From OGER Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 243\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>BiosampleId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Purple microbial (Photo- and chemosynthetic) m...</td>\n",
       "      <td>SAMN02228699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Purple microbial (Photo- and chemosynthetic) m...</td>\n",
       "      <td>SAMN02228700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Purple microbial (Photo- and chemosynthetic) m...</td>\n",
       "      <td>SAMN02228701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Purple microbial (Photo- and chemosynthetic) m...</td>\n",
       "      <td>SAMN02228702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Purple microbial (Photo- and chemosynthetic) m...</td>\n",
       "      <td>SAMN02228703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Description   BiosampleId\n",
       "0  Purple microbial (Photo- and chemosynthetic) m...  SAMN02228699\n",
       "1  Purple microbial (Photo- and chemosynthetic) m...  SAMN02228700\n",
       "2  Purple microbial (Photo- and chemosynthetic) m...  SAMN02228701\n",
       "3  Purple microbial (Photo- and chemosynthetic) m...  SAMN02228702\n",
       "4  Purple microbial (Photo- and chemosynthetic) m...  SAMN02228703"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structDataID = structuredDataDF.id.str.replace('BIOSAMPLE:','')\n",
    "structDataIdDF = pd.DataFrame(structDataID)\n",
    "structDataOgerInput = ogerInput.merge(structDataIdDF, how='inner', left_on='BiosampleId', right_on='id')\n",
    "structDataOgerInput = structDataOgerInput.drop(['id'], axis = 1)\n",
    "print('Number of rows: '+ str(len(nhOgerInput)))\n",
    "structDataOgerInput.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From OGER output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of row: 2429\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BiosampleId</th>\n",
       "      <th>Biolink</th>\n",
       "      <th>BeginTerm</th>\n",
       "      <th>EndTerm</th>\n",
       "      <th>TokenizedTerm</th>\n",
       "      <th>PreferredTerm</th>\n",
       "      <th>CURIE</th>\n",
       "      <th>S-Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>biolink:ChemicalSubstance</td>\n",
       "      <td>19</td>\n",
       "      <td>24</td>\n",
       "      <td>probe</td>\n",
       "      <td>probe</td>\n",
       "      <td>CHEBI:50406</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>61</td>\n",
       "      <td>65</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>biolink:AnatomicalEntity</td>\n",
       "      <td>70</td>\n",
       "      <td>73</td>\n",
       "      <td>gut</td>\n",
       "      <td>digestive tract</td>\n",
       "      <td>UBERON:0001555</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>biolink:OntologyClass</td>\n",
       "      <td>70</td>\n",
       "      <td>73</td>\n",
       "      <td>gut</td>\n",
       "      <td>watercourse</td>\n",
       "      <td>ENVO:00000029</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>biolink:AnatomicalEntity</td>\n",
       "      <td>70</td>\n",
       "      <td>73</td>\n",
       "      <td>gut</td>\n",
       "      <td>digestive system</td>\n",
       "      <td>UBERON:0001007</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    BiosampleId                    Biolink  BeginTerm  EndTerm TokenizedTerm  \\\n",
       "0  SAMN02393652  biolink:ChemicalSubstance         19       24         probe   \n",
       "1  SAMN02393652      biolink:OntologyClass         61       65          soil   \n",
       "2  SAMN02393652   biolink:AnatomicalEntity         70       73           gut   \n",
       "3  SAMN02393652      biolink:OntologyClass         70       73           gut   \n",
       "4  SAMN02393652   biolink:AnatomicalEntity         70       73           gut   \n",
       "\n",
       "      PreferredTerm           CURIE S-Type  \n",
       "0             probe     CHEBI:50406     S1  \n",
       "1              soil   ENVO:00001998     S1  \n",
       "2   digestive tract  UBERON:0001555     S1  \n",
       "3       watercourse   ENVO:00000029     S1  \n",
       "4  digestive system  UBERON:0001007     S1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structDataOgerOutput = ogerOutput.merge(structDataIdDF, how='inner', left_on='BiosampleId', right_on='id')\n",
    "structDataOgerOutput = structDataOgerOutput.drop(['id'], axis = 1)\n",
    "print('Number of row: '+ str(len(structDataOgerOutput)))\n",
    "structDataOgerOutput.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Form Phrases for each biosample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6 µs, sys: 1e+03 ns, total: 7 µs\n",
      "Wall time: 11 µs\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Function to phrase-ify\n",
    "def phrasify(dataFrame, columnNames):\n",
    "    count = 0\n",
    "    returnDF = pd.DataFrame(columns=columnNames)\n",
    "    \n",
    "    bsIdList = dataFrame.BiosampleId.unique()\n",
    "    \n",
    "    for i in bsIdList:\n",
    "        count += 1\n",
    "        tmpDF = pd.DataFrame()\n",
    "        bpList = []\n",
    "        phrase1 = []\n",
    "        phrase2 = []\n",
    "        curie = []\n",
    "\n",
    "        tmpDF = dataFrame.loc[(dataFrame.BiosampleId == i)]\n",
    "        \n",
    "        # Filter: Consider ONLY 3 or more DISTINCT terms AND if an RO CURIE is present\n",
    "        \n",
    "        #if len(tmpDF) > 2 \\\n",
    "        #and len(tmpDF.TokenizedTerm.unique()) > 2 \\\n",
    "        #and tmpDF.CURIE.str.contains('RO').any():\n",
    "            \n",
    "        tmpDF.sort_values(by=['BeginTerm'], inplace=True)\n",
    "\n",
    "        for row in tmpDF.iterrows():\n",
    "            bpList.append(row[1].BeginTerm)\n",
    "            phrase1.append(row[1].TokenizedTerm)\n",
    "            phrase2.append(row[1].PreferredTerm)\n",
    "            curie.append(row[1].CURIE)\n",
    "\n",
    "        returnDF = returnDF.append(pd.DataFrame([[i, bpList, phrase1, phrase2, curie]], columns=columnNames))\n",
    "            \n",
    "        if count % 1000 == 0:\n",
    "            print('Number of relevant rows: '+ str(len(returnDF))+ ' - '+ datetime.now().strftime(\"%d/%m/%Y %H:%M:%S\"))\n",
    "        \n",
    "        # *********** FOR DEBUG PURPOSES ONLY ****************\n",
    "        if count == 5000:\n",
    "            break;\n",
    "        \n",
    "    return returnDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****Number of rows: 238 ********\n",
      "CPU times: user 1.04 s, sys: 8.34 ms, total: 1.05 s\n",
      "Wall time: 1.09 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BiosampleId</th>\n",
       "      <th>BeginPosList</th>\n",
       "      <th>Phrase</th>\n",
       "      <th>StdPhrase</th>\n",
       "      <th>CURIE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02393652</td>\n",
       "      <td>[19, 61, 70, 70, 70, 208]</td>\n",
       "      <td>[probe, soil, gut, gut, gut, function]</td>\n",
       "      <td>[probe, soil, digestive tract, watercourse, di...</td>\n",
       "      <td>[CHEBI:50406, ENVO:00001998, UBERON:0001555, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02389811</td>\n",
       "      <td>[20, 48, 179, 189, 202, 202, 221, 221, 241, 25...</td>\n",
       "      <td>[aquarium, mesocosm, part of, flow, system, sy...</td>\n",
       "      <td>[aquarium, mesocosm, part of, mass wasting flo...</td>\n",
       "      <td>[ENVO:00002196, ENVO:01000620, BFO:0000050, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02251077</td>\n",
       "      <td>[65, 98, 188, 198, 216, 231, 259, 259, 273, 273]</td>\n",
       "      <td>[bacteria, lagoon, site, temperate, lagoon, fr...</td>\n",
       "      <td>[Bacteria, lagoon, site, temperate, lagoon, fr...</td>\n",
       "      <td>[NCBITaxon:2, ENVO:00000038, BFO:0000029, ENVO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA2620360</td>\n",
       "      <td>[42, 160, 200, 231, 231]</td>\n",
       "      <td>[seawater, Ocean, all, quality, quality]</td>\n",
       "      <td>[sea water, ocean, root, quality, quality]</td>\n",
       "      <td>[ENVO:00002149, ENVO:00000015, NCBITaxon:1, OB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02009298</td>\n",
       "      <td>[63, 63, 63, 68, 89, 93, 138, 143, 201, 236, 4...</td>\n",
       "      <td>[deep, deep, deep, sea, Sea, rift, halo, strat...</td>\n",
       "      <td>[undersea trough, ocean trench, increased dept...</td>\n",
       "      <td>[ENVO:01000654, ENVO:00000275, OBO:PATO_000159...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    BiosampleId                                       BeginPosList  \\\n",
       "0  SAMN02393652                          [19, 61, 70, 70, 70, 208]   \n",
       "0  SAMN02389811  [20, 48, 179, 189, 202, 202, 221, 221, 241, 25...   \n",
       "0  SAMN02251077   [65, 98, 188, 198, 216, 231, 259, 259, 273, 273]   \n",
       "0  SAMEA2620360                           [42, 160, 200, 231, 231]   \n",
       "0  SAMN02009298  [63, 63, 63, 68, 89, 93, 138, 143, 201, 236, 4...   \n",
       "\n",
       "                                              Phrase  \\\n",
       "0             [probe, soil, gut, gut, gut, function]   \n",
       "0  [aquarium, mesocosm, part of, flow, system, sy...   \n",
       "0  [bacteria, lagoon, site, temperate, lagoon, fr...   \n",
       "0           [seawater, Ocean, all, quality, quality]   \n",
       "0  [deep, deep, deep, sea, Sea, rift, halo, strat...   \n",
       "\n",
       "                                           StdPhrase  \\\n",
       "0  [probe, soil, digestive tract, watercourse, di...   \n",
       "0  [aquarium, mesocosm, part of, mass wasting flo...   \n",
       "0  [Bacteria, lagoon, site, temperate, lagoon, fr...   \n",
       "0         [sea water, ocean, root, quality, quality]   \n",
       "0  [undersea trough, ocean trench, increased dept...   \n",
       "\n",
       "                                               CURIE  \n",
       "0  [CHEBI:50406, ENVO:00001998, UBERON:0001555, E...  \n",
       "0  [ENVO:00002196, ENVO:01000620, BFO:0000050, EN...  \n",
       "0  [NCBITaxon:2, ENVO:00000038, BFO:0000029, ENVO...  \n",
       "0  [ENVO:00002149, ENVO:00000015, NCBITaxon:1, OB...  \n",
       "0  [ENVO:01000654, ENVO:00000275, OBO:PATO_000159...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "pCols = ['BiosampleId', 'BeginPosList', 'Phrase', 'StdPhrase', 'CURIE']\n",
    "phraseDF = phrasify(nhOgerOutput,pCols)\n",
    "print('*****Number of rows: '+ str(len(phraseDF)) + ' ********')\n",
    "phraseDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BiosampleId</th>\n",
       "      <th>BeginPosList</th>\n",
       "      <th>Phrase</th>\n",
       "      <th>StdPhrase</th>\n",
       "      <th>CURIE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN02393652</td>\n",
       "      <td>[19, 61, 70, 70, 70, 208]</td>\n",
       "      <td>[probe, soil, gut, gut, gut, function]</td>\n",
       "      <td>[probe, soil, digestive tract, watercourse, di...</td>\n",
       "      <td>[CHEBI:50406, ENVO:00001998, UBERON:0001555, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN02389811</td>\n",
       "      <td>[20, 48, 179, 189, 202, 202, 221, 221, 241, 25...</td>\n",
       "      <td>[aquarium, mesocosm, part of, flow, system, sy...</td>\n",
       "      <td>[aquarium, mesocosm, part of, mass wasting flo...</td>\n",
       "      <td>[ENVO:00002196, ENVO:01000620, BFO:0000050, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN02251077</td>\n",
       "      <td>[65, 98, 188, 198, 216, 231, 259, 259, 273, 273]</td>\n",
       "      <td>[bacteria, lagoon, site, temperate, lagoon, fr...</td>\n",
       "      <td>[Bacteria, lagoon, site, temperate, lagoon, fr...</td>\n",
       "      <td>[NCBITaxon:2, ENVO:00000038, BFO:0000029, ENVO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMEA2620360</td>\n",
       "      <td>[42, 160, 200, 231, 231]</td>\n",
       "      <td>[seawater, Ocean, all, quality, quality]</td>\n",
       "      <td>[sea water, ocean, root, quality, quality]</td>\n",
       "      <td>[ENVO:00002149, ENVO:00000015, NCBITaxon:1, OB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN02009298</td>\n",
       "      <td>[63, 63, 63, 68, 89, 93, 138, 143, 201, 236, 4...</td>\n",
       "      <td>[deep, deep, deep, sea, Sea, rift, halo, strat...</td>\n",
       "      <td>[undersea trough, ocean trench, increased dept...</td>\n",
       "      <td>[ENVO:01000654, ENVO:00000275, OBO:PATO_000159...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              BiosampleId                                       BeginPosList  \\\n",
       "0  BIOSAMPLE:SAMN02393652                          [19, 61, 70, 70, 70, 208]   \n",
       "0  BIOSAMPLE:SAMN02389811  [20, 48, 179, 189, 202, 202, 221, 221, 241, 25...   \n",
       "0  BIOSAMPLE:SAMN02251077   [65, 98, 188, 198, 216, 231, 259, 259, 273, 273]   \n",
       "0  BIOSAMPLE:SAMEA2620360                           [42, 160, 200, 231, 231]   \n",
       "0  BIOSAMPLE:SAMN02009298  [63, 63, 63, 68, 89, 93, 138, 143, 201, 236, 4...   \n",
       "\n",
       "                                              Phrase  \\\n",
       "0             [probe, soil, gut, gut, gut, function]   \n",
       "0  [aquarium, mesocosm, part of, flow, system, sy...   \n",
       "0  [bacteria, lagoon, site, temperate, lagoon, fr...   \n",
       "0           [seawater, Ocean, all, quality, quality]   \n",
       "0  [deep, deep, deep, sea, Sea, rift, halo, strat...   \n",
       "\n",
       "                                           StdPhrase  \\\n",
       "0  [probe, soil, digestive tract, watercourse, di...   \n",
       "0  [aquarium, mesocosm, part of, mass wasting flo...   \n",
       "0  [Bacteria, lagoon, site, temperate, lagoon, fr...   \n",
       "0         [sea water, ocean, root, quality, quality]   \n",
       "0  [undersea trough, ocean trench, increased dept...   \n",
       "\n",
       "                                               CURIE  \n",
       "0  [CHEBI:50406, ENVO:00001998, UBERON:0001555, E...  \n",
       "0  [ENVO:00002196, ENVO:01000620, BFO:0000050, EN...  \n",
       "0  [NCBITaxon:2, ENVO:00000038, BFO:0000029, ENVO...  \n",
       "0  [ENVO:00002149, ENVO:00000015, NCBITaxon:1, OB...  \n",
       "0  [ENVO:01000654, ENVO:00000275, OBO:PATO_000159...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add 'BIOSAMPLE:' to the id \n",
    "phraseDF.BiosampleId = 'BIOSAMPLE:' + phraseDF.BiosampleId\n",
    "phraseDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 18615\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tot_diss_nitro</th>\n",
       "      <th>salinity_meth</th>\n",
       "      <th>tidal_stage</th>\n",
       "      <th>num_replicons</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>rel_to_oxygen</th>\n",
       "      <th>potassium</th>\n",
       "      <th>sodium</th>\n",
       "      <th>env_package</th>\n",
       "      <th>biomass</th>\n",
       "      <th>...</th>\n",
       "      <th>id</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>taxonomy_name</th>\n",
       "      <th>package</th>\n",
       "      <th>norm_env_package</th>\n",
       "      <th>BiosampleId</th>\n",
       "      <th>BeginPosList</th>\n",
       "      <th>Phrase</th>\n",
       "      <th>StdPhrase</th>\n",
       "      <th>CURIE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706329</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706311</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706320</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706300</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>BIOSAMPLE:SAMEA1706316</td>\n",
       "      <td>449393</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  tot_diss_nitro salinity_meth tidal_stage num_replicons nitrate  \\\n",
       "0            NaN           NaN         NaN           NaN     NaN   \n",
       "1            NaN           NaN         NaN           NaN     NaN   \n",
       "2            NaN           NaN         NaN           NaN     NaN   \n",
       "3            NaN           NaN         NaN           NaN     NaN   \n",
       "4            NaN           NaN         NaN           NaN     NaN   \n",
       "\n",
       "  rel_to_oxygen potassium sodium env_package biomass  ...  \\\n",
       "0           NaN      2.33    NaN       water     NaN  ...   \n",
       "1           NaN      1.33    NaN       water     NaN  ...   \n",
       "2           NaN      2.77    NaN       water     NaN  ...   \n",
       "3           NaN      2.52    NaN       water     NaN  ...   \n",
       "4           NaN       3.4    NaN       water     NaN  ...   \n",
       "\n",
       "                       id taxonomy_id          taxonomy_name      package  \\\n",
       "0  BIOSAMPLE:SAMEA1706329      449393  freshwater metagenome  Generic.1.0   \n",
       "1  BIOSAMPLE:SAMEA1706311      449393  freshwater metagenome  Generic.1.0   \n",
       "2  BIOSAMPLE:SAMEA1706320      449393  freshwater metagenome  Generic.1.0   \n",
       "3  BIOSAMPLE:SAMEA1706300      449393  freshwater metagenome  Generic.1.0   \n",
       "4  BIOSAMPLE:SAMEA1706316      449393  freshwater metagenome  Generic.1.0   \n",
       "\n",
       "  norm_env_package BiosampleId BeginPosList Phrase StdPhrase  CURIE  \n",
       "0            water         NaN          NaN    NaN       NaN    NaN  \n",
       "1            water         NaN          NaN    NaN       NaN    NaN  \n",
       "2            water         NaN          NaN    NaN       NaN    NaN  \n",
       "3            water         NaN          NaN    NaN       NaN    NaN  \n",
       "4            water         NaN          NaN    NaN       NaN    NaN  \n",
       "\n",
       "[5 rows x 102 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#result = ogerInput.merge(phraseDF, how='inner', on='BiosampleId')\n",
    "result = structuredDataDF.merge(phraseDF, how='left', left_on='id', right_on='BiosampleId')\n",
    "print('Number of rows: '+ str(len(result)))\n",
    "result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv('../output/waterSampleWithEnitityPhrases.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
