{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Settings required from user:\n",
    "\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How should the mapping effort be scoped?\n",
    "Additional taxonomic filters may be applied below\n",
    "- soil\n",
    "- sediment\n",
    "- plant-associated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "scoping_col = \"env_package_normalization.EnvPackage\"\n",
    "scoping_value = \"soil\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What Biosample field should be mapped to ontology classes?\n",
    "- env_broad_scale\n",
    "- env_local_scale\n",
    "- env_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_broad_scale\"\n",
    "# biosample_col_to_map = \"env_medium\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# where do we require a single ontology and where can we use multiple?\n",
    "primary_onto_prefix = \"ENVO\"\n",
    "# 'gaz',\n",
    "# need to rank results\n",
    "# additional_preferred_ontologies = [\"ncbitaxon\", \"obi\", \"pato\", \"efo\"]\n",
    "additional_preferred_ontologies = [\"ncbitaxon\", \"po\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file = \"../../target/harmonized-table.db\"\n",
    "ncbitaxon_sqlite_file = \"../../../scoped-mapping/semantic-sql/db/ncbitaxon.db\"\n",
    "\n",
    "# /Users/MAM/Documents/gitrepos/biosample-analysis/target/harmonized-table.db\n",
    "\n",
    "chars_to_whiteout = \"._-\"\n",
    "my_query_fields = \"\"  # OLS weighted default\n",
    "my_row_req = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "salvage_max_dist = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ols_review_file = \"../local/ols_review.tsv\"\n",
    "ols_review_seperator = \"\\t\"\n",
    "\n",
    "strategy_col = \"strategy\"\n",
    "include_col = \"include\"\n",
    "\n",
    "first_pass_include_val = True\n",
    "# first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val = False\n",
    "# salvage_strategy_val    = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = biosample_col_to_map + \"_\" + scoping_col + \"_\" + scoping_value\n",
    "sssom_file = \"../\" + sssom_subject_prefix + \"_SSSOM.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# per_biosample_mapping_results_file = \"per_biosample_scoped_mapping_results.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "primary_ontology_sqlite_file = (\n",
    "    \"../../../scoped-mapping/semantic-sql/db/\" + primary_onto_prefix.lower() + \".db\"\n",
    ")\n",
    "primary_ontology_cnx = sqlite3.connect(primary_ontology_sqlite_file)\n",
    "additional_preferred_ontologies.insert(0, primary_onto_prefix.lower())\n",
    "additional_preferred_ontologies_str = \",\".join(additional_preferred_ontologies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ontology</th>\n",
       "      <th>id_pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO</td>\n",
       "      <td>BFO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CARO</td>\n",
       "      <td>CARO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI</td>\n",
       "      <td>CHEBI:\\d{4,6}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO</td>\n",
       "      <td>ENVO:\\d{7,8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAO</td>\n",
       "      <td>FAO:0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FOODON</td>\n",
       "      <td>FOODON:\\d{8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO</td>\n",
       "      <td>GO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IAO</td>\n",
       "      <td>IAO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NCBITaxon</td>\n",
       "      <td>NCBITaxon:\\d+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OBI</td>\n",
       "      <td>OBI:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PATO</td>\n",
       "      <td>PATO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCO</td>\n",
       "      <td>PCO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PO</td>\n",
       "      <td>PO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RO</td>\n",
       "      <td>RO:0002577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>UBERON</td>\n",
       "      <td>UBERON:\\d{7}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ontology     id_pattern\n",
       "0         BFO      BFO:\\d{7}\n",
       "1        CARO     CARO:\\d{7}\n",
       "2       CHEBI  CHEBI:\\d{4,6}\n",
       "3        ENVO   ENVO:\\d{7,8}\n",
       "4         FAO    FAO:0000001\n",
       "5      FOODON   FOODON:\\d{8}\n",
       "6          GO       GO:\\d{7}\n",
       "7         IAO      IAO:\\d{7}\n",
       "8   NCBITaxon  NCBITaxon:\\d+\n",
       "9         OBI      OBI:\\d{7}\n",
       "10       PATO     PATO:\\d{7}\n",
       "11        PCO      PCO:\\d{7}\n",
       "12         PO       PO:\\d{7}\n",
       "13         RO     RO:0002577\n",
       "14     UBERON   UBERON:\\d{7}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select * from id_patterns\"\n",
    "[id_patterns_frame, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "id_patterns_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_patterns = dict(zip(id_patterns_frame.ontology, id_patterns_frame.id_pattern))\n",
    "# id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- env_package = env_package annotation from NCBI Biosample file XXX\n",
    "- count = number of biosamples using that env_package annotation\n",
    "- lhs = checklist info\n",
    "- rhs = potential package info\n",
    "- extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "- remaining_string = rhs/string, with potential OBO IDs removed\n",
    "- remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "- rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "- EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "- is_canonical = false when EnvPackage is NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('non_bsattribute_columns',), ('harmonized_attrib_pivot',), ('biosample',), ('biosample_sample_taxon_summary',), ('all_tax_labels',), ('ids_labs_selected_ontolgies',), ('id_patterns',), ('package_dictionary',), ('env_package_normalization',), ('repaired_env_package',)]\n"
     ]
    }
   ],
   "source": [
    "cursor = biosample_cnx.cursor()\n",
    "\n",
    "cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "print(cursor.fetchall())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"select * from env_package_normalization\"\n",
    "[env_package_normalization, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "# env_package_normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data. These values were set at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_medium\n",
      "env_package_normalization.EnvPackage\n",
      "soil\n"
     ]
    }
   ],
   "source": [
    "print(biosample_col_to_map)\n",
    "print(scoping_col)\n",
    "print(scoping_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count\n",
       "0                       soil  12596\n",
       "1                       Soil    439\n",
       "2                  peat soil    242\n",
       "3                rhizosphere    238\n",
       "4                  ENVO:soil    214\n",
       "..                       ...    ...\n",
       "105        rhizosphere?soil?      1\n",
       "106              soil sample      1\n",
       "107              soil, water      1\n",
       "108  surface sediment, water      1\n",
       "109                  topsoil      1\n",
       "\n",
       "[110 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = (\n",
    "    \"select \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalization on\n",
    "    b.env_package = env_package_normalization.env_package\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"' group by \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    ")\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   repeated                 splitted  part_count  seq\n",
       "0                      soil                     soil           1    1\n",
       "0                      Soil                     Soil           1    1\n",
       "0                 peat soil                peat soil           1    1\n",
       "0               rhizosphere              rhizosphere           1    1\n",
       "0                 ENVO:soil                ENVO:soil           1    1\n",
       "..                      ...                      ...         ...  ...\n",
       "0         rhizosphere?soil?        rhizosphere?soil?           1    1\n",
       "0               soil sample              soil sample           1    1\n",
       "0               soil, water              soil, water           1    1\n",
       "0   surface sediment, water  surface sediment, water           1    1\n",
       "0                   topsoil                  topsoil           1    1\n",
       "\n",
       "[110 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name=\"PandaZ\"):\n",
    "    split_check = row[1]\n",
    "    if split_check is None:\n",
    "        split_check = \"\"\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    seq_list = list(range(1, splitted_count + 1))\n",
    "    as_frame[\"part_count\"] = splitted_count\n",
    "    as_frame[\"seq\"] = seq_list\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[[\"repeated\", \"splitted\", \"part_count\", \"seq\"]]\n",
    "concat_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    concat_frame, left_on=biosample_col_to_map, right_on=\"repeated\", how=\"left\"\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `OBO` IDs have been entered\n",
    "In the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # this helps by standardizing what could be an ID\n",
    "# but if the RHS is text, then envo should just be removed\n",
    "mapping_candidates[\"term_tidy\"] = mapping_candidates.splitted.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\",\n",
    "    primary_onto_prefix + \":\",\n",
    "    regex=True,\n",
    "    case=False,\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td></td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td></td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td></td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td></td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count                 repeated  \\\n",
       "0                       soil  12596                     soil   \n",
       "1                       Soil    439                     Soil   \n",
       "2                  peat soil    242                peat soil   \n",
       "3                rhizosphere    238              rhizosphere   \n",
       "4                  ENVO:soil    214                ENVO:soil   \n",
       "..                       ...    ...                      ...   \n",
       "105        rhizosphere?soil?      1        rhizosphere?soil?   \n",
       "106              soil sample      1              soil sample   \n",
       "107              soil, water      1              soil, water   \n",
       "108  surface sediment, water      1  surface sediment, water   \n",
       "109                  topsoil      1                  topsoil   \n",
       "\n",
       "                    splitted  part_count  seq                term_tidy  \\\n",
       "0                       soil           1    1                     soil   \n",
       "1                       Soil           1    1                     Soil   \n",
       "2                  peat soil           1    1                peat soil   \n",
       "3                rhizosphere           1    1              rhizosphere   \n",
       "4                  ENVO:soil           1    1                ENVO:soil   \n",
       "..                       ...         ...  ...                      ...   \n",
       "105        rhizosphere?soil?           1    1        rhizosphere?soil?   \n",
       "106              soil sample           1    1              soil sample   \n",
       "107              soil, water           1    1              soil, water   \n",
       "108  surface sediment, water           1    1  surface sediment, water   \n",
       "109                  topsoil           1    1                  topsoil   \n",
       "\n",
       "                      string extract         remaining_string  \\\n",
       "0                       soil                             soil   \n",
       "1                       Soil                             Soil   \n",
       "2                  peat soil                        peat soil   \n",
       "3                rhizosphere                      rhizosphere   \n",
       "4                  ENVO:soil                        ENVO:soil   \n",
       "..                       ...     ...                      ...   \n",
       "105        rhizosphere?soil?                rhizosphere?soil?   \n",
       "106              soil sample                      soil sample   \n",
       "107              soil, water                      soil, water   \n",
       "108  surface sediment, water          surface sediment, water   \n",
       "109                  topsoil                          topsoil   \n",
       "\n",
       "           remaining_tidied  \n",
       "0                      soil  \n",
       "1                      soil  \n",
       "2                 peat soil  \n",
       "3               rhizosphere  \n",
       "4                      soil  \n",
       "..                      ...  \n",
       "105        rhizosphere soil  \n",
       "106             soil sample  \n",
       "107              soil water  \n",
       "108  surface sediment water  \n",
       "109                 topsoil  \n",
       "\n",
       "[110 rows x 11 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(\n",
    "    mapping_candidates[\"term_tidy\"], id_patterns[primary_onto_prefix]\n",
    ")\n",
    "\n",
    "# candidate_series_decomposition\n",
    "\n",
    "mapping_candidates = pd.concat(\n",
    "    [mapping_candidates, candidate_series_decomposition], axis=1\n",
    ")\n",
    "\n",
    "mapping_candidates[\n",
    "    \"remaining_tidied\"\n",
    "] = mapping_candidates.remaining_tidied.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\", \"\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "_Start by connecting to the rdftab database from which the terms and label-like annotations will be obtained?_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000001</td>\n",
       "      <td>entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000002</td>\n",
       "      <td>continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFO:0000003</td>\n",
       "      <td>occurrent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFO:0000004</td>\n",
       "      <td>independent continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFO:0000006</td>\n",
       "      <td>spatial region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13129</th>\n",
       "      <td>UBERON:3000972</td>\n",
       "      <td>head external integument structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13130</th>\n",
       "      <td>UBERON:3000977</td>\n",
       "      <td>body external integument structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13131</th>\n",
       "      <td>UBERON:3010200</td>\n",
       "      <td>vasculature of respiratory integument</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13132</th>\n",
       "      <td>UBERON:4100000</td>\n",
       "      <td>skeletal element projection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13133</th>\n",
       "      <td>UBERON:8410024</td>\n",
       "      <td>intestinal junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13134 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              subject                                  value\n",
       "0         BFO:0000001                                 entity\n",
       "1         BFO:0000002                             continuant\n",
       "2         BFO:0000003                              occurrent\n",
       "3         BFO:0000004                 independent continuant\n",
       "4         BFO:0000006                         spatial region\n",
       "...               ...                                    ...\n",
       "13129  UBERON:3000972     head external integument structure\n",
       "13130  UBERON:3000977     body external integument structure\n",
       "13131  UBERON:3010200  vasculature of respiratory integument\n",
       "13132  UBERON:4100000            skeletal element projection\n",
       "13133  UBERON:8410024                    intestinal junction\n",
       "\n",
       "[13134 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select stanza as subject, value from ids_labs_selected_ontolgies\"\n",
    "\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td></td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td></td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td></td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td></td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count                 repeated  \\\n",
       "0                       soil  12596                     soil   \n",
       "1                       Soil    439                     Soil   \n",
       "2                  peat soil    242                peat soil   \n",
       "3                rhizosphere    238              rhizosphere   \n",
       "4                  ENVO:soil    214                ENVO:soil   \n",
       "..                       ...    ...                      ...   \n",
       "110        rhizosphere?soil?      1        rhizosphere?soil?   \n",
       "111              soil sample      1              soil sample   \n",
       "112              soil, water      1              soil, water   \n",
       "113  surface sediment, water      1  surface sediment, water   \n",
       "114                  topsoil      1                  topsoil   \n",
       "\n",
       "                    splitted  part_count  seq                term_tidy  \\\n",
       "0                       soil           1    1                     soil   \n",
       "1                       Soil           1    1                     Soil   \n",
       "2                  peat soil           1    1                peat soil   \n",
       "3                rhizosphere           1    1              rhizosphere   \n",
       "4                  ENVO:soil           1    1                ENVO:soil   \n",
       "..                       ...         ...  ...                      ...   \n",
       "110        rhizosphere?soil?           1    1        rhizosphere?soil?   \n",
       "111              soil sample           1    1              soil sample   \n",
       "112              soil, water           1    1              soil, water   \n",
       "113  surface sediment, water           1    1  surface sediment, water   \n",
       "114                  topsoil           1    1                  topsoil   \n",
       "\n",
       "                      string extract         remaining_string  \\\n",
       "0                       soil                             soil   \n",
       "1                       Soil                             Soil   \n",
       "2                  peat soil                        peat soil   \n",
       "3                rhizosphere                      rhizosphere   \n",
       "4                  ENVO:soil                        ENVO:soil   \n",
       "..                       ...     ...                      ...   \n",
       "110        rhizosphere?soil?                rhizosphere?soil?   \n",
       "111              soil sample                      soil sample   \n",
       "112              soil, water                      soil, water   \n",
       "113  surface sediment, water          surface sediment, water   \n",
       "114                  topsoil                          topsoil   \n",
       "\n",
       "           remaining_tidied subject value  \n",
       "0                      soil     NaN   NaN  \n",
       "1                      soil     NaN   NaN  \n",
       "2                 peat soil     NaN   NaN  \n",
       "3               rhizosphere     NaN   NaN  \n",
       "4                      soil     NaN   NaN  \n",
       "..                      ...     ...   ...  \n",
       "110        rhizosphere soil     NaN   NaN  \n",
       "111             soil sample     NaN   NaN  \n",
       "112              soil water     NaN   NaN  \n",
       "113  surface sediment water     NaN   NaN  \n",
       "114                 topsoil     NaN   NaN  \n",
       "\n",
       "[115 rows x 13 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"extract\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td></td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td></td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil water</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td></td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td></td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count                 repeated  \\\n",
       "0                       soil  12596                     soil   \n",
       "1                       Soil    439                     Soil   \n",
       "2                  peat soil    242                peat soil   \n",
       "3                rhizosphere    238              rhizosphere   \n",
       "4                  ENVO:soil    214                ENVO:soil   \n",
       "..                       ...    ...                      ...   \n",
       "110        rhizosphere?soil?      1        rhizosphere?soil?   \n",
       "111              soil sample      1              soil sample   \n",
       "112              soil, water      1              soil, water   \n",
       "113  surface sediment, water      1  surface sediment, water   \n",
       "114                  topsoil      1                  topsoil   \n",
       "\n",
       "                    splitted  part_count  seq                term_tidy  \\\n",
       "0                       soil           1    1                     soil   \n",
       "1                       Soil           1    1                     Soil   \n",
       "2                  peat soil           1    1                peat soil   \n",
       "3                rhizosphere           1    1              rhizosphere   \n",
       "4                  ENVO:soil           1    1                ENVO:soil   \n",
       "..                       ...         ...  ...                      ...   \n",
       "110        rhizosphere?soil?           1    1        rhizosphere?soil?   \n",
       "111              soil sample           1    1              soil sample   \n",
       "112              soil, water           1    1              soil, water   \n",
       "113  surface sediment, water           1    1  surface sediment, water   \n",
       "114                  topsoil           1    1                  topsoil   \n",
       "\n",
       "                      string extract         remaining_string  \\\n",
       "0                       soil                             soil   \n",
       "1                       Soil                             Soil   \n",
       "2                  peat soil                        peat soil   \n",
       "3                rhizosphere                      rhizosphere   \n",
       "4                  ENVO:soil                        ENVO:soil   \n",
       "..                       ...     ...                      ...   \n",
       "110        rhizosphere?soil?                rhizosphere?soil?   \n",
       "111              soil sample                      soil sample   \n",
       "112              soil, water                      soil, water   \n",
       "113  surface sediment, water          surface sediment, water   \n",
       "114                  topsoil                          topsoil   \n",
       "\n",
       "           remaining_tidied subject value  cosine  \n",
       "0                      soil     NaN           1.0  \n",
       "1                      soil     NaN           1.0  \n",
       "2                 peat soil     NaN           1.0  \n",
       "3               rhizosphere     NaN           1.0  \n",
       "4                      soil     NaN           1.0  \n",
       "..                      ...     ...   ...     ...  \n",
       "110        rhizosphere soil     NaN           1.0  \n",
       "111             soil sample     NaN           1.0  \n",
       "112              soil water     NaN           1.0  \n",
       "113  surface sediment water     NaN           1.0  \n",
       "114                 topsoil     NaN           1.0  \n",
       "\n",
       "[115 rows x 14 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates[\"value\"] = mapping_candidates[\"value\"].fillna(\"\")\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates[\"remaining_tidied\"].fillna(\n",
    "    \"\"\n",
    ")\n",
    "mapping_candidates[\"cosine\"] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(\n",
    "        my_row[\"remaining_tidied\"].lower(), my_row[\"value\"].lower()\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names are meaningful and unique\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td></td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td></td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil water</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td></td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td></td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count                 repeated  \\\n",
       "0                       soil  12596                     soil   \n",
       "1                       soil  12596                     soil   \n",
       "2                       Soil    439                     Soil   \n",
       "3                       Soil    439                     Soil   \n",
       "4                  peat soil    242                peat soil   \n",
       "..                       ...    ...                      ...   \n",
       "151        rhizosphere?soil?      1        rhizosphere?soil?   \n",
       "152              soil sample      1              soil sample   \n",
       "153              soil, water      1              soil, water   \n",
       "154  surface sediment, water      1  surface sediment, water   \n",
       "155                  topsoil      1                  topsoil   \n",
       "\n",
       "                    splitted  part_count  seq                term_tidy  \\\n",
       "0                       soil           1    1                     soil   \n",
       "1                       soil           1    1                     soil   \n",
       "2                       Soil           1    1                     Soil   \n",
       "3                       Soil           1    1                     Soil   \n",
       "4                  peat soil           1    1                peat soil   \n",
       "..                       ...         ...  ...                      ...   \n",
       "151        rhizosphere?soil?           1    1        rhizosphere?soil?   \n",
       "152              soil sample           1    1              soil sample   \n",
       "153              soil, water           1    1              soil, water   \n",
       "154  surface sediment, water           1    1  surface sediment, water   \n",
       "155                  topsoil           1    1                  topsoil   \n",
       "\n",
       "                      string extract         remaining_string  \\\n",
       "0                       soil                             soil   \n",
       "1                       soil                             soil   \n",
       "2                       Soil                             Soil   \n",
       "3                       Soil                             Soil   \n",
       "4                  peat soil                        peat soil   \n",
       "..                       ...     ...                      ...   \n",
       "151        rhizosphere?soil?                rhizosphere?soil?   \n",
       "152              soil sample                      soil sample   \n",
       "153              soil, water                      soil, water   \n",
       "154  surface sediment, water          surface sediment, water   \n",
       "155                  topsoil                          topsoil   \n",
       "\n",
       "           remaining_tidied term_id lab_from_id  lfi_cosine term_id_from_lab  \\\n",
       "0                      soil     NaN                     1.0    ENVO:00001998   \n",
       "1                      soil     NaN                     1.0    ENVO:00001998   \n",
       "2                      soil     NaN                     1.0    ENVO:00001998   \n",
       "3                      soil     NaN                     1.0    ENVO:00001998   \n",
       "4                 peat soil     NaN                     1.0    ENVO:00005774   \n",
       "..                      ...     ...         ...         ...              ...   \n",
       "151        rhizosphere soil     NaN                     1.0              NaN   \n",
       "152             soil sample     NaN                     1.0              NaN   \n",
       "153              soil water     NaN                     1.0              NaN   \n",
       "154  surface sediment water     NaN                     1.0              NaN   \n",
       "155                 topsoil     NaN                     1.0              NaN   \n",
       "\n",
       "         value  \n",
       "0         soil  \n",
       "1         soil  \n",
       "2         soil  \n",
       "3         soil  \n",
       "4    peat soil  \n",
       "..         ...  \n",
       "151        NaN  \n",
       "152        NaN  \n",
       "153        NaN  \n",
       "154        NaN  \n",
       "155        NaN  \n",
       "\n",
       "[156 rows x 16 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "]\n",
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"remaining_tidied\", right_on=\"value\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "    \"term_id_from_lab\",\n",
    "    \"value\",\n",
    "]\n",
    "\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.drop_duplicates()\n",
    "\n",
    "mc_tifl_split = mapping_candidates[\"term_id_from_lab\"].str.split(\":\")\n",
    "\n",
    "import math\n",
    "\n",
    "prefix_when_possible = []\n",
    "for one_split in mc_tifl_split:\n",
    "    if type(one_split) == list:\n",
    "        os_first = one_split[0]\n",
    "        prefix_when_possible.append(os_first)\n",
    "    else:\n",
    "        prefix_when_possible.append(\"\")\n",
    "\n",
    "prefix_when_possible = [each_string.lower() for each_string in prefix_when_possible]\n",
    "\n",
    "# prefix_when_possible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/biosample-analysis/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1596: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/Users/MAM/Documents/gitrepos/biosample-analysis/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1745: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n"
     ]
    }
   ],
   "source": [
    "if \"onto_prefix\" in mapping_candidates.columns:\n",
    "    mapping_candidates = mapping_candidates.drop(\"onto_prefix\", 1)\n",
    "\n",
    "mapping_candidates.loc[:, \"onto_prefix\"] = pd.Series(prefix_when_possible).values\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>envo</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>po</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  onto_prefix  rank\n",
       "0        envo     0\n",
       "1   ncbitaxon     1\n",
       "2          po     2"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranks = list(range(len(additional_preferred_ontologies)))\n",
    "\n",
    "d = {\"onto_prefix\": additional_preferred_ontologies, \"rank\": ranks}\n",
    "onto_rank_frame = pd.DataFrame(d)\n",
    "\n",
    "onto_rank_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_rank_frame, on=\"onto_prefix\", how=\"left\"\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_maps = mapping_candidates.groupby(biosample_col_to_map)[\"rank\"].min().reset_index()\n",
    "# best_maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    best_maps, on=[biosample_col_to_map, \"rank\"]\n",
    ")\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates[\"consensus_id\"] = mapping_candidates[\"term_id_from_lab\"]\n",
    "mapping_candidates[\"consensus_lab\"] = mapping_candidates[\"value\"]\n",
    "\n",
    "flag = mapping_candidates[\"consensus_id\"].isnull() & (\n",
    "    (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates.lfi_cosine.le(my_max_string_dist)\n",
    "    )\n",
    "    | (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates[\"remaining_tidied\"].eq(\"\")\n",
    "    )\n",
    ")\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"term_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_id\"] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"lab_from_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_lab\"] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~flag\n",
    "mapping_candidates[\"id_or_lab_ok\"] = antiflag\n",
    "\n",
    "mapping_candidates[\"assembled_consensus\"] = (\n",
    "    mapping_candidates[\"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + mapping_candidates[\"consensus_id\"]\n",
    "    + \"]\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>...</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12596</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>242</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>True</td>\n",
       "      <td>peat soil [ENVO:00005774]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>True</td>\n",
       "      <td>rhizosphere [ENVO:00005801]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere?soil?</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil sample</td>\n",
       "      <td>soil sample</td>\n",
       "      <td></td>\n",
       "      <td>soil sample</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>soil, water</td>\n",
       "      <td>soil, water</td>\n",
       "      <td></td>\n",
       "      <td>soil, water</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td></td>\n",
       "      <td>surface sediment, water</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td></td>\n",
       "      <td>topsoil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_medium  count                 repeated  \\\n",
       "0                       soil  12596                     soil   \n",
       "1                       Soil    439                     Soil   \n",
       "2                  peat soil    242                peat soil   \n",
       "3                rhizosphere    238              rhizosphere   \n",
       "4                  ENVO:soil    214                ENVO:soil   \n",
       "..                       ...    ...                      ...   \n",
       "105        rhizosphere?soil?      1        rhizosphere?soil?   \n",
       "106              soil sample      1              soil sample   \n",
       "107              soil, water      1              soil, water   \n",
       "108  surface sediment, water      1  surface sediment, water   \n",
       "109                  topsoil      1                  topsoil   \n",
       "\n",
       "                    splitted  part_count  seq                term_tidy  \\\n",
       "0                       soil           1    1                     soil   \n",
       "1                       Soil           1    1                     Soil   \n",
       "2                  peat soil           1    1                peat soil   \n",
       "3                rhizosphere           1    1              rhizosphere   \n",
       "4                  ENVO:soil           1    1                ENVO:soil   \n",
       "..                       ...         ...  ...                      ...   \n",
       "105        rhizosphere?soil?           1    1        rhizosphere?soil?   \n",
       "106              soil sample           1    1              soil sample   \n",
       "107              soil, water           1    1              soil, water   \n",
       "108  surface sediment, water           1    1  surface sediment, water   \n",
       "109                  topsoil           1    1                  topsoil   \n",
       "\n",
       "                      string extract         remaining_string  ...  \\\n",
       "0                       soil                             soil  ...   \n",
       "1                       Soil                             Soil  ...   \n",
       "2                  peat soil                        peat soil  ...   \n",
       "3                rhizosphere                      rhizosphere  ...   \n",
       "4                  ENVO:soil                        ENVO:soil  ...   \n",
       "..                       ...     ...                      ...  ...   \n",
       "105        rhizosphere?soil?                rhizosphere?soil?  ...   \n",
       "106              soil sample                      soil sample  ...   \n",
       "107              soil, water                      soil, water  ...   \n",
       "108  surface sediment, water          surface sediment, water  ...   \n",
       "109                  topsoil                          topsoil  ...   \n",
       "\n",
       "    lab_from_id lfi_cosine term_id_from_lab        value onto_prefix rank  \\\n",
       "0                      1.0    ENVO:00001998         soil        envo  0.0   \n",
       "1                      1.0    ENVO:00001998         soil        envo  0.0   \n",
       "2                      1.0    ENVO:00005774    peat soil        envo  0.0   \n",
       "3                      1.0    ENVO:00005801  rhizosphere        envo  0.0   \n",
       "4                      1.0    ENVO:00001998         soil        envo  0.0   \n",
       "..          ...        ...              ...          ...         ...  ...   \n",
       "105                    1.0              NaN          NaN              NaN   \n",
       "106                    1.0              NaN          NaN              NaN   \n",
       "107                    1.0              NaN          NaN              NaN   \n",
       "108                    1.0              NaN          NaN              NaN   \n",
       "109                    1.0              NaN          NaN              NaN   \n",
       "\n",
       "      consensus_id  consensus_lab id_or_lab_ok          assembled_consensus  \n",
       "0    ENVO:00001998           soil         True         soil [ENVO:00001998]  \n",
       "1    ENVO:00001998           soil         True         soil [ENVO:00001998]  \n",
       "2    ENVO:00005774      peat soil         True    peat soil [ENVO:00005774]  \n",
       "3    ENVO:00005801    rhizosphere         True  rhizosphere [ENVO:00005801]  \n",
       "4    ENVO:00001998           soil         True         soil [ENVO:00001998]  \n",
       "..             ...            ...          ...                          ...  \n",
       "105            NaN            NaN        False                          NaN  \n",
       "106            NaN            NaN        False                          NaN  \n",
       "107            NaN            NaN        False                          NaN  \n",
       "108            NaN            NaN        False                          NaN  \n",
       "109            NaN            NaN        False                          NaN  \n",
       "\n",
       "[110 rows x 22 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>162</td>\n",
       "      <td>organic soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>113</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>99</td>\n",
       "      <td>dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>90</td>\n",
       "      <td>mine tailings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>80</td>\n",
       "      <td>mineral soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>1</td>\n",
       "      <td>mud soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>1</td>\n",
       "      <td>paddy soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>1</td>\n",
       "      <td>replicate 1 soil from baodongyu of taoyuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1</td>\n",
       "      <td>replicate 1 soil from gushi of taoyuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    count                            remaining_tidied\n",
       "0     162                                organic soil\n",
       "1     113                                            \n",
       "2      99                                         dna\n",
       "3      90                               mine tailings\n",
       "4      80                                mineral soil\n",
       "..    ...                                         ...\n",
       "60      1                                    mud soil\n",
       "61      1                                  paddy soil\n",
       "62      1  replicate 1 soil from baodongyu of taoyuan\n",
       "63      1      replicate 1 soil from gushi of taoyuan\n",
       "64      1                                     topsoil\n",
       "\n",
       "[65 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = ~mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[[\"remaining_tidied\", \"count\"]]\n",
    "\n",
    "sum_by_needed = needs_search.groupby(\"remaining_tidied\")[\"count\"].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed[\"remaining_tidied\"] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values(\"count\", ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "# sum_by_needed.to_sql(\"sum_by_needed\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed[\"remaining_tidied\"])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per unique post-tidied submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n",
    "`datetime.now().replace(microsecond=0).isoformat()` ?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-05T19:10:38\n",
      "65\n",
      "1.0833333333333333\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now().replace(microsecond=0).isoformat())\n",
    "print(len(sum_by_needed.index))\n",
    "print((len(sum_by_needed.index)) / 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    ebs_raw_list,\n",
    "    bad_chars=chars_to_whiteout,\n",
    "    cat_name=biosample_col_to_map,\n",
    "    ontoprefix=additional_preferred_ontologies_str,\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=my_string_dist_arg,\n",
    ")\n",
    "# ebs_search_res.to_sql(\n",
    "#     \"ols_search_results\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       category            raw          query          name  string_dist_rank  \\\n",
       "45   env_medium       bacteria       bacteria      Bacteria                 1   \n",
       "305  env_medium  mine tailings  mine tailings  mine tailing                 1   \n",
       "425  env_medium           peat           peat          peat                 1   \n",
       "553  env_medium     sandy soil     sandy soil    sandy soil                 1   \n",
       "\n",
       "     string_dist         obo_id         label  search_rank ontology_prefix  \\\n",
       "45         0.000    NCBITaxon:2      Bacteria            1       NCBITAXON   \n",
       "305        0.036  ENVO:00000003  mine tailing            4            ENVO   \n",
       "425        0.000  ENVO:00005774     peat soil            1            ENVO   \n",
       "553        0.000  ENVO:00002229      arenosol            1            ENVO   \n",
       "\n",
       "                 scope     type                                           iri  \\\n",
       "45               label    label    http://purl.obolibrary.org/obo/NCBITaxon_2   \n",
       "305              label    label  http://purl.obolibrary.org/obo/ENVO_00000003   \n",
       "425  hasRelatedSynonym     None  http://purl.obolibrary.org/obo/ENVO_00005774   \n",
       "553            synonym  synonym  http://purl.obolibrary.org/obo/ENVO_00002229   \n",
       "\n",
       "    ontology_name  \n",
       "45      ncbitaxon  \n",
       "305          envo  \n",
       "425          envo  \n",
       "553          envo  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    ebs_search_res, max_string_dist=my_max_string_dist\n",
    ")\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_medium</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>bacterium A9 13-13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.541</td>\n",
       "      <td>NCBITaxon:1469932</td>\n",
       "      <td>bacterium A9 13-13</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1469932</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.622</td>\n",
       "      <td>NCBITaxon:12424</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_12424</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Pasteurellaceae bacterium 13/13</td>\n",
       "      <td>3</td>\n",
       "      <td>0.667</td>\n",
       "      <td>NCBITaxon:1347399</td>\n",
       "      <td>Pasteurellaceae bacterium 13/13</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1347399</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Bacteriophage 13</td>\n",
       "      <td>4</td>\n",
       "      <td>0.742</td>\n",
       "      <td>NCBITaxon:12424</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>equivalent name</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_12424</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>soil erosion</td>\n",
       "      <td>1</td>\n",
       "      <td>0.631</td>\n",
       "      <td>ENVO:01000706</td>\n",
       "      <td>soil erosion</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000706</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>luvisol</td>\n",
       "      <td>2</td>\n",
       "      <td>0.833</td>\n",
       "      <td>ENVO:00002248</td>\n",
       "      <td>luvisol</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002248</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>lixisol</td>\n",
       "      <td>3</td>\n",
       "      <td>0.833</td>\n",
       "      <td>ENVO:00002242</td>\n",
       "      <td>lixisol</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002242</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>stagnosol</td>\n",
       "      <td>4</td>\n",
       "      <td>0.856</td>\n",
       "      <td>ENVO:00002274</td>\n",
       "      <td>stagnosol</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002274</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>mineral horizon</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>ENVO:03600011</td>\n",
       "      <td>mineral horizon</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03600011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>556 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       category      raw    query                             name  \\\n",
       "0    env_medium                                                      \n",
       "1    env_medium       13       13               bacterium A9 13-13   \n",
       "3    env_medium       13       13                         Phage 13   \n",
       "2    env_medium       13       13  Pasteurellaceae bacterium 13/13   \n",
       "4    env_medium       13       13                 Bacteriophage 13   \n",
       "..          ...      ...      ...                              ...   \n",
       "646  env_medium  topsoil  topsoil                     soil erosion   \n",
       "647  env_medium  topsoil  topsoil                          luvisol   \n",
       "648  env_medium  topsoil  topsoil                          lixisol   \n",
       "645  env_medium  topsoil  topsoil                        stagnosol   \n",
       "644  env_medium  topsoil  topsoil                  mineral horizon   \n",
       "\n",
       "     string_dist_rank  string_dist             obo_id  \\\n",
       "0                   1        0.000                      \n",
       "1                   1        0.541  NCBITaxon:1469932   \n",
       "3                   2        0.622    NCBITaxon:12424   \n",
       "2                   3        0.667  NCBITaxon:1347399   \n",
       "4                   4        0.742    NCBITaxon:12424   \n",
       "..                ...          ...                ...   \n",
       "646                 1        0.631      ENVO:01000706   \n",
       "647                 2        0.833      ENVO:00002248   \n",
       "648                 3        0.833      ENVO:00002242   \n",
       "645                 4        0.856      ENVO:00002274   \n",
       "644                 5        1.000      ENVO:03600011   \n",
       "\n",
       "                               label  search_rank ontology_prefix  \\\n",
       "0                                               1                   \n",
       "1                 bacterium A9 13-13            1       NCBITAXON   \n",
       "3                           Phage 13            3       NCBITAXON   \n",
       "2    Pasteurellaceae bacterium 13/13            2       NCBITAXON   \n",
       "4                           Phage 13            3       NCBITAXON   \n",
       "..                               ...          ...             ...   \n",
       "646                     soil erosion            3            ENVO   \n",
       "647                          luvisol            4            ENVO   \n",
       "648                          lixisol            5            ENVO   \n",
       "645                        stagnosol            2            ENVO   \n",
       "644                  mineral horizon            1            ENVO   \n",
       "\n",
       "               scope             type  \\\n",
       "0                NaN              NaN   \n",
       "1              label            label   \n",
       "3              label            label   \n",
       "2              label            label   \n",
       "4    hasExactSynonym  equivalent name   \n",
       "..               ...              ...   \n",
       "646            label            label   \n",
       "647            label            label   \n",
       "648            label            label   \n",
       "645            label            label   \n",
       "644            label            label   \n",
       "\n",
       "                                                  iri ontology_name  \n",
       "0                                                                    \n",
       "1    http://purl.obolibrary.org/obo/NCBITaxon_1469932     ncbitaxon  \n",
       "3      http://purl.obolibrary.org/obo/NCBITaxon_12424     ncbitaxon  \n",
       "2    http://purl.obolibrary.org/obo/NCBITaxon_1347399     ncbitaxon  \n",
       "4      http://purl.obolibrary.org/obo/NCBITaxon_12424     ncbitaxon  \n",
       "..                                                ...           ...  \n",
       "646      http://purl.obolibrary.org/obo/ENVO_01000706          envo  \n",
       "647      http://purl.obolibrary.org/obo/ENVO_00002248          envo  \n",
       "648      http://purl.obolibrary.org/obo/ENVO_00002242          envo  \n",
       "645      http://purl.obolibrary.org/obo/ENVO_00002274          envo  \n",
       "644      http://purl.obolibrary.org/obo/ENVO_03600011          envo  \n",
       "\n",
       "[556 rows x 14 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(\n",
    "    ebs_search_res, my_best_acceptable\n",
    ")\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS\n",
    "what would it be like to use a local OLS instance? (latency-wise)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['',\n",
       " '13',\n",
       " '15',\n",
       " '16s rrna gene sequence',\n",
       " '22',\n",
       " '24',\n",
       " '4',\n",
       " '6',\n",
       " 'alfisols',\n",
       " 'biomat',\n",
       " 'cryoconite',\n",
       " 'decomposed litter oe layer',\n",
       " 'dna',\n",
       " 'fermentation humic sustrate',\n",
       " 'forest soil biochar',\n",
       " 'genome',\n",
       " 'genomic bacterial dna of the ssu rrna gene v5 v7 hypervariable region',\n",
       " 'hypolith scappings from under rocks',\n",
       " 'juvenile from soil aluvial loam',\n",
       " 'juvenile from soil diluvial sand',\n",
       " 'juvenile from soil loess loam',\n",
       " 'juveniles meloidogyne hapla from soil diluvial sand',\n",
       " 'leaves',\n",
       " 'litters',\n",
       " 'mineral soil',\n",
       " 'mixed fungi',\n",
       " 'molecular ecology sample w1 3',\n",
       " 'molecular ecology sample w1 6',\n",
       " 'molecular ecology sample x1 3',\n",
       " 'molecular ecology sample x1 6',\n",
       " 'moss',\n",
       " 'mud soil',\n",
       " 'not applicable',\n",
       " 'organic soil',\n",
       " 'paddy soil',\n",
       " 'permafrost soil',\n",
       " 'picea litter',\n",
       " 'pinus litter',\n",
       " 'replicate 1 soil from baodongyu of taoyuan',\n",
       " 'replicate 1 soil from gushi of taoyuan',\n",
       " 'replicate 1 soil from jiaxing',\n",
       " 'replicate 1 soil from leizhou',\n",
       " 'replicate 1 soil from yingtan',\n",
       " 'replicate 2 soil from baodongyu of taoyuan',\n",
       " 'replicate 2 soil from gushi of taoyuan',\n",
       " 'replicate 2 soil from jiaxing',\n",
       " 'replicate 2 soil from leizhou',\n",
       " 'replicate 2 soil from yingtan',\n",
       " 'rhizosphere soil',\n",
       " 'rna',\n",
       " 'soil aluvial loam',\n",
       " 'soil and ice',\n",
       " 'soil collected during dry season',\n",
       " 'soil collected during rainy season',\n",
       " 'soil core',\n",
       " 'soil dna',\n",
       " 'soil sample',\n",
       " 'soil water',\n",
       " 'soybean soil',\n",
       " 'surface sediment water',\n",
       " 'topsoil']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings[\"raw\"])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "print(len(still_unmapped))\n",
    "\n",
    "still_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-05T19:11:35\n",
      "65\n",
      "1.0833333333333333\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now().replace(microsecond=0).isoformat())\n",
    "print(len(sum_by_needed.index))\n",
    "print((len(sum_by_needed.index)) / 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    still_unmapped,\n",
    "    bad_chars=\"._-\",\n",
    "    cat_name=\"salvage\",\n",
    "    ontoprefix=\"\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=2,\n",
    ")\n",
    "\n",
    "# salvage_search_res.to_sql(\n",
    "#     \"salvage_search_res\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns\n",
    "At the very least, will require some review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172971</td>\n",
       "      <td>Worry Score 4</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172971</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>salvage</td>\n",
       "      <td>dna</td>\n",
       "      <td>dna</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C449</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C449</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome</td>\n",
       "      <td>genome</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C16629</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C16629</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>salvage</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>CO_366:0000273</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>CO_366</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_366:0000273</td>\n",
       "      <td>co_366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>salvage</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00000134</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000134</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005801</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0013221</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0013221</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>salvage</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>1</td>\n",
       "      <td>0.155</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    category                     raw                   query  \\\n",
       "11   salvage                      15                      15   \n",
       "65   salvage                       4                       4   \n",
       "92   salvage                       6                       6   \n",
       "111  salvage              cryoconite              cryoconite   \n",
       "129  salvage                     dna                     dna   \n",
       "166  salvage                  genome                  genome   \n",
       "295  salvage                  leaves                  leaves   \n",
       "524  salvage          not applicable          not applicable   \n",
       "580  salvage              paddy soil              paddy soil   \n",
       "592  salvage         permafrost soil         permafrost soil   \n",
       "778  salvage        rhizosphere soil        rhizosphere soil   \n",
       "789  salvage                     rna                     rna   \n",
       "968  salvage  surface sediment water  surface sediment water   \n",
       "\n",
       "                 name  string_dist_rank  string_dist          obo_id  \\\n",
       "11                 15                 1        0.000    NCIT:C113429   \n",
       "65                  4                 1        0.000    NCIT:C172971   \n",
       "92                  6                 1        0.000    NCIT:C179016   \n",
       "111   cryoconite hole                 1        0.198   ENVO:03000039   \n",
       "129               DNA                 1        0.000       NCIT:C449   \n",
       "166            Genome                 1        0.000     NCIT:C16629   \n",
       "295            leaves                 1        0.000  CO_366:0000273   \n",
       "524    Not Applicable                 1        0.000     NCIT:C48660   \n",
       "580   rice paddy soil                 1        0.198   ENVO:00005740   \n",
       "592        permafrost                 1        0.198   ENVO:00000134   \n",
       "778       rhizosphere                 1        0.184   ENVO:00005801   \n",
       "789               RNA                 1        0.000    OMIT:0013221   \n",
       "968  sediment surface                 1        0.155   ENVO:01001192   \n",
       "\n",
       "                           label  search_rank ontology_prefix  \\\n",
       "11                       Fifteen            5            NCIT   \n",
       "65                 Worry Score 4            3            NCIT   \n",
       "92   COVID-19 Disease Severity 6            3            NCIT   \n",
       "111              cryoconite hole            2            ENVO   \n",
       "129                          DNA            1            NCIT   \n",
       "166                       Genome            1            NCIT   \n",
       "295                       leaves            1          CO_366   \n",
       "524               Not Applicable            1            NCIT   \n",
       "580             paddy field soil            1            ENVO   \n",
       "592                   permafrost            1            ENVO   \n",
       "778                  rhizosphere            1            ENVO   \n",
       "789                          RNA            1            OMIT   \n",
       "968             sediment surface            2            ENVO   \n",
       "\n",
       "                   scope        type  \\\n",
       "11               synonym     synonym   \n",
       "65               synonym     synonym   \n",
       "92       hasExactSynonym        None   \n",
       "111                label       label   \n",
       "129                label       label   \n",
       "166                label       label   \n",
       "295                label       label   \n",
       "524                label       label   \n",
       "580  has_related_synonym  annotation   \n",
       "592                label       label   \n",
       "778                label       label   \n",
       "789                label       label   \n",
       "968                label       label   \n",
       "\n",
       "                                                iri ontology_name  \n",
       "11      http://purl.obolibrary.org/obo/NCIT_C113429          ncit  \n",
       "65      http://purl.obolibrary.org/obo/NCIT_C172971          ncit  \n",
       "92      http://purl.obolibrary.org/obo/NCIT_C179016          ncit  \n",
       "111    http://purl.obolibrary.org/obo/ENVO_03000039          envo  \n",
       "129        http://purl.obolibrary.org/obo/NCIT_C449          ncit  \n",
       "166      http://purl.obolibrary.org/obo/NCIT_C16629          ncit  \n",
       "295  http://www.cropontology.org/rdf/CO_366:0000273        co_366  \n",
       "524      http://purl.obolibrary.org/obo/NCIT_C48660          ncit  \n",
       "580    http://purl.obolibrary.org/obo/ENVO_00005740          envo  \n",
       "592    http://purl.obolibrary.org/obo/ENVO_00000134          envo  \n",
       "778    http://purl.obolibrary.org/obo/ENVO_00005801          envo  \n",
       "789     http://purl.obolibrary.org/obo/OMIT_0013221          omit  \n",
       "968    http://purl.obolibrary.org/obo/ENVO_01001192          envo  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    salvage_search_res, max_string_dist=salvage_max_dist\n",
    ")\n",
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the mappings for review!\n",
    "\n",
    "_Note that the merge on ID or label mappings are not included here_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_best_acceptable[include_col] = first_pass_include_val\n",
    "salvage_best_acceptable[include_col] = salvage_include_val\n",
    "best_and_salvage = pd.concat([my_best_acceptable, salvage_best_acceptable])\n",
    "\n",
    "best_and_salvage[\"ols_curation_notes\"] = \"\"\n",
    "\n",
    "# best_and_salvage.to_csv(ols_review_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "best_and_salvage = best_and_salvage[\n",
    "    [\n",
    "        \"category\",\n",
    "        \"raw\",\n",
    "        \"query\",\n",
    "        \"name\",\n",
    "        \"obo_id\",\n",
    "        \"label\",\n",
    "        \"ontology_prefix\",\n",
    "        \"include\",\n",
    "        \"ols_curation_notes\",\n",
    "        \"scope\",\n",
    "        \"type\",\n",
    "        \"iri\",\n",
    "        \"ontology_name\",\n",
    "        \"string_dist_rank\",\n",
    "        \"string_dist\",\n",
    "        \"search_rank\",\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>include</th>\n",
       "      <th>ols_curation_notes</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>search_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>salvage</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>CO_366:0000273</td>\n",
       "      <td>leaves</td>\n",
       "      <td>CO_366</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_366:0000273</td>\n",
       "      <td>co_366</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>salvage</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>ENVO:00000134</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000134</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005801</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>salvage</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.155</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT:C172971</td>\n",
       "      <td>Worry Score 4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172971</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>salvage</td>\n",
       "      <td>dna</td>\n",
       "      <td>dna</td>\n",
       "      <td>DNA</td>\n",
       "      <td>NCIT:C449</td>\n",
       "      <td>DNA</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C449</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome</td>\n",
       "      <td>genome</td>\n",
       "      <td>Genome</td>\n",
       "      <td>NCIT:C16629</td>\n",
       "      <td>Genome</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C16629</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>RNA</td>\n",
       "      <td>OMIT:0013221</td>\n",
       "      <td>RNA</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0013221</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       category                     raw                   query  \\\n",
       "295     salvage                  leaves                  leaves   \n",
       "111     salvage              cryoconite              cryoconite   \n",
       "580     salvage              paddy soil              paddy soil   \n",
       "592     salvage         permafrost soil         permafrost soil   \n",
       "778     salvage        rhizosphere soil        rhizosphere soil   \n",
       "968     salvage  surface sediment water  surface sediment water   \n",
       "11      salvage                      15                      15   \n",
       "65      salvage                       4                       4   \n",
       "92      salvage                       6                       6   \n",
       "129     salvage                     dna                     dna   \n",
       "166     salvage                  genome                  genome   \n",
       "524     salvage          not applicable          not applicable   \n",
       "789     salvage                     rna                     rna   \n",
       "305  env_medium           mine tailings           mine tailings   \n",
       "425  env_medium                    peat                    peat   \n",
       "553  env_medium              sandy soil              sandy soil   \n",
       "45   env_medium                bacteria                bacteria   \n",
       "\n",
       "                 name          obo_id                        label  \\\n",
       "295            leaves  CO_366:0000273                       leaves   \n",
       "111   cryoconite hole   ENVO:03000039              cryoconite hole   \n",
       "580   rice paddy soil   ENVO:00005740             paddy field soil   \n",
       "592        permafrost   ENVO:00000134                   permafrost   \n",
       "778       rhizosphere   ENVO:00005801                  rhizosphere   \n",
       "968  sediment surface   ENVO:01001192             sediment surface   \n",
       "11                 15    NCIT:C113429                      Fifteen   \n",
       "65                  4    NCIT:C172971                Worry Score 4   \n",
       "92                  6    NCIT:C179016  COVID-19 Disease Severity 6   \n",
       "129               DNA       NCIT:C449                          DNA   \n",
       "166            Genome     NCIT:C16629                       Genome   \n",
       "524    Not Applicable     NCIT:C48660               Not Applicable   \n",
       "789               RNA    OMIT:0013221                          RNA   \n",
       "305      mine tailing   ENVO:00000003                 mine tailing   \n",
       "425              peat   ENVO:00005774                    peat soil   \n",
       "553        sandy soil   ENVO:00002229                     arenosol   \n",
       "45           Bacteria     NCBITaxon:2                     Bacteria   \n",
       "\n",
       "    ontology_prefix  include ols_curation_notes                scope  \\\n",
       "295          CO_366    False                                   label   \n",
       "111            ENVO    False                                   label   \n",
       "580            ENVO    False                     has_related_synonym   \n",
       "592            ENVO    False                                   label   \n",
       "778            ENVO    False                                   label   \n",
       "968            ENVO    False                                   label   \n",
       "11             NCIT    False                                 synonym   \n",
       "65             NCIT    False                                 synonym   \n",
       "92             NCIT    False                         hasExactSynonym   \n",
       "129            NCIT    False                                   label   \n",
       "166            NCIT    False                                   label   \n",
       "524            NCIT    False                                   label   \n",
       "789            OMIT    False                                   label   \n",
       "305            ENVO     True                                   label   \n",
       "425            ENVO     True                       hasRelatedSynonym   \n",
       "553            ENVO     True                                 synonym   \n",
       "45        NCBITAXON     True                                   label   \n",
       "\n",
       "           type                                             iri ontology_name  \\\n",
       "295       label  http://www.cropontology.org/rdf/CO_366:0000273        co_366   \n",
       "111       label    http://purl.obolibrary.org/obo/ENVO_03000039          envo   \n",
       "580  annotation    http://purl.obolibrary.org/obo/ENVO_00005740          envo   \n",
       "592       label    http://purl.obolibrary.org/obo/ENVO_00000134          envo   \n",
       "778       label    http://purl.obolibrary.org/obo/ENVO_00005801          envo   \n",
       "968       label    http://purl.obolibrary.org/obo/ENVO_01001192          envo   \n",
       "11      synonym     http://purl.obolibrary.org/obo/NCIT_C113429          ncit   \n",
       "65      synonym     http://purl.obolibrary.org/obo/NCIT_C172971          ncit   \n",
       "92         None     http://purl.obolibrary.org/obo/NCIT_C179016          ncit   \n",
       "129       label        http://purl.obolibrary.org/obo/NCIT_C449          ncit   \n",
       "166       label      http://purl.obolibrary.org/obo/NCIT_C16629          ncit   \n",
       "524       label      http://purl.obolibrary.org/obo/NCIT_C48660          ncit   \n",
       "789       label     http://purl.obolibrary.org/obo/OMIT_0013221          omit   \n",
       "305       label    http://purl.obolibrary.org/obo/ENVO_00000003          envo   \n",
       "425        None    http://purl.obolibrary.org/obo/ENVO_00005774          envo   \n",
       "553     synonym    http://purl.obolibrary.org/obo/ENVO_00002229          envo   \n",
       "45        label      http://purl.obolibrary.org/obo/NCBITaxon_2     ncbitaxon   \n",
       "\n",
       "     string_dist_rank  string_dist  search_rank  \n",
       "295                 1        0.000            1  \n",
       "111                 1        0.198            2  \n",
       "580                 1        0.198            1  \n",
       "592                 1        0.198            1  \n",
       "778                 1        0.184            1  \n",
       "968                 1        0.155            2  \n",
       "11                  1        0.000            5  \n",
       "65                  1        0.000            3  \n",
       "92                  1        0.000            3  \n",
       "129                 1        0.000            1  \n",
       "166                 1        0.000            1  \n",
       "524                 1        0.000            1  \n",
       "789                 1        0.000            1  \n",
       "305                 1        0.036            4  \n",
       "425                 1        0.000            1  \n",
       "553                 1        0.000            1  \n",
       "45                  1        0.000            1  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_and_salvage = best_and_salvage.sort_values(\n",
    "    [\"include\", \"ontology_prefix\"], ascending=(True, True)\n",
    ")\n",
    "best_and_salvage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_and_salvage.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## print mapped column and scoping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_package_normalization.EnvPackage\n",
      "soil\n",
      "env_medium\n"
     ]
    }
   ],
   "source": [
    "print(scoping_col)\n",
    "print(scoping_value)\n",
    "print(biosample_col_to_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemExit",
     "evalue": "Don't skip reviewing and saving!",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m Don't skip reviewing and saving!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/biosample-analysis/venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3425: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "raise SystemExit(\"Don't skip reviewing and saving!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do some review in a spreadsheet application!\n",
    "Specifically, columns `include` and `ols_curation_notes`\n",
    "\n",
    "Then Save\n",
    "\n",
    "**Add counts back in to motivate manual lookups if necessary**\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now read the reviewed spreadsheet back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# curated = pd.read_csv(ols_review_file, sep=ols_review_seperator)\n",
    "# curated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OR copy and paste"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>include</th>\n",
       "      <th>ols_curation_notes</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>search_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>salvage</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>CO_366:0000273</td>\n",
       "      <td>leaves</td>\n",
       "      <td>CO_366</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_366:0000273</td>\n",
       "      <td>co_366</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>salvage</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>ENVO:00000134</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000134</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005801</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>salvage</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.155</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT:C172971</td>\n",
       "      <td>Worry Score 4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172971</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>salvage</td>\n",
       "      <td>dna</td>\n",
       "      <td>dna</td>\n",
       "      <td>DNA</td>\n",
       "      <td>NCIT:C449</td>\n",
       "      <td>DNA</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C449</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome</td>\n",
       "      <td>genome</td>\n",
       "      <td>Genome</td>\n",
       "      <td>NCIT:C16629</td>\n",
       "      <td>Genome</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C16629</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>RNA</td>\n",
       "      <td>OMIT:0013221</td>\n",
       "      <td>RNA</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0013221</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      category                     raw                   query  \\\n",
       "0      salvage              paddy soil              paddy soil   \n",
       "1      salvage                  leaves                  leaves   \n",
       "2      salvage              cryoconite              cryoconite   \n",
       "3      salvage         permafrost soil         permafrost soil   \n",
       "4      salvage        rhizosphere soil        rhizosphere soil   \n",
       "5      salvage  surface sediment water  surface sediment water   \n",
       "6      salvage                      15                      15   \n",
       "7      salvage                       4                       4   \n",
       "8      salvage                       6                       6   \n",
       "9      salvage                     dna                     dna   \n",
       "10     salvage                  genome                  genome   \n",
       "11     salvage          not applicable          not applicable   \n",
       "12     salvage                     rna                     rna   \n",
       "13  env_medium           mine tailings           mine tailings   \n",
       "14  env_medium                    peat                    peat   \n",
       "15  env_medium              sandy soil              sandy soil   \n",
       "16  env_medium                bacteria                bacteria   \n",
       "\n",
       "                name          obo_id                        label  \\\n",
       "0    rice paddy soil   ENVO:00005740             paddy field soil   \n",
       "1             leaves  CO_366:0000273                       leaves   \n",
       "2    cryoconite hole   ENVO:03000039              cryoconite hole   \n",
       "3         permafrost   ENVO:00000134                   permafrost   \n",
       "4        rhizosphere   ENVO:00005801                  rhizosphere   \n",
       "5   sediment surface   ENVO:01001192             sediment surface   \n",
       "6                 15    NCIT:C113429                      Fifteen   \n",
       "7                  4    NCIT:C172971                Worry Score 4   \n",
       "8                  6    NCIT:C179016  COVID-19 Disease Severity 6   \n",
       "9                DNA       NCIT:C449                          DNA   \n",
       "10            Genome     NCIT:C16629                       Genome   \n",
       "11    Not Applicable     NCIT:C48660               Not Applicable   \n",
       "12               RNA    OMIT:0013221                          RNA   \n",
       "13      mine tailing   ENVO:00000003                 mine tailing   \n",
       "14              peat   ENVO:00005774                    peat soil   \n",
       "15        sandy soil   ENVO:00002229                     arenosol   \n",
       "16          Bacteria     NCBITaxon:2                     Bacteria   \n",
       "\n",
       "   ontology_prefix  include  ols_curation_notes                scope  \\\n",
       "0             ENVO     True                 NaN  has_related_synonym   \n",
       "1           CO_366    False                 NaN                label   \n",
       "2             ENVO    False                 NaN                label   \n",
       "3             ENVO    False                 NaN                label   \n",
       "4             ENVO    False                 NaN                label   \n",
       "5             ENVO    False                 NaN                label   \n",
       "6             NCIT    False                 NaN              synonym   \n",
       "7             NCIT    False                 NaN              synonym   \n",
       "8             NCIT    False                 NaN      hasExactSynonym   \n",
       "9             NCIT    False                 NaN                label   \n",
       "10            NCIT    False                 NaN                label   \n",
       "11            NCIT    False                 NaN                label   \n",
       "12            OMIT    False                 NaN                label   \n",
       "13            ENVO     True                 NaN                label   \n",
       "14            ENVO     True                 NaN    hasRelatedSynonym   \n",
       "15            ENVO     True                 NaN              synonym   \n",
       "16       NCBITAXON     True                 NaN                label   \n",
       "\n",
       "          type                                             iri ontology_name  \\\n",
       "0   annotation    http://purl.obolibrary.org/obo/ENVO_00005740          envo   \n",
       "1        label  http://www.cropontology.org/rdf/CO_366:0000273        co_366   \n",
       "2        label    http://purl.obolibrary.org/obo/ENVO_03000039          envo   \n",
       "3        label    http://purl.obolibrary.org/obo/ENVO_00000134          envo   \n",
       "4        label    http://purl.obolibrary.org/obo/ENVO_00005801          envo   \n",
       "5        label    http://purl.obolibrary.org/obo/ENVO_01001192          envo   \n",
       "6      synonym     http://purl.obolibrary.org/obo/NCIT_C113429          ncit   \n",
       "7      synonym     http://purl.obolibrary.org/obo/NCIT_C172971          ncit   \n",
       "8          NaN     http://purl.obolibrary.org/obo/NCIT_C179016          ncit   \n",
       "9        label        http://purl.obolibrary.org/obo/NCIT_C449          ncit   \n",
       "10       label      http://purl.obolibrary.org/obo/NCIT_C16629          ncit   \n",
       "11       label      http://purl.obolibrary.org/obo/NCIT_C48660          ncit   \n",
       "12       label     http://purl.obolibrary.org/obo/OMIT_0013221          omit   \n",
       "13       label    http://purl.obolibrary.org/obo/ENVO_00000003          envo   \n",
       "14         NaN    http://purl.obolibrary.org/obo/ENVO_00005774          envo   \n",
       "15     synonym    http://purl.obolibrary.org/obo/ENVO_00002229          envo   \n",
       "16       label      http://purl.obolibrary.org/obo/NCBITaxon_2     ncbitaxon   \n",
       "\n",
       "    string_dist_rank  string_dist  search_rank  \n",
       "0                  1        0.198            1  \n",
       "1                  1        0.000            1  \n",
       "2                  1        0.198            2  \n",
       "3                  1        0.198            1  \n",
       "4                  1        0.184            1  \n",
       "5                  1        0.155            2  \n",
       "6                  1        0.000            5  \n",
       "7                  1        0.000            3  \n",
       "8                  1        0.000            3  \n",
       "9                  1        0.000            1  \n",
       "10                 1        0.000            1  \n",
       "11                 1        0.000            1  \n",
       "12                 1        0.000            1  \n",
       "13                 1        0.036            4  \n",
       "14                 1        0.000            1  \n",
       "15                 1        0.000            1  \n",
       "16                 1        0.000            1  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curated = pd.read_clipboard()\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter on `include`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = curated[\"include\"]\n",
    "curated = curated.loc[flag]\n",
    "\n",
    "curated.columns = \"ols_\" + curated.columns\n",
    "# curated.to_sql(\"curated\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "# curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged = mapping_candidates.merge(\n",
    "    curated, how=\"outer\", left_on=\"remaining_tidied\", right_on=\"ols_raw\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = (\n",
    "    ~merge_search_merged[\"id_or_lab_ok\"]\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].eq(\"\")\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].isna()\n",
    ")\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_obo_id\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_id\"] = replacement\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_label\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_lab\"] = replacement\n",
    "\n",
    "replacement = (\n",
    "    merge_search_merged.loc[flag, \"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + merge_search_merged.loc[flag, \"consensus_id\"]\n",
    "    + \"]\"\n",
    ")\n",
    "\n",
    "merge_search_merged.loc[flag, \"assembled_consensus\"] = replacement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge_search_merged.to_sql(\n",
    "#     \"merge_search_merged\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "summary = merge_search_merged[[biosample_col_to_map, \"count\", \"assembled_consensus\"]]\n",
    "summary = summary.sort_values(\"count\", ascending=False)\n",
    "# summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "frame_for_sssom = merge_search_merged\n",
    "\n",
    "na_flag = frame_for_sssom[\"splitted\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "na_flag = frame_for_sssom[\"consensus_id\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "raw_queries = list(frame_for_sssom[\"splitted\"])\n",
    "urlencodeds = list(map(urllib.parse.quote, raw_queries))\n",
    "urlencodeds = [sssom_subject_prefix + \":\" + item for item in urlencodeds]\n",
    "\n",
    "results_rows = len(frame_for_sssom.index)\n",
    "iso8601_stamp = datetime.now().replace(microsecond=0).isoformat()\n",
    "\n",
    "sssom_frame = {\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"subject_label\": frame_for_sssom[\"splitted\"],\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"object_id\": frame_for_sssom[\"consensus_id\"],\n",
    "    \"object_label\": frame_for_sssom[\"consensus_lab\"],\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    \"mapping_date\": [iso8601_stamp] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "    \"subject_id\": urlencodeds,\n",
    "}\n",
    "\n",
    "sssom_frame = pd.DataFrame(sssom_frame)\n",
    "\n",
    "sssom_frame.to_csv(sssom_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "# sssom_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From here down requires refactoring and generalization \n",
    "\n",
    "That may even be true for some of the blocks above this!\n",
    "\n",
    "## Join together pipe-delimited env_broad_scale mappings \n",
    "_Any other constraints? If part_count is > 1, do we need to check for empty or null env_broad_scale, repeated or splitted?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    110\n",
       "Name: part_count, dtype: int64"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_flag = merge_search_merged[\"part_count\"] > 1\n",
    "multi_part_flag.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_medium, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_frame = merge_search_merged.loc[multi_part_flag]\n",
    "\n",
    "multi_part_frame = multi_part_frame[\n",
    "    [\n",
    "        biosample_col_to_map,\n",
    "        \"count\",\n",
    "        \"repeated\",\n",
    "        \"splitted\",\n",
    "        \"part_count\",\n",
    "        \"seq\",\n",
    "        \"consensus_id\",\n",
    "        \"consensus_lab\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "multi_part_frame = multi_part_frame.sort_values(\n",
    "    [biosample_col_to_map, \"seq\"], ascending=(True, True)\n",
    ")\n",
    "\n",
    "multi_part_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: part_count, dtype: int64)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "part_count = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\n",
    "    \"part_count\"\n",
    "].max()\n",
    "part_count.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: seq, dtype: int64)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_max = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\"seq\"].max()\n",
    "seq_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_medium, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parts_check = pd.DataFrame(dict(part_count=part_count, seq_max=seq_max)).reset_index()\n",
    "\n",
    "all_parts_flag = parts_check[\"seq_max\"] == parts_check[\"part_count\"]\n",
    "ebs_with_all_parts = parts_check.loc[all_parts_flag, biosample_col_to_map]\n",
    "ebs_with_all_parts\n",
    "\n",
    "all_parts_flag = multi_part_frame[biosample_col_to_map].isin(list(ebs_with_all_parts))\n",
    "all_parts_frame = multi_part_frame.loc[all_parts_flag]\n",
    "all_parts_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_pipesep_ebs = list(set(list(all_parts_frame[biosample_col_to_map])))\n",
    "unique_pipesep_ebs.sort()\n",
    "repipe_dict_list = []\n",
    "for one_pipesep in unique_pipesep_ebs:\n",
    "    flag = all_parts_frame[biosample_col_to_map] == one_pipesep\n",
    "    temp_frame = all_parts_frame.loc[flag]\n",
    "    ids_list = list(temp_frame[\"consensus_id\"])\n",
    "    lab_list = list(temp_frame[\"consensus_lab\"])\n",
    "    s = \"|\"\n",
    "    ids_repipe = s.join(ids_list)\n",
    "    lab_repipe = s.join(lab_list)\n",
    "\n",
    "    repipe_dict = {\n",
    "        biosample_col_to_map: one_pipesep,\n",
    "        \"consensus_id\": ids_repipe,\n",
    "        \"consensus_lab\": lab_repipe,\n",
    "    }\n",
    "    repipe_dict_list.append(repipe_dict)\n",
    "repipe_frame = pd.DataFrame(repipe_dict_list)\n",
    "repipe_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_min_cols = sssom_frame[[\"subject_label\", \"object_id\", \"object_label\"]]\n",
    "\n",
    "sssom_min_cols.columns = [biosample_col_to_map, \"consensus_id\", \"consensus_lab\"]\n",
    "\n",
    "# sssom_min_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "repipe_rows = len(repipe_frame.index)\n",
    "if repipe_rows > 0:\n",
    "    direct_and_repipe = pd.concat([sssom_min_cols, repipe_frame])\n",
    "else:\n",
    "    direct_and_repipe = sssom_min_cols\n",
    "# direct_and_repipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "direct_and_repipe.to_sql(\n",
    "    \"direct_and_repipe\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## refactor this even more?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-05 19:30:31.127529\n",
      "2021-07-05 19:30:31.367504\n",
      "0:00:00.239975\n"
     ]
    }
   ],
   "source": [
    "q = (\n",
    "    \"\"\"\n",
    "select\n",
    "\tb.id, b.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\",\n",
    "\tdar.consensus_id ,\n",
    "\tdar.consensus_lab\n",
    "from\n",
    "\tdirect_and_repipe dar\n",
    "left join biosample b on\n",
    "\tb.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\" = dar.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "\tb.taxonomy_id = stic.biosample_taxid\n",
    "join env_package_normalization on\n",
    "\tb.env_package = env_package_normalization.env_package\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"'\"\n",
    ")\n",
    "\n",
    "[per_biosample_scoped_mapping_results, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>631934</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>631935</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>631936</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>631937</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>631938</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15266</th>\n",
       "      <td>1906787</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15267</th>\n",
       "      <td>1906788</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15268</th>\n",
       "      <td>1906789</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15269</th>\n",
       "      <td>1906790</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15270</th>\n",
       "      <td>1906791</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalization.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15271 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id             env_medium   consensus_id          consensus_lab  \\\n",
       "0       631934  oil contaminated soil  ENVO:00002875  oil contaminated soil   \n",
       "1       631935  oil contaminated soil  ENVO:00002875  oil contaminated soil   \n",
       "2       631936  oil contaminated soil  ENVO:00002875  oil contaminated soil   \n",
       "3       631937  oil contaminated soil  ENVO:00002875  oil contaminated soil   \n",
       "4       631938        greenhouse soil  ENVO:00005780        greenhouse soil   \n",
       "...        ...                    ...            ...                    ...   \n",
       "15266  1906787            arable soil  ENVO:00005742            arable soil   \n",
       "15267  1906788            arable soil  ENVO:00005742            arable soil   \n",
       "15268  1906789            arable soil  ENVO:00005742            arable soil   \n",
       "15269  1906790            arable soil  ENVO:00005742            arable soil   \n",
       "15270  1906791            arable soil  ENVO:00005742            arable soil   \n",
       "\n",
       "                                           notes  \n",
       "0      env_package_normalization.EnvPackage=soil  \n",
       "1      env_package_normalization.EnvPackage=soil  \n",
       "2      env_package_normalization.EnvPackage=soil  \n",
       "3      env_package_normalization.EnvPackage=soil  \n",
       "4      env_package_normalization.EnvPackage=soil  \n",
       "...                                          ...  \n",
       "15266  env_package_normalization.EnvPackage=soil  \n",
       "15267  env_package_normalization.EnvPackage=soil  \n",
       "15268  env_package_normalization.EnvPackage=soil  \n",
       "15269  env_package_normalization.EnvPackage=soil  \n",
       "15270  env_package_normalization.EnvPackage=soil  \n",
       "\n",
       "[15271 rows x 5 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_biosample_scoped_mapping_results[\"notes\"] = scoping_col + \"=\" + scoping_value\n",
    "\n",
    "table_name = \"repaired_\" + biosample_col_to_map\n",
    "\n",
    "per_biosample_scoped_mapping_results.to_sql(\n",
    "    table_name, biosample_cnx, if_exists=\"append\", index=False\n",
    ")\n",
    "\n",
    "per_biosample_scoped_mapping_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
