{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import timeit\n",
    "import pyarrow\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from itertools import compress  \n",
    "\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.externals.six import StringIO  \n",
    "from IPython.display import Image  \n",
    "import pydotplus\n",
    "from sklearn.tree import export_graphviz\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import OneHotEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we read the data from a compressed parquet file into a pandas df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seconds 294.7669460773468\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "df = pd.read_parquet(\"harmonized-table.parquet.gzip\")\n",
    "\n",
    "exec_time = (time.time() - start_time)\n",
    "print('seconds ' + str(exec_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 14300584 entries, 0 to 14300583\n",
      "Columns: 464 entries, accession to trophic_level\n",
      "dtypes: object(464)\n",
      "memory usage: 49.4+ GB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seconds 5109.506125926971\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "df_describe = df.describe(include='all')\n",
    "\n",
    "exec_time = (time.time() - start_time)\n",
    "print('seconds ' + str(exec_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>gap_consent_short_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>dose</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>...</th>\n",
       "      <th>porosity</th>\n",
       "      <th>birth_date</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>label</th>\n",
       "      <th>type_strain</th>\n",
       "      <th>family_relationship</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>14300583</td>\n",
       "      <td>0</td>\n",
       "      <td>4079113</td>\n",
       "      <td>185928</td>\n",
       "      <td>13196</td>\n",
       "      <td>16138</td>\n",
       "      <td>3162</td>\n",
       "      <td>16333</td>\n",
       "      <td>12891</td>\n",
       "      <td>20305</td>\n",
       "      <td>...</td>\n",
       "      <td>999</td>\n",
       "      <td>33672</td>\n",
       "      <td>4292</td>\n",
       "      <td>2689</td>\n",
       "      <td>29968</td>\n",
       "      <td>204128</td>\n",
       "      <td>49937</td>\n",
       "      <td>1584</td>\n",
       "      <td>26878</td>\n",
       "      <td>5102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>14300583</td>\n",
       "      <td>0</td>\n",
       "      <td>616</td>\n",
       "      <td>9008</td>\n",
       "      <td>384</td>\n",
       "      <td>5710</td>\n",
       "      <td>500</td>\n",
       "      <td>2811</td>\n",
       "      <td>571</td>\n",
       "      <td>686</td>\n",
       "      <td>...</td>\n",
       "      <td>109</td>\n",
       "      <td>4446</td>\n",
       "      <td>875</td>\n",
       "      <td>1278</td>\n",
       "      <td>1226</td>\n",
       "      <td>5283</td>\n",
       "      <td>37670</td>\n",
       "      <td>14</td>\n",
       "      <td>1046</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SAMN01127300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRU</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Ustekinumab 45 mg</td>\n",
       "      <td>not collected</td>\n",
       "      <td>20</td>\n",
       "      <td>Missing: Not provided</td>\n",
       "      <td>missing</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>15.1 ÃÂµg/L</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>0.25 g</td>\n",
       "      <td>9606</td>\n",
       "      <td>missing</td>\n",
       "      <td>Yes</td>\n",
       "      <td>proband</td>\n",
       "      <td>heterotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1300152</td>\n",
       "      <td>20296</td>\n",
       "      <td>2907</td>\n",
       "      <td>428</td>\n",
       "      <td>418</td>\n",
       "      <td>2257</td>\n",
       "      <td>978</td>\n",
       "      <td>1547</td>\n",
       "      <td>...</td>\n",
       "      <td>146</td>\n",
       "      <td>2621</td>\n",
       "      <td>140</td>\n",
       "      <td>169</td>\n",
       "      <td>3927</td>\n",
       "      <td>92901</td>\n",
       "      <td>293</td>\n",
       "      <td>1053</td>\n",
       "      <td>4343</td>\n",
       "      <td>1057</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession urogenit_disord gap_consent_short_name phenotype  \\\n",
       "count       14300583               0                4079113    185928   \n",
       "unique      14300583               0                    616      9008   \n",
       "top     SAMN01127300             NaN                    GRU   Unknown   \n",
       "freq               1             NaN                1300152     20296   \n",
       "\n",
       "                     dose      tot_nitro tot_depth_water_col  \\\n",
       "count               13196          16138                3162   \n",
       "unique                384           5710                 500   \n",
       "top     Ustekinumab 45 mg  not collected                  20   \n",
       "freq                 2907            428                 418   \n",
       "\n",
       "                      nitrate samp_store_dur    perturbation  ... porosity  \\\n",
       "count                   16333          12891           20305  ...      999   \n",
       "unique                   2811            571             686  ...      109   \n",
       "top     Missing: Not provided        missing  not applicable  ...      0.3   \n",
       "freq                     2257            978            1547  ...      146   \n",
       "\n",
       "            birth_date     tot_phosp   water_content samp_vol_we_dna_ext  \\\n",
       "count            33672          4292            2689               29968   \n",
       "unique            4446           875            1278                1226   \n",
       "top     not applicable  15.1 ÃÂµg/L  not applicable              0.25 g   \n",
       "freq              2621           140             169                3927   \n",
       "\n",
       "       host_taxid    label type_strain family_relationship trophic_level  \n",
       "count      204128    49937        1584               26878          5102  \n",
       "unique       5283    37670          14                1046            51  \n",
       "top          9606  missing         Yes             proband   heterotroph  \n",
       "freq        92901      293        1053                4343          1057  \n",
       "\n",
       "[4 rows x 464 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_describe.to_csv(\"harmonized-table.parquet_describe.tsv\", sep=\"\\t\", header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "300      13\n",
       "105      10\n",
       "1         8\n",
       "492       8\n",
       "689       7\n",
       "         ..\n",
       "2388      1\n",
       "56998     1\n",
       "1367      1\n",
       "344       1\n",
       "621       1\n",
       "Name: freq, Length: 347, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_describe.iloc[3,].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>douche</th>\n",
       "      <th>menarche</th>\n",
       "      <th>attribute</th>\n",
       "      <th>hrt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       urogenit_disord douche menarche attribute  hrt\n",
       "count                0      0        0         0    0\n",
       "unique               0      0        0         0    0\n",
       "top                NaN    NaN      NaN       NaN  NaN\n",
       "freq               NaN    NaN      NaN       NaN  NaN"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_describe.loc[:, df_describe.isna().any()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a first pass at values corresponding to 'no information', candidates for NA replacement for ML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_labels = ['NaN', 'Unknown', 'not collected', 'Missing: Not provided', 'missing', 'not applicable', 'not provided', '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count                      1545145\n",
       "unique                       17586\n",
       "top       human-associated habitat\n",
       "freq                        116964\n",
       "Name: env_broad_scale, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['env_broad_scale'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "human-associated habitat    116964\n",
       "missing                     108895\n",
       "urban biome                  86064\n",
       "not applicable               72041\n",
       "Bos taurus                   25803\n",
       "                             ...  \n",
       "AMC2-6                           1\n",
       "lake environment25               1\n",
       "ratGE4.4                         1\n",
       "alpine grassland 116             1\n",
       "Organism2                        1\n",
       "Name: env_broad_scale, Length: 17586, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['env_broad_scale'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feces                     146111\n",
       "faeces                    103521\n",
       "soil                       99029\n",
       "missing                    98177\n",
       "not applicable             71522\n",
       "                           ...  \n",
       "131955                         1\n",
       "Stool_02-047-4_6_month         1\n",
       "feces_GE4.4                    1\n",
       "Gut_St.Denijs_010              1\n",
       "feces_21SM.w2d                 1\n",
       "Name: env_medium, Length: 22520, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['env_medium'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting sklearn.preprocessing\n",
      "\u001b[31m  Could not find a version that satisfies the requirement sklearn.preprocessing (from versions: )\u001b[0m\n",
      "\u001b[31mNo matching distribution found for sklearn.preprocessing\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install sklearn.preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "bad input shape (14300584, 464)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-7947ff290990>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mlaen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpreprocessing\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mLabelEncoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_encode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlaen\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/preprocessing/label.py\u001b[0m in \u001b[0;36mfit_transform\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    233\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlike\u001b[0m \u001b[0mof\u001b[0m \u001b[0mshape\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mn_samples\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m         \"\"\"\n\u001b[0;32m--> 235\u001b[0;31m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcolumn_or_1d\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwarn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    236\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclasses_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_encode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    237\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcolumn_or_1d\u001b[0;34m(y, warn)\u001b[0m\n\u001b[1;32m    795\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    796\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 797\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"bad input shape {0}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    798\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    799\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: bad input shape (14300584, 464)"
     ]
    }
   ],
   "source": [
    "laen = preprocessing.LabelEncoder()\n",
    "df_encode = laen.fit_transform(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#onehotencoder = OneHotEncoder(categorical_features = [0])\n",
    "\n",
    "#df_encode = OneHotEncoder().fit_transform(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_encode.to_parquet('harmonized-table__onehot.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.indexes.base.Index"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#features = df.columns.drop('env_medium')\n",
    "#type(features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_human_index = df['taxonomy_id'].isin(['9606'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[93,\n",
       " 94,\n",
       " 95,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 115,\n",
       " 225,\n",
       " 226,\n",
       " 227,\n",
       " 228,\n",
       " 229,\n",
       " 230,\n",
       " 231,\n",
       " 232,\n",
       " 233,\n",
       " 234,\n",
       " 235,\n",
       " 236,\n",
       " 237,\n",
       " 238,\n",
       " 239,\n",
       " 240,\n",
       " 241,\n",
       " 242,\n",
       " 243,\n",
       " 244,\n",
       " 245,\n",
       " 246,\n",
       " 247,\n",
       " 248,\n",
       " 249,\n",
       " 250,\n",
       " 251,\n",
       " 252,\n",
       " 253,\n",
       " 254,\n",
       " 255,\n",
       " 256,\n",
       " 257,\n",
       " 258,\n",
       " 259,\n",
       " 260,\n",
       " 261,\n",
       " 262,\n",
       " 263,\n",
       " 264,\n",
       " 265,\n",
       " 266,\n",
       " 267,\n",
       " 268,\n",
       " 269,\n",
       " 335,\n",
       " 338,\n",
       " 339,\n",
       " 347,\n",
       " 348,\n",
       " 355,\n",
       " 356,\n",
       " 357,\n",
       " 358,\n",
       " 359,\n",
       " 361,\n",
       " 362,\n",
       " 363,\n",
       " 364,\n",
       " 365,\n",
       " 366,\n",
       " 367,\n",
       " 368,\n",
       " 369,\n",
       " 370,\n",
       " 371,\n",
       " 372,\n",
       " 373,\n",
       " 374,\n",
       " 375,\n",
       " 376,\n",
       " 377,\n",
       " 378,\n",
       " 379,\n",
       " 380,\n",
       " 381,\n",
       " 382,\n",
       " 383,\n",
       " 384,\n",
       " 385,\n",
       " 386,\n",
       " 387,\n",
       " 388,\n",
       " 389,\n",
       " 390,\n",
       " 391,\n",
       " 392,\n",
       " 393,\n",
       " 394,\n",
       " 395,\n",
       " 396,\n",
       " 397,\n",
       " 398,\n",
       " 399,\n",
       " 400,\n",
       " 401,\n",
       " 402,\n",
       " 403,\n",
       " 404,\n",
       " 405,\n",
       " 406,\n",
       " 407,\n",
       " 408,\n",
       " 409,\n",
       " 410,\n",
       " 411,\n",
       " 412,\n",
       " 413,\n",
       " 414,\n",
       " 415,\n",
       " 416,\n",
       " 417,\n",
       " 418,\n",
       " 419,\n",
       " 420,\n",
       " 421,\n",
       " 422,\n",
       " 423,\n",
       " 424,\n",
       " 425,\n",
       " 426,\n",
       " 427,\n",
       " 428,\n",
       " 429,\n",
       " 430,\n",
       " 431,\n",
       " 432,\n",
       " 433,\n",
       " 434,\n",
       " 435,\n",
       " 436,\n",
       " 437,\n",
       " 438,\n",
       " 439,\n",
       " 440,\n",
       " 441,\n",
       " 442,\n",
       " 443,\n",
       " 444,\n",
       " 445,\n",
       " 446,\n",
       " 447,\n",
       " 448,\n",
       " 449,\n",
       " 450,\n",
       " 451,\n",
       " 452,\n",
       " 453,\n",
       " 454,\n",
       " 455,\n",
       " 456,\n",
       " 457,\n",
       " 458,\n",
       " 459,\n",
       " 460,\n",
       " 461,\n",
       " 462,\n",
       " 463,\n",
       " 464,\n",
       " 465,\n",
       " 466,\n",
       " 467,\n",
       " 468,\n",
       " 469,\n",
       " 470,\n",
       " 471,\n",
       " 472,\n",
       " 473,\n",
       " 474,\n",
       " 475,\n",
       " 476,\n",
       " 477,\n",
       " 478,\n",
       " 479,\n",
       " 480,\n",
       " 481,\n",
       " 482,\n",
       " 483,\n",
       " 484,\n",
       " 485,\n",
       " 486,\n",
       " 487,\n",
       " 488,\n",
       " 489,\n",
       " 490,\n",
       " 491,\n",
       " 492,\n",
       " 493,\n",
       " 494,\n",
       " 495,\n",
       " 496,\n",
       " 497,\n",
       " 498,\n",
       " 499,\n",
       " 500,\n",
       " 501,\n",
       " 502,\n",
       " 503,\n",
       " 504,\n",
       " 505,\n",
       " 506,\n",
       " 507,\n",
       " 508,\n",
       " 509,\n",
       " 510,\n",
       " 511,\n",
       " 512,\n",
       " 513,\n",
       " 514,\n",
       " 515,\n",
       " 516,\n",
       " 517,\n",
       " 518,\n",
       " 519,\n",
       " 520,\n",
       " 521,\n",
       " 522,\n",
       " 585,\n",
       " 586,\n",
       " 587,\n",
       " 588,\n",
       " 589,\n",
       " 590,\n",
       " 698,\n",
       " 699,\n",
       " 711,\n",
       " 712,\n",
       " 713,\n",
       " 714,\n",
       " 715,\n",
       " 716,\n",
       " 717,\n",
       " 718,\n",
       " 719,\n",
       " 720,\n",
       " 721,\n",
       " 722,\n",
       " 723,\n",
       " 724,\n",
       " 725,\n",
       " 726,\n",
       " 727,\n",
       " 728,\n",
       " 729,\n",
       " 730,\n",
       " 731,\n",
       " 732,\n",
       " 733,\n",
       " 734,\n",
       " 735,\n",
       " 736,\n",
       " 737,\n",
       " 738,\n",
       " 739,\n",
       " 740,\n",
       " 741,\n",
       " 742,\n",
       " 743,\n",
       " 744,\n",
       " 745,\n",
       " 746,\n",
       " 747,\n",
       " 748,\n",
       " 749,\n",
       " 750,\n",
       " 751,\n",
       " 752,\n",
       " 753,\n",
       " 754,\n",
       " 755,\n",
       " 756,\n",
       " 757,\n",
       " 758,\n",
       " 759,\n",
       " 760,\n",
       " 761,\n",
       " 762,\n",
       " 763,\n",
       " 764,\n",
       " 765,\n",
       " 766,\n",
       " 767,\n",
       " 768,\n",
       " 769,\n",
       " 770,\n",
       " 771,\n",
       " 772,\n",
       " 773,\n",
       " 774,\n",
       " 775,\n",
       " 776,\n",
       " 777,\n",
       " 778,\n",
       " 779,\n",
       " 780,\n",
       " 781,\n",
       " 782,\n",
       " 783,\n",
       " 784,\n",
       " 785,\n",
       " 786,\n",
       " 787,\n",
       " 788,\n",
       " 789,\n",
       " 790,\n",
       " 791,\n",
       " 792,\n",
       " 793,\n",
       " 794,\n",
       " 795,\n",
       " 796,\n",
       " 797,\n",
       " 798,\n",
       " 799,\n",
       " 800,\n",
       " 801,\n",
       " 802,\n",
       " 803,\n",
       " 804,\n",
       " 805,\n",
       " 806,\n",
       " 807,\n",
       " 808,\n",
       " 809,\n",
       " 810,\n",
       " 811,\n",
       " 812,\n",
       " 813,\n",
       " 814,\n",
       " 815,\n",
       " 816,\n",
       " 817,\n",
       " 818,\n",
       " 819,\n",
       " 820,\n",
       " 821,\n",
       " 822,\n",
       " 823,\n",
       " 824,\n",
       " 825,\n",
       " 826,\n",
       " 827,\n",
       " 828,\n",
       " 829,\n",
       " 830,\n",
       " 831,\n",
       " 832,\n",
       " 833,\n",
       " 834,\n",
       " 835,\n",
       " 836,\n",
       " 837,\n",
       " 838,\n",
       " 839,\n",
       " 840,\n",
       " 841,\n",
       " 842,\n",
       " 843,\n",
       " 844,\n",
       " 845,\n",
       " 846,\n",
       " 847,\n",
       " 848,\n",
       " 849,\n",
       " 850,\n",
       " 851,\n",
       " 852,\n",
       " 853,\n",
       " 854,\n",
       " 855,\n",
       " 856,\n",
       " 857,\n",
       " 858,\n",
       " 859,\n",
       " 860,\n",
       " 861,\n",
       " 862,\n",
       " 863,\n",
       " 864,\n",
       " 865,\n",
       " 866,\n",
       " 867,\n",
       " 868,\n",
       " 869,\n",
       " 870,\n",
       " 871,\n",
       " 872,\n",
       " 873,\n",
       " 874,\n",
       " 875,\n",
       " 876,\n",
       " 877,\n",
       " 878,\n",
       " 879,\n",
       " 880,\n",
       " 881,\n",
       " 882,\n",
       " 883,\n",
       " 884,\n",
       " 885,\n",
       " 886,\n",
       " 887,\n",
       " 888,\n",
       " 889,\n",
       " 890,\n",
       " 891,\n",
       " 892,\n",
       " 893,\n",
       " 894,\n",
       " 895,\n",
       " 896,\n",
       " 897,\n",
       " 898,\n",
       " 899,\n",
       " 900,\n",
       " 901,\n",
       " 902,\n",
       " 903,\n",
       " 904,\n",
       " 905,\n",
       " 906,\n",
       " 907,\n",
       " 908,\n",
       " 909,\n",
       " 910,\n",
       " 911,\n",
       " 912,\n",
       " 913,\n",
       " 914,\n",
       " 915,\n",
       " 916,\n",
       " 917,\n",
       " 918,\n",
       " 919,\n",
       " 920,\n",
       " 921,\n",
       " 922,\n",
       " 923,\n",
       " 924,\n",
       " 925,\n",
       " 926,\n",
       " 927,\n",
       " 928,\n",
       " 929,\n",
       " 930,\n",
       " 931,\n",
       " 932,\n",
       " 933,\n",
       " 934,\n",
       " 935,\n",
       " 936,\n",
       " 937,\n",
       " 938,\n",
       " 939,\n",
       " 940,\n",
       " 941,\n",
       " 942,\n",
       " 943,\n",
       " 944,\n",
       " 945,\n",
       " 946,\n",
       " 947,\n",
       " 948,\n",
       " 949,\n",
       " 950,\n",
       " 951,\n",
       " 952,\n",
       " 953,\n",
       " 954,\n",
       " 955,\n",
       " 956,\n",
       " 957,\n",
       " 958,\n",
       " 959,\n",
       " 960,\n",
       " 961,\n",
       " 962,\n",
       " 963,\n",
       " 964,\n",
       " 965,\n",
       " 966,\n",
       " 967,\n",
       " 968,\n",
       " 969,\n",
       " 970,\n",
       " 971,\n",
       " 972,\n",
       " 973,\n",
       " 974,\n",
       " 975,\n",
       " 976,\n",
       " 977,\n",
       " 978,\n",
       " 979,\n",
       " 980,\n",
       " 981,\n",
       " 982,\n",
       " 983,\n",
       " 984,\n",
       " 985,\n",
       " 986,\n",
       " 987,\n",
       " 988,\n",
       " 989,\n",
       " 990,\n",
       " 991,\n",
       " 992,\n",
       " 993,\n",
       " 994,\n",
       " 995,\n",
       " 996,\n",
       " 997,\n",
       " 998,\n",
       " 999,\n",
       " 1000,\n",
       " 1001,\n",
       " 1002,\n",
       " 1003,\n",
       " 1004,\n",
       " 1005,\n",
       " 1006,\n",
       " 1007,\n",
       " 1008,\n",
       " 1009,\n",
       " 1010,\n",
       " 1011,\n",
       " 1012,\n",
       " 1013,\n",
       " 1014,\n",
       " 1015,\n",
       " 1016,\n",
       " 1017,\n",
       " 1018,\n",
       " 1019,\n",
       " 1020,\n",
       " 1021,\n",
       " 1022,\n",
       " 1023,\n",
       " 1024,\n",
       " 1025,\n",
       " 1026,\n",
       " 1027,\n",
       " 1028,\n",
       " 1029,\n",
       " 1030,\n",
       " 1031,\n",
       " 1032,\n",
       " 1033,\n",
       " 1034,\n",
       " 1035,\n",
       " 1036,\n",
       " 1037,\n",
       " 1038,\n",
       " 1039,\n",
       " 1040,\n",
       " 1041,\n",
       " 1042,\n",
       " 1043,\n",
       " 1044,\n",
       " 1045,\n",
       " 1046,\n",
       " 1047,\n",
       " 1048,\n",
       " 1049,\n",
       " 1050,\n",
       " 1051,\n",
       " 1052,\n",
       " 1053,\n",
       " 1054,\n",
       " 1055,\n",
       " 1056,\n",
       " 1057,\n",
       " 1058,\n",
       " 1059,\n",
       " 1060,\n",
       " 1061,\n",
       " 1062,\n",
       " 1063,\n",
       " 1064,\n",
       " 1065,\n",
       " 1066,\n",
       " 1067,\n",
       " 1068,\n",
       " 1069,\n",
       " 1070,\n",
       " 1071,\n",
       " 1072,\n",
       " 1073,\n",
       " 1074,\n",
       " 1075,\n",
       " 1076,\n",
       " 1077,\n",
       " 1078,\n",
       " 1079,\n",
       " 1080,\n",
       " 1081,\n",
       " 1082,\n",
       " 1083,\n",
       " 1084,\n",
       " 1085,\n",
       " 1086,\n",
       " 1087,\n",
       " 1088,\n",
       " 1089,\n",
       " 1090,\n",
       " 1091,\n",
       " 1092,\n",
       " 1093,\n",
       " 1094,\n",
       " 1095,\n",
       " 1096,\n",
       " 1097,\n",
       " 1098,\n",
       " 1099,\n",
       " 1100,\n",
       " 1101,\n",
       " 1102,\n",
       " 1103,\n",
       " 1104,\n",
       " 1105,\n",
       " 1106,\n",
       " 1107,\n",
       " 1108,\n",
       " 1109,\n",
       " 1110,\n",
       " 1111,\n",
       " 1112,\n",
       " 1113,\n",
       " 1114,\n",
       " 1115,\n",
       " 1116,\n",
       " 1117,\n",
       " 1118,\n",
       " 1119,\n",
       " 1120,\n",
       " 1121,\n",
       " 1122,\n",
       " 1123,\n",
       " 1124,\n",
       " 1125,\n",
       " 1126,\n",
       " 1127,\n",
       " 1128,\n",
       " 1129,\n",
       " 1130,\n",
       " 1131,\n",
       " 1132,\n",
       " 1133,\n",
       " 1134,\n",
       " 1135,\n",
       " 1136,\n",
       " 1137,\n",
       " 1138,\n",
       " 1139,\n",
       " 1140,\n",
       " 1141,\n",
       " 1142,\n",
       " 1143,\n",
       " 1144,\n",
       " 1145,\n",
       " 1146,\n",
       " 1147,\n",
       " 1148,\n",
       " 1149,\n",
       " 1150,\n",
       " 1151,\n",
       " 1152,\n",
       " 1153,\n",
       " 1154,\n",
       " 1155,\n",
       " 1156,\n",
       " 1157,\n",
       " 1158,\n",
       " 1159,\n",
       " 1160,\n",
       " 1161,\n",
       " 1162,\n",
       " 1163,\n",
       " 1164,\n",
       " 1165,\n",
       " 1166,\n",
       " 1167,\n",
       " 1168,\n",
       " 1169,\n",
       " 1170,\n",
       " 1171,\n",
       " 1172,\n",
       " 1173,\n",
       " 1174,\n",
       " 1175,\n",
       " 1176,\n",
       " 1177,\n",
       " 1178,\n",
       " 1179,\n",
       " 1180,\n",
       " 1181,\n",
       " 1182,\n",
       " 1183,\n",
       " 1184,\n",
       " 1185,\n",
       " 1186,\n",
       " 1187,\n",
       " 1188,\n",
       " 1189,\n",
       " 1190,\n",
       " 1191,\n",
       " 1192,\n",
       " 1193,\n",
       " 1194,\n",
       " 1195,\n",
       " 1196,\n",
       " 1197,\n",
       " 1198,\n",
       " 1199,\n",
       " 1200,\n",
       " 1201,\n",
       " 1202,\n",
       " 1203,\n",
       " 1204,\n",
       " 1205,\n",
       " 1206,\n",
       " 1207,\n",
       " 1208,\n",
       " 1209,\n",
       " 1210,\n",
       " 1211,\n",
       " 1212,\n",
       " 1213,\n",
       " 1214,\n",
       " 1215,\n",
       " 1216,\n",
       " 1217,\n",
       " 1218,\n",
       " 1219,\n",
       " 1220,\n",
       " 1221,\n",
       " 1222,\n",
       " 1223,\n",
       " 1224,\n",
       " 1225,\n",
       " 1226,\n",
       " 1227,\n",
       " 1228,\n",
       " 1229,\n",
       " 1230,\n",
       " 1231,\n",
       " 1232,\n",
       " 1233,\n",
       " 1259,\n",
       " 1260,\n",
       " 1261,\n",
       " 1262,\n",
       " 1263,\n",
       " 1264,\n",
       " 1265,\n",
       " 1266,\n",
       " 1267,\n",
       " 1268,\n",
       " 1269,\n",
       " 1270,\n",
       " 1292,\n",
       " 1293,\n",
       " 1294,\n",
       " 1295,\n",
       " 1300,\n",
       " 1401,\n",
       " 1402,\n",
       " 1403,\n",
       " 1404,\n",
       " 1405,\n",
       " 1406,\n",
       " 1407,\n",
       " 1408,\n",
       " 1409,\n",
       " 1410,\n",
       " 1411,\n",
       " 1412,\n",
       " 1413,\n",
       " 1414,\n",
       " 1415,\n",
       " 1416,\n",
       " 1417,\n",
       " 1418,\n",
       " 1419,\n",
       " 1420,\n",
       " 1421,\n",
       " 1422,\n",
       " 1423,\n",
       " 1424,\n",
       " 1425,\n",
       " 1426,\n",
       " 1427,\n",
       " 1428,\n",
       " 1429,\n",
       " 1430,\n",
       " 1431,\n",
       " 1432,\n",
       " 1433,\n",
       " 1434,\n",
       " 1435,\n",
       " 1436,\n",
       " 1437,\n",
       " 1438,\n",
       " 1439,\n",
       " 1440,\n",
       " 1441,\n",
       " 1442,\n",
       " 1443,\n",
       " 1444,\n",
       " 1445,\n",
       " 1446,\n",
       " 1447,\n",
       " 1448,\n",
       " 1449,\n",
       " 1450,\n",
       " 1451,\n",
       " 1452,\n",
       " 1453,\n",
       " 1454,\n",
       " 1455,\n",
       " 1456,\n",
       " 1457,\n",
       " 1458,\n",
       " 1459,\n",
       " 1460,\n",
       " 1461,\n",
       " 1462,\n",
       " 1463,\n",
       " 1464,\n",
       " 1465,\n",
       " 1466,\n",
       " 1467,\n",
       " 1468,\n",
       " 1469,\n",
       " 1470,\n",
       " 1471,\n",
       " 1472,\n",
       " 1473,\n",
       " 1474,\n",
       " 1475,\n",
       " 1476,\n",
       " 1477,\n",
       " 1478,\n",
       " 1479,\n",
       " 1480,\n",
       " 1481,\n",
       " 1482,\n",
       " 1483,\n",
       " 1484,\n",
       " 1485,\n",
       " 1486,\n",
       " 1487,\n",
       " 1488,\n",
       " 1489,\n",
       " 1490,\n",
       " 1491,\n",
       " 1492,\n",
       " 1493,\n",
       " 1494,\n",
       " 1495,\n",
       " 1496,\n",
       " 1497,\n",
       " 1498,\n",
       " 1499,\n",
       " 1500,\n",
       " 1501,\n",
       " 1502,\n",
       " 1503,\n",
       " 1504,\n",
       " 1505,\n",
       " 1506,\n",
       " 1507,\n",
       " 1508,\n",
       " 1509,\n",
       " 1510,\n",
       " 1511,\n",
       " 1512,\n",
       " 1513,\n",
       " 1514,\n",
       " 1515,\n",
       " 1516,\n",
       " 1517,\n",
       " 1518,\n",
       " 1519,\n",
       " 1520,\n",
       " 1521,\n",
       " 1522,\n",
       " 1523,\n",
       " 1524,\n",
       " 1732,\n",
       " 1733,\n",
       " 1734,\n",
       " 1735,\n",
       " 1736,\n",
       " 1737,\n",
       " 1738,\n",
       " 1739,\n",
       " 1740,\n",
       " 1741,\n",
       " 1742,\n",
       " 1743,\n",
       " 1744,\n",
       " 1745,\n",
       " 1746,\n",
       " 1747,\n",
       " 1748,\n",
       " 1749,\n",
       " 1750,\n",
       " 1751,\n",
       " 1752,\n",
       " 1755,\n",
       " 1756,\n",
       " 1757,\n",
       " 1758,\n",
       " 1759,\n",
       " 1760,\n",
       " 1761,\n",
       " 1762,\n",
       " 1763,\n",
       " 1764,\n",
       " 1765,\n",
       " 1766,\n",
       " 1767,\n",
       " 1768,\n",
       " 1769,\n",
       " 1770,\n",
       " 1771,\n",
       " 1772,\n",
       " 1773,\n",
       " 1774,\n",
       " 1775,\n",
       " 1776,\n",
       " 1777,\n",
       " 1778,\n",
       " 1779,\n",
       " 1780,\n",
       " 1781,\n",
       " 1782,\n",
       " 1783,\n",
       " 1784,\n",
       " 1785,\n",
       " 1786,\n",
       " 1787,\n",
       " 1788,\n",
       " 1789,\n",
       " 1790,\n",
       " 1791,\n",
       " 1792,\n",
       " 1793,\n",
       " 1794,\n",
       " 1795,\n",
       " 1796,\n",
       " 1797,\n",
       " 1798,\n",
       " 1799,\n",
       " 1800,\n",
       " 1801,\n",
       " 1810,\n",
       " 1811,\n",
       " 1812,\n",
       " 1813,\n",
       " 1814,\n",
       " 1815,\n",
       " 1816,\n",
       " 1817,\n",
       " 1818,\n",
       " 1819,\n",
       " 1820,\n",
       " 1821,\n",
       " 1822,\n",
       " 1823,\n",
       " 1824,\n",
       " 1826,\n",
       " 1827,\n",
       " 1873,\n",
       " 1874,\n",
       " 1875,\n",
       " 1876,\n",
       " 1877,\n",
       " 1894,\n",
       " 1895,\n",
       " 1940,\n",
       " 1941,\n",
       " 1951,\n",
       " 1952,\n",
       " 1953,\n",
       " 1954,\n",
       " 1955,\n",
       " 1958,\n",
       " ...]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_index = list(compress(range(len(df_human_index )), df_human_index ))\n",
    "df_human_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nonhuman = df.drop(df_human_index)\n",
    "                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7480877, 464)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_nonhuman.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nonhuman.to_parquet('harmonized-table__nonhuman.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['missing', None, 'MIENS_16S', 'water', 'host-associated',\n",
       "       'host associated', 'soil', 'Soil', 'MIMARKS_16S',\n",
       "       'miscellaneous natural or artificial environment', 'Human-gut',\n",
       "       'sediment', 'human-gut', 'MIMARKS_water', 'human-associated',\n",
       "       'MIGS/MIMS/MIMARKS.human-associated',\n",
       "       'MIGS/MIMS/MIMARKS.host-associated', 'MIGS/MIMS/MIMARKS.sediment',\n",
       "       'MIGS/MIMS/MIMARKS.wastewater', 'MIGS/MIMS/MIMARKS.miscellaneous',\n",
       "       'MIGS/MIMS/MIMARKS.soil', 'MIGS/MIMS/MIMARKS.human-oral',\n",
       "       'MIGS/MIMS/MIMARKS.plant-associated', 'MIGS/MIMS/MIMARKS.water',\n",
       "       'MIGS/MIMS/MIMARKS.human-gut', 'MIGS/MIMS/MIMARKS.microbial',\n",
       "       'seawater', 'MIGS/MIMS/MIMARKS.human-skin', 'wastewater/sludge',\n",
       "       'plant-associated', 'human-oral', 'MIGS/MIMS/MIMARKS.air',\n",
       "       'microbial mat/biofilm', 'MIGS/MIMS/MIMARKS.human-vaginal',\n",
       "       'not applicable', 'Host-associated', 'mouse-gut', 'not collected',\n",
       "       'Not available', 'air', 'wastewater|sludge', 'miscellaneous',\n",
       "       'built environment', 'unknown', 'arctic', 'MIGS.plant-associated',\n",
       "       'wastewater', 'misc environment', 'human-skin', 'human-vaginal',\n",
       "       'Human-associated', 'fresh water', 'freshwater sediment',\n",
       "       'sea water', 'sterile water', 'Coastal water body (ENVO:02000049)',\n",
       "       'sea water, [ENVO:00002149]', 'saliva', 'soil-associated',\n",
       "       'fermentation-associated', 'Sediment', 'biofilm', 'rock', 'plant',\n",
       "       'mice gut', 'Water', 'Human-oral', 'Unspecified',\n",
       "       'wastewater sludge', 'marine sediment (ENVO:00002113)', 'Unknown',\n",
       "       'Sea water', 'human-not providedsopharyngeal',\n",
       "       'microbial mat|biofilm', 'cave', 'misc-environment',\n",
       "       'ENVO:00000016', 'default', 'human gut',\n",
       "       'home, outdoor environment', 'Human_Gut', 'Mouse_Gut', 'mimarks',\n",
       "       'Human-vaginal', 'CV', 'human_skin', 'built', 'host_associated',\n",
       "       'env_package', 'gut microbiome', 'gut', 'Deep-sea water'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.env_package.unique()\n",
    "env_packages = ['soil', 'Soil', 'sediment', 'plant-associated', 'water', 'seawater', 'wastewater', 'fresh water', 'freshwater sediment', 'sea water', 'sterile water', 'Coastal water body (ENVO:02000049)', 'sea water, [ENVO:00002149]', 'soil-associated', 'Sediment', 'biofilm', 'rock', 'plant', 'Water', 'marine sediment (ENVO:00002113)', 'Unknown', 'Sea water', 'cave', 'Deep-sea water', 'MIGS/MIMS/MIMARKS.sediment', 'MIGS/MIMS/MIMARKS.soil', 'MIGS/MIMS/MIMARKS.plant-associated', 'MIGS/MIMS/MIMARKS.water', 'MIGS/MIMS/MIMARKS.plant-associated', 'MIGS/MIMS/MIMARKS.air', 'ENVO:00000016']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0           False\n",
       "1           False\n",
       "2           False\n",
       "3           False\n",
       "4           False\n",
       "            ...  \n",
       "14300579    False\n",
       "14300580    False\n",
       "14300581    False\n",
       "14300582    False\n",
       "14300583    False\n",
       "Name: env_package, Length: 14300584, dtype: bool"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_env_index = df['env_package'].isin(env_packages)\n",
    "df_env_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34313, 464)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv = df[df_env_index]\n",
    "df_onlyenv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyenv.to_parquet('harmonized-table__envonly.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seconds 1.3723068237304688\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "df_onlyenv_describe = df_onlyenv.describe(include='all')\n",
    "\n",
    "exec_time = (time.time() - start_time)\n",
    "print('seconds ' + str(exec_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>gap_consent_short_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>dose</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>...</th>\n",
       "      <th>porosity</th>\n",
       "      <th>birth_date</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>label</th>\n",
       "      <th>type_strain</th>\n",
       "      <th>family_relationship</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>34313</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2130</td>\n",
       "      <td>683</td>\n",
       "      <td>4955</td>\n",
       "      <td>204</td>\n",
       "      <td>196</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>454</td>\n",
       "      <td>14</td>\n",
       "      <td>7033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>34313</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>638</td>\n",
       "      <td>248</td>\n",
       "      <td>878</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>367</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SAMEA104176466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missing: Not Provided</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Missing: Not provided</td>\n",
       "      <td>02/01/2011-12/01/2011</td>\n",
       "      <td>Anti bird net applied, plants removed in bucke...</td>\n",
       "      <td>...</td>\n",
       "      <td>36.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.53microMolePerLiter</td>\n",
       "      <td>80</td>\n",
       "      <td>10 L</td>\n",
       "      <td>4577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chemolithotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>276</td>\n",
       "      <td>24</td>\n",
       "      <td>2001</td>\n",
       "      <td>60</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "      <td>14</td>\n",
       "      <td>4739</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession urogenit_disord gap_consent_short_name phenotype dose  \\\n",
       "count            34313               0                      0         0    0   \n",
       "unique           34313               0                      0         0    0   \n",
       "top     SAMEA104176466             NaN                    NaN       NaN  NaN   \n",
       "freq                 1             NaN                    NaN       NaN  NaN   \n",
       "\n",
       "                    tot_nitro tot_depth_water_col                nitrate  \\\n",
       "count                    2130                 683                   4955   \n",
       "unique                    638                 248                    878   \n",
       "top     Missing: Not Provided                 0.5  Missing: Not provided   \n",
       "freq                      276                  24                   2001   \n",
       "\n",
       "               samp_store_dur  \\\n",
       "count                     204   \n",
       "unique                      8   \n",
       "top     02/01/2011-12/01/2011   \n",
       "freq                       60   \n",
       "\n",
       "                                             perturbation  ... porosity  \\\n",
       "count                                                 196  ...       22   \n",
       "unique                                                 15  ...       16   \n",
       "top     Anti bird net applied, plants removed in bucke...  ...     36.3   \n",
       "freq                                                   70  ...        3   \n",
       "\n",
       "       birth_date              tot_phosp water_content samp_vol_we_dna_ext  \\\n",
       "count           0                     24           454                  14   \n",
       "unique          0                      4           367                   1   \n",
       "top           NaN  0.53microMolePerLiter            80                10 L   \n",
       "freq          NaN                      6            46                  14   \n",
       "\n",
       "       host_taxid label type_strain family_relationship    trophic_level  \n",
       "count        7033     0           0                   0                2  \n",
       "unique         21     0           0                   0                2  \n",
       "top          4577   NaN         NaN                 NaN  chemolithotroph  \n",
       "freq         4739   NaN         NaN                 NaN                1  \n",
       "\n",
       "[4 rows x 464 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>gap_consent_short_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>dose</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>...</th>\n",
       "      <th>porosity</th>\n",
       "      <th>birth_date</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>label</th>\n",
       "      <th>type_strain</th>\n",
       "      <th>family_relationship</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>SAMEA1706329</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>SAMEA1706311</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>SAMEA1706320</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>SAMEA1706300</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>SAMEA1706316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession urogenit_disord gap_consent_short_name phenotype  dose  \\\n",
       "19559  SAMEA1706329            None                   None      None  None   \n",
       "19560  SAMEA1706311            None                   None      None  None   \n",
       "19561  SAMEA1706320            None                   None      None  None   \n",
       "19562  SAMEA1706300            None                   None      None  None   \n",
       "19563  SAMEA1706316            None                   None      None  None   \n",
       "\n",
       "      tot_nitro tot_depth_water_col nitrate samp_store_dur perturbation  ...  \\\n",
       "19559      None                None    None           None         None  ...   \n",
       "19560      None                None    None           None         None  ...   \n",
       "19561      None                None    None           None         None  ...   \n",
       "19562      None                None    None           None         None  ...   \n",
       "19563      None                None    None           None         None  ...   \n",
       "\n",
       "      porosity birth_date tot_phosp water_content samp_vol_we_dna_ext  \\\n",
       "19559     None       None      None          None                None   \n",
       "19560     None       None      None          None                None   \n",
       "19561     None       None      None          None                None   \n",
       "19562     None       None      None          None                None   \n",
       "19563     None       None      None          None                None   \n",
       "\n",
       "      host_taxid label type_strain family_relationship trophic_level  \n",
       "19559       None  None        None                None          None  \n",
       "19560       None  None        None                None          None  \n",
       "19561       None  None        None                None          None  \n",
       "19562       None  None        None                None          None  \n",
       "19563       None  None        None                None          None  \n",
       "\n",
       "[5 rows x 464 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>gap_consent_short_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>dose</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>...</th>\n",
       "      <th>porosity</th>\n",
       "      <th>birth_date</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>label</th>\n",
       "      <th>type_strain</th>\n",
       "      <th>family_relationship</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>\"This sample (TARA_A100000165) was collected during the Tara Oceans expedition (2009-2013) at station TARA_018 (latitudeN=35.759, longitudeE=14.2574) on date/time=2009-11-02T08:13, using a PUMP (High Volume Peristaltic Pump).  The sample material (saline water (ENVO:00002010), including plankton (ENVO:xxxxxxxx)) was collected at a depth of 3-7 m, targeting a surface water layer (ENVO:00002042) in the marine biome (ENVO:00000447). The sample was size-fractionated (tbd-tbd micrometres), and stored in liquid nitrogen for later detection of prokaryote nucleic acid sequences by pyrosequencing methods, and for later metagenomics analysis. This sample has replicate sample(s): TARA_A100000166.\"</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\"This sample (TARA_A100000220) was collected during the Tara Oceans expedition (2009-2013) at station TARA_019 (latitudeN=34.2116, longitudeE=13.7684) on date/time=2009-11-11T16:24, using a PUMP (High Volume Peristaltic Pump).  The sample material (saline water (ENVO:00002010), including plankton (ENVO:xxxxxxxx)) was collected at a depth of 3-7 m, targeting a surface water layer (ENVO:00002042) in the marine biome (ENVO:00000447). The sample was size-fractionated (0.22-1.6 micrometres), preserved on board with RNA-Later, and stored in liquid nitrogen for later detection of prokaryote nucleic acid sequences by pyrosequencing methods, and for later metagenomics/transcriptomics analysis.\"</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\"This sample (TARA_A100001026) was collected during the Tara Oceans expedition (2009-2013) at station TARA_032 (latitudeN=23.36, longitudeE=37.2183) on date/time=2010-01-11T07:21, using a PUMP (High Volume Peristaltic Pump).  The sample material (saline water (ENVO:00002010), including plankton (ENVO:xxxxxxxx)) was collected at a depth of 3-7 m, targeting a surface water layer (ENVO:00002042) in the marine biome (ENVO:00000447). The sample was size-fractionated (0.22-1.6 micrometres), preserved on board with RNA-Later, and stored in liquid nitrogen for later detection of prokaryote nucleic acid sequences by pyrosequencing methods, and for later metagenomics/transcriptomics analysis. This sample has replicate sample(s): TARA_A100001027.\"</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\"This sample (TARA_A100001195) was collected during the Tara Oceans expedition (2009-2013) at station TARA_033 (latitudeN=21.9467, longitudeE=38.2517) on date/time=2010-01-13T07:16, using a PUMP (High Volume Peristaltic Pump).  The sample material (saline water (ENVO:00002010), including plankton (ENVO:xxxxxxxx)) was collected at a depth of 3-7 m, targeting a surface water layer (ENVO:00002042) in the marine biome (ENVO:00000447). The sample was size-fractionated (0.22-1.6 micrometres), preserved on board with RNA-Later, and stored in liquid nitrogen for later detection of prokaryote nucleic acid sequences by pyrosequencing methods, and for later metagenomics/transcriptomics analysis. This sample has replicate sample(s): TARA_A100001197.\"</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>\"This sample (TARA_A100001401) was collected during the Tara Oceans expedition (2009-2013) at station TARA_031 (latitudeN=27.16, longitudeE=27.1433) on date/time=2010-01-09T07:15, using a PUMP (High Volume Peristaltic Pump).  The sample material (saline water (ENVO:00002010), including plankton (ENVO:xxxxxxxx)) was collected at a depth of 3-7 m, targeting a surface water layer (ENVO:00002042) in the marine biome (ENVO:00000447). The sample was size-fractionated (0.22-1.6 micrometres), preserved on board with RNA-Later, and stored in liquid nitrogen for later detection of prokaryote nucleic acid sequences by pyrosequencing methods, and for later metagenomics/transcriptomics analysis. This sample has replicate sample(s): TARA_A100001454.\"</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>within</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wp15_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yes</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zooplankton</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zooplankton individuals per liter</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243009 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    accession  \\\n",
       "\"This sample (TARA_A100000165) was collected du...        NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...        NaN   \n",
       "...                                                       ...   \n",
       "within                                                    NaN   \n",
       "wp15_2                                                    NaN   \n",
       "yes                                                       NaN   \n",
       "zooplankton                                               NaN   \n",
       "zooplankton individuals per liter                         NaN   \n",
       "\n",
       "                                                    urogenit_disord  \\\n",
       "\"This sample (TARA_A100000165) was collected du...              NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...              NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...              NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...              NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...              NaN   \n",
       "...                                                             ...   \n",
       "within                                                          NaN   \n",
       "wp15_2                                                          NaN   \n",
       "yes                                                             NaN   \n",
       "zooplankton                                                     NaN   \n",
       "zooplankton individuals per liter                               NaN   \n",
       "\n",
       "                                                    gap_consent_short_name  \\\n",
       "\"This sample (TARA_A100000165) was collected du...                     NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...                     NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...                     NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...                     NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...                     NaN   \n",
       "...                                                                    ...   \n",
       "within                                                                 NaN   \n",
       "wp15_2                                                                 NaN   \n",
       "yes                                                                    NaN   \n",
       "zooplankton                                                            NaN   \n",
       "zooplankton individuals per liter                                      NaN   \n",
       "\n",
       "                                                    phenotype  dose  \\\n",
       "\"This sample (TARA_A100000165) was collected du...        NaN   NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...        NaN   NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...        NaN   NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...        NaN   NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...        NaN   NaN   \n",
       "...                                                       ...   ...   \n",
       "within                                                    NaN   NaN   \n",
       "wp15_2                                                    NaN   NaN   \n",
       "yes                                                       NaN   NaN   \n",
       "zooplankton                                               NaN   NaN   \n",
       "zooplankton individuals per liter                         NaN   NaN   \n",
       "\n",
       "                                                    tot_nitro  \\\n",
       "\"This sample (TARA_A100000165) was collected du...        NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...        NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...        NaN   \n",
       "...                                                       ...   \n",
       "within                                                    NaN   \n",
       "wp15_2                                                    NaN   \n",
       "yes                                                       NaN   \n",
       "zooplankton                                               NaN   \n",
       "zooplankton individuals per liter                         NaN   \n",
       "\n",
       "                                                    tot_depth_water_col  \\\n",
       "\"This sample (TARA_A100000165) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...                  NaN   \n",
       "...                                                                 ...   \n",
       "within                                                              NaN   \n",
       "wp15_2                                                              NaN   \n",
       "yes                                                                 NaN   \n",
       "zooplankton                                                         NaN   \n",
       "zooplankton individuals per liter                                   NaN   \n",
       "\n",
       "                                                    nitrate  samp_store_dur  \\\n",
       "\"This sample (TARA_A100000165) was collected du...      NaN             NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...      NaN             NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...      NaN             NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...      NaN             NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...      NaN             NaN   \n",
       "...                                                     ...             ...   \n",
       "within                                                  NaN             NaN   \n",
       "wp15_2                                                  NaN             NaN   \n",
       "yes                                                     NaN             NaN   \n",
       "zooplankton                                             NaN             NaN   \n",
       "zooplankton individuals per liter                       NaN             NaN   \n",
       "\n",
       "                                                    perturbation  ...  \\\n",
       "\"This sample (TARA_A100000165) was collected du...           NaN  ...   \n",
       "\"This sample (TARA_A100000220) was collected du...           NaN  ...   \n",
       "\"This sample (TARA_A100001026) was collected du...           NaN  ...   \n",
       "\"This sample (TARA_A100001195) was collected du...           NaN  ...   \n",
       "\"This sample (TARA_A100001401) was collected du...           NaN  ...   \n",
       "...                                                          ...  ...   \n",
       "within                                                       NaN  ...   \n",
       "wp15_2                                                       NaN  ...   \n",
       "yes                                                          NaN  ...   \n",
       "zooplankton                                                  NaN  ...   \n",
       "zooplankton individuals per liter                            NaN  ...   \n",
       "\n",
       "                                                    porosity  birth_date  \\\n",
       "\"This sample (TARA_A100000165) was collected du...       NaN         NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...       NaN         NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...       NaN         NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...       NaN         NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...       NaN         NaN   \n",
       "...                                                      ...         ...   \n",
       "within                                                   NaN         NaN   \n",
       "wp15_2                                                   NaN         NaN   \n",
       "yes                                                      NaN         NaN   \n",
       "zooplankton                                              NaN         NaN   \n",
       "zooplankton individuals per liter                        NaN         NaN   \n",
       "\n",
       "                                                    tot_phosp  water_content  \\\n",
       "\"This sample (TARA_A100000165) was collected du...        NaN            NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...        NaN            NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...        NaN            NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...        NaN            NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...        NaN            NaN   \n",
       "...                                                       ...            ...   \n",
       "within                                                    NaN            NaN   \n",
       "wp15_2                                                    NaN            NaN   \n",
       "yes                                                       NaN            NaN   \n",
       "zooplankton                                               NaN            NaN   \n",
       "zooplankton individuals per liter                         NaN            NaN   \n",
       "\n",
       "                                                    samp_vol_we_dna_ext  \\\n",
       "\"This sample (TARA_A100000165) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...                  NaN   \n",
       "...                                                                 ...   \n",
       "within                                                              NaN   \n",
       "wp15_2                                                              NaN   \n",
       "yes                                                                 NaN   \n",
       "zooplankton                                                         NaN   \n",
       "zooplankton individuals per liter                                   NaN   \n",
       "\n",
       "                                                    host_taxid  label  \\\n",
       "\"This sample (TARA_A100000165) was collected du...         NaN    NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...         NaN    NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...         NaN    NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...         NaN    NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...         NaN    NaN   \n",
       "...                                                        ...    ...   \n",
       "within                                                     NaN    NaN   \n",
       "wp15_2                                                     NaN    NaN   \n",
       "yes                                                        NaN    NaN   \n",
       "zooplankton                                                NaN    NaN   \n",
       "zooplankton individuals per liter                          NaN    NaN   \n",
       "\n",
       "                                                    type_strain  \\\n",
       "\"This sample (TARA_A100000165) was collected du...          NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...          NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...          NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...          NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...          NaN   \n",
       "...                                                         ...   \n",
       "within                                                      NaN   \n",
       "wp15_2                                                      NaN   \n",
       "yes                                                         NaN   \n",
       "zooplankton                                                 NaN   \n",
       "zooplankton individuals per liter                           NaN   \n",
       "\n",
       "                                                    family_relationship  \\\n",
       "\"This sample (TARA_A100000165) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100000220) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001026) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001195) was collected du...                  NaN   \n",
       "\"This sample (TARA_A100001401) was collected du...                  NaN   \n",
       "...                                                                 ...   \n",
       "within                                                              NaN   \n",
       "wp15_2                                                              NaN   \n",
       "yes                                                                 NaN   \n",
       "zooplankton                                                         NaN   \n",
       "zooplankton individuals per liter                                   NaN   \n",
       "\n",
       "                                                    trophic_level  \n",
       "\"This sample (TARA_A100000165) was collected du...            NaN  \n",
       "\"This sample (TARA_A100000220) was collected du...            NaN  \n",
       "\"This sample (TARA_A100001026) was collected du...            NaN  \n",
       "\"This sample (TARA_A100001195) was collected du...            NaN  \n",
       "\"This sample (TARA_A100001401) was collected du...            NaN  \n",
       "...                                                           ...  \n",
       "within                                                        NaN  \n",
       "wp15_2                                                        NaN  \n",
       "yes                                                           NaN  \n",
       "zooplankton                                                   NaN  \n",
       "zooplankton individuals per liter                             NaN  \n",
       "\n",
       "[243009 rows x 464 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_onlyenv_notNone = df_onlyenv[df_onlyenv != 'None'].index.values\n",
    "#df_onlyenv_notNone.shape\n",
    "\n",
    "#df_onlyenv.value_counts()\n",
    "\n",
    "#df_onlyenv.groupby(\"None\").count()\n",
    "\n",
    "df_onlyenv.apply(pd.value_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/frame.py:3065: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>urogenit_disord</th>\n",
       "      <th>gap_consent_short_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>dose</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>...</th>\n",
       "      <th>porosity</th>\n",
       "      <th>birth_date</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>label</th>\n",
       "      <th>type_strain</th>\n",
       "      <th>family_relationship</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>34313</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2130</td>\n",
       "      <td>683</td>\n",
       "      <td>4955</td>\n",
       "      <td>204</td>\n",
       "      <td>196</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>454</td>\n",
       "      <td>14</td>\n",
       "      <td>7033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>34313</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>638</td>\n",
       "      <td>248</td>\n",
       "      <td>878</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>367</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SAMN13905284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missing: Not Provided</td>\n",
       "      <td>18.64meter</td>\n",
       "      <td>Missing: Not provided</td>\n",
       "      <td>02/01/2011-12/01/2011</td>\n",
       "      <td>Anti bird net applied, plants removed in bucke...</td>\n",
       "      <td>...</td>\n",
       "      <td>36.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.6microMolePerLiter</td>\n",
       "      <td>80</td>\n",
       "      <td>10 L</td>\n",
       "      <td>4577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heterotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>276</td>\n",
       "      <td>24</td>\n",
       "      <td>2001</td>\n",
       "      <td>60</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "      <td>14</td>\n",
       "      <td>4739</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           accession urogenit_disord gap_consent_short_name phenotype dose  \\\n",
       "count          34313               0                      0         0    0   \n",
       "unique         34313               0                      0         0    0   \n",
       "top     SAMN13905284             NaN                    NaN       NaN  NaN   \n",
       "freq               1             NaN                    NaN       NaN  NaN   \n",
       "\n",
       "                    tot_nitro tot_depth_water_col                nitrate  \\\n",
       "count                    2130                 683                   4955   \n",
       "unique                    638                 248                    878   \n",
       "top     Missing: Not Provided          18.64meter  Missing: Not provided   \n",
       "freq                      276                  24                   2001   \n",
       "\n",
       "               samp_store_dur  \\\n",
       "count                     204   \n",
       "unique                      8   \n",
       "top     02/01/2011-12/01/2011   \n",
       "freq                       60   \n",
       "\n",
       "                                             perturbation  ... porosity  \\\n",
       "count                                                 196  ...       22   \n",
       "unique                                                 15  ...       16   \n",
       "top     Anti bird net applied, plants removed in bucke...  ...     36.3   \n",
       "freq                                                   70  ...        3   \n",
       "\n",
       "       birth_date             tot_phosp water_content samp_vol_we_dna_ext  \\\n",
       "count           0                    24           454                  14   \n",
       "unique          0                     4           367                   1   \n",
       "top           NaN  1.6microMolePerLiter            80                10 L   \n",
       "freq          NaN                     6            46                  14   \n",
       "\n",
       "       host_taxid label type_strain family_relationship trophic_level  \n",
       "count        7033     0           0                   0             2  \n",
       "unique         21     0           0                   0             2  \n",
       "top          4577   NaN         NaN                 NaN   heterotroph  \n",
       "freq         4739   NaN         NaN                 NaN             1  \n",
       "\n",
       "[4 rows x 464 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_cp = df_onlyenv\n",
    "cats = df_onlyenv_cp.columns\n",
    "df_onlyenv_cp[cats] = df_onlyenv_cp[cats].astype(\"category\")\n",
    "df_onlyenv_cp[cats].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyenv_value_counts = df_onlyenv.apply(pd.value_counts, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyenv_describe.to_csv(\"harmonized-table.parquet_onlyenv_describe.tsv\", sep=\"\\t\", header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SAMEA104176466' nan nan nan nan 'Missing: Not Provided' '0.5'\n",
      " 'Missing: Not provided' '02/01/2011-12/01/2011'\n",
      " 'Anti bird net applied, plants removed in buckets before sampling because of difficulty in access'\n",
      " 'lifeguard preservation solution' nan nan nan\n",
      " 'Palacios C, Zettler E, Amils R, Amaral-Zettler L (2008) Contrasting Microbial Community Assembly Hypotheses: A Reconciling Tale from the Rio Tinto. PLoS ONE 3(12): e3853. doi:10.1371/journal.pone.0003853'\n",
      " nan '67 kg N ha-1' nan nan nan nan '0.0' 'not applicable'\n",
      " '-1 degree Celsius' 'Missing: Not provide' nan nan nan 'Healthy' nan nan\n",
      " nan nan nan nan nan nan 'silt' 'Missing: Not provided' nan nan\n",
      " 'Missing: Not provided' '8.9' nan nan nan nan '457.904'\n",
      " 'Missing: Not provided' nan nan nan nan nan 'not applicable' nan nan nan\n",
      " nan 'Nipponbare' nan nan 'ENVO:cropland biome' nan nan\n",
      " '0.05 microgram per liter' nan nan nan 'multiple corer' nan nan nan nan\n",
      " '1,485,358 bp' '14_OA11.15' nan nan nan nan nan nan\n",
      " 'Missing: Not Provided' nan 'Missing: Not provided' nan nan nan nan nan\n",
      " 'Missing: Not provided' 'soil' '44.47microMolePerLiter' nan nan nan nan\n",
      " nan nan nan nan 'Missing: Not provided' nan 'bathroom' nan 'sandy loam'\n",
      " 'D' nan nan nan '0.01' nan nan nan nan 'Missing: Not Provided' 'DSMP' nan\n",
      " nan nan 'Missing: Not provided' nan nan '0.1' nan nan nan nan\n",
      " 'Missing: Not provided' nan 'gravimetric' nan 'not applicable'\n",
      " 'Meta-omics study of the global deep-sea benthic diversity with focus on the Clarion-Clipperton Fracture Zone'\n",
      " nan nan 'water' nan nan '7.210^41.710^4 microbes mlÃ¢Â\\x88Â\\x921'\n",
      " 'USA:AK' 'not applicable' nan nan nan nan 'Missing: Not Provided' nan\n",
      " 'Missing: Not Provided' nan nan nan nan nan '-80C' 'Generic' nan\n",
      " 'Missing: Not collected' nan nan nan 'Missing: Not provided'\n",
      " '10.21milligramPerMeterCubed' nan nan nan\n",
      " 'zooplankton individuals per liter' 'UBERON:feces' 'UBERON:feces'\n",
      " 'Missing: Not provided' nan 'Missing: Not provided' nan\n",
      " 'silt loam (2-4% slopes; fine-silty, mixed, superactive, mesic mollic oxyaquic hapludalf)'\n",
      " 'UBERON:feces' nan nan 'fresh water' '8' nan nan 'grass/herbaceous cover'\n",
      " 'live' 'Missing: Not provided' nan nan nan '0' nan '0.3355' nan nan nan\n",
      " nan nan nan nan nan nan '0.248' nan '27.48 kilogram per cubic meter'\n",
      " '2020-07-31T16:05:46' nan 'Missing: Not provided'\n",
      " 'Missing: Not collected' 'Generic.1.0' 'lake water'\n",
      " 'Missing: Not Provided' nan nan nan nan nan nan 'PRJNA316748'\n",
      " 'Missing:Not reported' '449393' nan nan nan nan nan nan\n",
      " 'Missing: Not provided' nan nan nan 'Missing: Not provided' nan nan nan\n",
      " 'Missing: Not provided' nan nan nan nan 'Missing: Not provided'\n",
      " '10.1128/AEM.06330-11' nan nan nan nan nan '25818237'\n",
      " 'bacteria and archaea' nan 'Pipette method' nan nan nan nan nan nan\n",
      " '2004' nan nan nan nan nan nan nan nan nan nan 'B7' nan nan nan nan\n",
      " 'Missing: Not collected' nan nan nan nan nan nan nan '9' nan nan nan nan\n",
      " nan '187' nan 'aerobe' 'Missing: Not provided' '5338'\n",
      " '1.72microMolePerLiter' nan 'Gray' nan nan nan 'not applicable'\n",
      " 'Dry_combustion' 'Dry_combustion' 'SRA:ERS1300917|BioSample:SAMEA4389468'\n",
      " nan nan 'yes' 'free living' nan nan nan nan nan\n",
      " 'For accompanying metadata and related data products, visit NEON data portal at data.neonscience.org and search for data product NEON.DP1.20221. Field data are located in NEON.DP1.20221, Zooplankton DNA barcode. For other documentation including protocols and data user guides, go to data.neonscience.org/documents'\n",
      " nan\n",
      " 'antibiotic resistance: streptomycine, ampiciline, cycloserine, cycloheximide'\n",
      " 'montaine steppe' nan '0.01 microgram per liter' nan\n",
      " 'Missing: Not provided' 'maize' nan '-80 degrees Celcius' nan\n",
      " 'silty clay loam' 'Cambisol' 'environmental' 'missing: not collected'\n",
      " 'Cunninghamia lanceolata plantation' nan nan nan 'freshwater metagenome'\n",
      " 'Missing: Not collected' nan 'Chloroform fumigation-incubation' nan nan\n",
      " nan nan '12.0' nan nan nan 'plant-associated habitat' nan nan\n",
      " 'Missing: Not Provided' nan nan nan 'bioproject' nan nan 'A' nan nan nan\n",
      " nan nan 'Missing: Not collected' nan nan nan nan 'US Taxonomy' nan nan\n",
      " 'MIMARKS-survey' nan 'Missing: Not reported' 'moderately well' nan nan\n",
      " nan 'http://soils.usda.gov/technical/classification/taxonomy' nan\n",
      " '12.5825535 N 116.6509938 W' nan 'not applicable' '2.9' nan nan nan nan\n",
      " '12.5 knots' nan nan 'Not applicable' nan nan nan nan 'maize' nan\n",
      " 'not applicable' 'not applicable' nan nan 'Generic'\n",
      " 'Missing: Not Provided' nan 'soil' nan '8.2 db'\n",
      " 'gene|marine metagenome|uncultured microorganism|uncultured organism|ecotype|biological replicate|technical replicate|interspecies interaction between organisms'\n",
      " nan 'not applicable' nan nan '.1,g' nan 'aerobic' nan nan nan nan nan\n",
      " '316748' nan nan nan '3.51 % silt and clay; 96.49 % sand' nan nan nan nan\n",
      " nan 'Zhiqin Pei and Christina Weissbecker'\n",
      " 'Sample from marine sediment metagenome' 'North west' nan 'Oct-2011' nan\n",
      " '15443275' '975' nan nan nan 'Missing: Not provided' nan\n",
      " 'one forest plot had the size 25.8 m x 25.8 m with 400 tree individuals planted in 1.29 m distance, at each forest plot four tree individuals of each tree species were randomly selected for sampling, at each sampling tree four soil cores were retrieved at 20-30 cm distance in each direction, all four cores form one composite sample for each sampling tree'\n",
      " 'Missing: Not provided' nan 'No sea ice' nan 'not applicable' nan\n",
      " 'DRI, Murray lab' nan '1.0 g/kg soil dw' '95' nan nan nan nan nan nan\n",
      " '36.3' nan '0.53microMolePerLiter' '80' '10 L' '4577' nan nan nan\n",
      " 'chemolithotroph']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "numpy.ndarray"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_onlyenv_describe.loc['top', :].to_numpy())\n",
    "type(df_onlyenv_describe.loc['top', :].to_numpy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0,   5,   6,   7,   8,   9,  10,  14,  16,  21,  22,  23,  24,\n",
       "        28,  37,  38,  41,  42,  47,  48,  54,  59,  62,  65,  69,  74,\n",
       "        75,  82,  84,  90,  91,  92, 101, 103, 105, 106, 110, 115, 116,\n",
       "       120, 123, 128, 130, 132, 133, 136, 139, 140, 141, 146, 148, 154,\n",
       "       155, 157, 161, 162, 166, 167, 168, 169, 171, 173, 174, 177, 178,\n",
       "       181, 182, 183, 187, 189, 199, 201, 202, 204, 205, 206, 207, 208,\n",
       "       215, 216, 217, 224, 228, 232, 237, 238, 244, 245, 247, 254, 265,\n",
       "       270, 278, 284, 286, 287, 288, 289, 291, 295, 296, 297, 298, 301,\n",
       "       302, 308, 310, 311, 313, 315, 316, 318, 320, 321, 322, 323, 324,\n",
       "       328, 329, 331, 336, 340, 343, 347, 350, 356, 361, 364, 366, 367,\n",
       "       371, 373, 375, 376, 381, 384, 389, 391, 392, 395, 396, 398, 400,\n",
       "       401, 403, 406, 408, 414, 418, 424, 425, 426, 428, 430, 431, 435,\n",
       "       437, 438, 440, 442, 444, 446, 447, 454, 456, 457, 458, 459, 463])"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#index = df_onlyenv_describe['top'].index[df_onlyenv_describe['top'].apply(np.isnan)]\n",
    "df_onlyenv_index_notnan = np.where(df_onlyenv_describe.loc['top', :].notnull())[0]#np.where(df_onlyenv_describe.loc['top', :].isnan())\n",
    "#index = list(np.where(df_onlyenv_describe.loc['top', :].isnan()[0]))\n",
    "\n",
    "df_onlyenv_index_notnan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyenv_nona = df_onlyenv[df_onlyenv.columns[df_onlyenv_index_notnan]]\n",
    "\n",
    "#gets opposite, so nans\n",
    "#df_onlyenv_nona = df_onlyenv.drop(df_onlyenv.columns[df_onlyenv_index_notnan], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>heavy_metals_meth</th>\n",
       "      <th>agrochem_addition</th>\n",
       "      <th>nitro</th>\n",
       "      <th>...</th>\n",
       "      <th>host_tissue_sampled</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>petroleum_hydrocarb</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>porosity</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>34313</td>\n",
       "      <td>2130</td>\n",
       "      <td>683</td>\n",
       "      <td>4955</td>\n",
       "      <td>204</td>\n",
       "      <td>196</td>\n",
       "      <td>4767</td>\n",
       "      <td>27</td>\n",
       "      <td>411</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>325</td>\n",
       "      <td>334</td>\n",
       "      <td>5</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>454</td>\n",
       "      <td>14</td>\n",
       "      <td>7033</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>34313</td>\n",
       "      <td>638</td>\n",
       "      <td>248</td>\n",
       "      <td>878</td>\n",
       "      <td>8</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>367</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SAMEA104176466</td>\n",
       "      <td>Missing: Not Provided</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Missing: Not provided</td>\n",
       "      <td>02/01/2011-12/01/2011</td>\n",
       "      <td>Anti bird net applied, plants removed in bucke...</td>\n",
       "      <td>lifeguard preservation solution</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>67 kg N ha-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>DRI, Murray lab</td>\n",
       "      <td>1.0 g/kg soil dw</td>\n",
       "      <td>95</td>\n",
       "      <td>36.3</td>\n",
       "      <td>0.53microMolePerLiter</td>\n",
       "      <td>80</td>\n",
       "      <td>10 L</td>\n",
       "      <td>4577</td>\n",
       "      <td>chemolithotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>276</td>\n",
       "      <td>24</td>\n",
       "      <td>2001</td>\n",
       "      <td>60</td>\n",
       "      <td>70</td>\n",
       "      <td>3139</td>\n",
       "      <td>27</td>\n",
       "      <td>355</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>310</td>\n",
       "      <td>328</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "      <td>14</td>\n",
       "      <td>4739</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 169 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession              tot_nitro tot_depth_water_col  \\\n",
       "count            34313                   2130                 683   \n",
       "unique           34313                    638                 248   \n",
       "top     SAMEA104176466  Missing: Not Provided                 0.5   \n",
       "freq                 1                    276                  24   \n",
       "\n",
       "                      nitrate         samp_store_dur  \\\n",
       "count                    4955                    204   \n",
       "unique                    878                      8   \n",
       "top     Missing: Not provided  02/01/2011-12/01/2011   \n",
       "freq                     2001                     60   \n",
       "\n",
       "                                             perturbation  \\\n",
       "count                                                 196   \n",
       "unique                                                 15   \n",
       "top     Anti bird net applied, plants removed in bucke...   \n",
       "freq                                                   70   \n",
       "\n",
       "                       samp_mat_process  \\\n",
       "count                              4767   \n",
       "unique                               13   \n",
       "top     lifeguard preservation solution   \n",
       "freq                               3139   \n",
       "\n",
       "                                        heavy_metals_meth agrochem_addition  \\\n",
       "count                                                  27               411   \n",
       "unique                                                  1                 3   \n",
       "top     Palacios C, Zettler E, Amils R, Amaral-Zettler...      67 kg N ha-1   \n",
       "freq                                                   27               355   \n",
       "\n",
       "       nitro  ... host_tissue_sampled   samp_store_loc petroleum_hydrocarb  \\\n",
       "count     55  ...                 325              334                   5   \n",
       "unique    43  ...                   4                2                   4   \n",
       "top      0.0  ...      not applicable  DRI, Murray lab    1.0 g/kg soil dw   \n",
       "freq       9  ...                 310              328                   2   \n",
       "\n",
       "       extreme_salinity porosity              tot_phosp water_content  \\\n",
       "count                23       22                     24           454   \n",
       "unique               11       16                      4           367   \n",
       "top                  95     36.3  0.53microMolePerLiter            80   \n",
       "freq                  6        3                      6            46   \n",
       "\n",
       "       samp_vol_we_dna_ext host_taxid    trophic_level  \n",
       "count                   14       7033                2  \n",
       "unique                   1         21                2  \n",
       "top                   10 L       4577  chemolithotroph  \n",
       "freq                    14       4739                1  \n",
       "\n",
       "[4 rows x 169 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_describe = df_onlyenv_nona.describe(include='all')\n",
    "df_onlyenv_nona_describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>tot_nitro</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>samp_store_dur</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>heavy_metals_meth</th>\n",
       "      <th>agrochem_addition</th>\n",
       "      <th>nitro</th>\n",
       "      <th>...</th>\n",
       "      <th>host_tissue_sampled</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>petroleum_hydrocarb</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>porosity</th>\n",
       "      <th>tot_phosp</th>\n",
       "      <th>water_content</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>trophic_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>SAMEA1706329</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sterivex filtration</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>SAMEA1706311</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sterivex filtration</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>SAMEA1706320</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sterivex filtration</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>SAMEA1706300</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sterivex filtration</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>SAMEA1706316</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sterivex filtration</td>\n",
       "      <td>Palacios C, Zettler E, Amils R, Amaral-Zettler...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 169 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          accession tot_nitro tot_depth_water_col nitrate samp_store_dur  \\\n",
       "19559  SAMEA1706329      None                None    None           None   \n",
       "19560  SAMEA1706311      None                None    None           None   \n",
       "19561  SAMEA1706320      None                None    None           None   \n",
       "19562  SAMEA1706300      None                None    None           None   \n",
       "19563  SAMEA1706316      None                None    None           None   \n",
       "\n",
       "      perturbation     samp_mat_process  \\\n",
       "19559         None  Sterivex filtration   \n",
       "19560         None  Sterivex filtration   \n",
       "19561         None  Sterivex filtration   \n",
       "19562         None  Sterivex filtration   \n",
       "19563         None  Sterivex filtration   \n",
       "\n",
       "                                       heavy_metals_meth agrochem_addition  \\\n",
       "19559  Palacios C, Zettler E, Amils R, Amaral-Zettler...              None   \n",
       "19560  Palacios C, Zettler E, Amils R, Amaral-Zettler...              None   \n",
       "19561  Palacios C, Zettler E, Amils R, Amaral-Zettler...              None   \n",
       "19562  Palacios C, Zettler E, Amils R, Amaral-Zettler...              None   \n",
       "19563  Palacios C, Zettler E, Amils R, Amaral-Zettler...              None   \n",
       "\n",
       "      nitro  ... host_tissue_sampled samp_store_loc petroleum_hydrocarb  \\\n",
       "19559  None  ...                None           None                None   \n",
       "19560  None  ...                None           None                None   \n",
       "19561  None  ...                None           None                None   \n",
       "19562  None  ...                None           None                None   \n",
       "19563  None  ...                None           None                None   \n",
       "\n",
       "      extreme_salinity porosity tot_phosp water_content samp_vol_we_dna_ext  \\\n",
       "19559             None     None      None          None                None   \n",
       "19560             None     None      None          None                None   \n",
       "19561             None     None      None          None                None   \n",
       "19562             None     None      None          None                None   \n",
       "19563             None     None      None          None                None   \n",
       "\n",
       "      host_taxid trophic_level  \n",
       "19559       None          None  \n",
       "19560       None          None  \n",
       "19561       None          None  \n",
       "19562       None          None  \n",
       "19563       None          None  \n",
       "\n",
       "[5 rows x 169 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "19559       water\n",
       "19560       water\n",
       "19561       water\n",
       "19562       water\n",
       "19563       water\n",
       "            ...  \n",
       "14275236     soil\n",
       "14275237     soil\n",
       "14275238     soil\n",
       "14275239     soil\n",
       "14275240     soil\n",
       "Name: env_package, Length: 34313, dtype: object"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_series = df_onlyenv_nona['env_package']\n",
    "print(type(env_series))\n",
    "env_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "accession              False\n",
       "tot_nitro              False\n",
       "tot_depth_water_col    False\n",
       "nitrate                False\n",
       "samp_store_dur         False\n",
       "                       ...  \n",
       "tot_phosp              False\n",
       "water_content           True\n",
       "samp_vol_we_dna_ext    False\n",
       "host_taxid              True\n",
       "trophic_level          False\n",
       "Name: top, Length: 169, dtype: bool"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_index = df_onlyenv_nona_describe.loc['top',:].str.isnumeric()\n",
    "df_onlyenv_nona_num_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyenv_nona_num = df_onlyenv_nona[df_onlyenv_nona.columns[df_onlyenv_nona_num_index]]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>49</td>\n",
       "      <td>5781</td>\n",
       "      <td>34313</td>\n",
       "      <td>13</td>\n",
       "      <td>183</td>\n",
       "      <td>12</td>\n",
       "      <td>172</td>\n",
       "      <td>52</td>\n",
       "      <td>538</td>\n",
       "      <td>34313</td>\n",
       "      <td>1257</td>\n",
       "      <td>23</td>\n",
       "      <td>454</td>\n",
       "      <td>7033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>79</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>34313</td>\n",
       "      <td>121</td>\n",
       "      <td>11</td>\n",
       "      <td>367</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>449393</td>\n",
       "      <td>25818237</td>\n",
       "      <td>2004</td>\n",
       "      <td>9</td>\n",
       "      <td>187</td>\n",
       "      <td>5338</td>\n",
       "      <td>316748</td>\n",
       "      <td>15443275</td>\n",
       "      <td>975</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "      <td>4577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>49</td>\n",
       "      <td>5452</td>\n",
       "      <td>6029</td>\n",
       "      <td>12</td>\n",
       "      <td>150</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>328</td>\n",
       "      <td>1</td>\n",
       "      <td>708</td>\n",
       "      <td>6</td>\n",
       "      <td>46</td>\n",
       "      <td>4739</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       samp_salinity altitude taxonomy_id ref_biomaterial  fire  \\\n",
       "count             49     5781       34313              13   183   \n",
       "unique             1        2          79               2     7   \n",
       "top                8        0      449393        25818237  2004   \n",
       "freq              49     5452        6029              12   150   \n",
       "\n",
       "       pool_dna_extracts microbial_biomass light_intensity entrez_value  \\\n",
       "count                 12               172              52          538   \n",
       "unique                 1               150              14           12   \n",
       "top                    9               187            5338       316748   \n",
       "freq                  12                 3               6          328   \n",
       "\n",
       "       biosample_id annual_season_precpt extreme_salinity water_content  \\\n",
       "count         34313                 1257               23           454   \n",
       "unique        34313                  121               11           367   \n",
       "top        15443275                  975               95            80   \n",
       "freq              1                  708                6            46   \n",
       "\n",
       "       host_taxid  \n",
       "count        7033  \n",
       "unique         21  \n",
       "top          4577  \n",
       "freq         4739  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_describe = df_onlyenv_nona_num.describe(include='all')\n",
    "df_onlyenv_nona_num_describe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>0</td>\n",
       "      <td>329</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0 m</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       samp_salinity altitude taxonomy_id ref_biomaterial fire  \\\n",
       "count              0      329           0               0    0   \n",
       "unique             0        1           0               0    0   \n",
       "top              NaN      0 m         NaN             NaN  NaN   \n",
       "freq             NaN      329         NaN             NaN  NaN   \n",
       "\n",
       "       pool_dna_extracts microbial_biomass light_intensity entrez_value  \\\n",
       "count                  0                 0               0            0   \n",
       "unique                 0                 0               0            0   \n",
       "top                  NaN               NaN             NaN          NaN   \n",
       "freq                 NaN               NaN             NaN          NaN   \n",
       "\n",
       "       biosample_id annual_season_precpt extreme_salinity water_content  \\\n",
       "count             0                    0                0             0   \n",
       "unique            0                    0                0             0   \n",
       "top             NaN                  NaN              NaN           NaN   \n",
       "freq            NaN                  NaN              NaN           NaN   \n",
       "\n",
       "       host_taxid  \n",
       "count           0  \n",
       "unique          0  \n",
       "top           NaN  \n",
       "freq          NaN  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num.where(df_onlyenv_nona_num == '0 m').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_onlyenv_nona_num_str_row_index = df_onlyenv_nona_num.applymap(np.isreal).all(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     34313\n",
       "unique        1\n",
       "top       False\n",
       "freq      34313\n",
       "dtype: object"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_onlyenv_nona_num_str_row_index.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_onlyenv_nona_num_nostr = df_onlyenv_nona_num[df_onlyenv_nona_num_str_row_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>449393</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26206</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>449393</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26207</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>449393</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26208</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>449393</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26209</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>449393</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>26210</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      samp_salinity altitude taxonomy_id ref_biomaterial  fire  \\\n",
       "19559          None     None      449393            None  None   \n",
       "19560          None     None      449393            None  None   \n",
       "19561          None     None      449393            None  None   \n",
       "19562          None     None      449393            None  None   \n",
       "19563          None     None      449393            None  None   \n",
       "\n",
       "      pool_dna_extracts microbial_biomass light_intensity entrez_value  \\\n",
       "19559              None              None            None         None   \n",
       "19560              None              None            None         None   \n",
       "19561              None              None            None         None   \n",
       "19562              None              None            None         None   \n",
       "19563              None              None            None         None   \n",
       "\n",
       "      biosample_id annual_season_precpt extreme_salinity water_content  \\\n",
       "19559        26206                 None             None          None   \n",
       "19560        26207                 None             None          None   \n",
       "19561        26208                 None             None          None   \n",
       "19562        26209                 None             None          None   \n",
       "19563        26210                 None             None          None   \n",
       "\n",
       "      host_taxid  \n",
       "19559       None  \n",
       "19560       None  \n",
       "19561       None  \n",
       "19562       None  \n",
       "19563       None  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_onlyenv_nona_num_nostr = pd.to_numeric(df_onlyenv_nona_num_nostr)\n",
    "df_onlyenv_nona_num_nostr = df_onlyenv_nona_num.apply(pd.to_numeric, errors=\"coerce\")\n",
    "#df_onlyenv_nona_num_nostr = df_onlyenv_nona_num.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26209</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       samp_salinity  altitude  taxonomy_id  ref_biomaterial  fire  \\\n",
       "19559            NaN       NaN       449393              NaN   NaN   \n",
       "19560            NaN       NaN       449393              NaN   NaN   \n",
       "19561            NaN       NaN       449393              NaN   NaN   \n",
       "19562            NaN       NaN       449393              NaN   NaN   \n",
       "19563            NaN       NaN       449393              NaN   NaN   \n",
       "\n",
       "       pool_dna_extracts  microbial_biomass  light_intensity  entrez_value  \\\n",
       "19559                NaN                NaN              NaN           NaN   \n",
       "19560                NaN                NaN              NaN           NaN   \n",
       "19561                NaN                NaN              NaN           NaN   \n",
       "19562                NaN                NaN              NaN           NaN   \n",
       "19563                NaN                NaN              NaN           NaN   \n",
       "\n",
       "       biosample_id  annual_season_precpt  extreme_salinity  water_content  \\\n",
       "19559         26206                   NaN               NaN            NaN   \n",
       "19560         26207                   NaN               NaN            NaN   \n",
       "19561         26208                   NaN               NaN            NaN   \n",
       "19562         26209                   NaN               NaN            NaN   \n",
       "19563         26210                   NaN               NaN            NaN   \n",
       "\n",
       "       host_taxid  \n",
       "19559         NaN  \n",
       "19560         NaN  \n",
       "19561         NaN  \n",
       "19562         NaN  \n",
       "19563         NaN  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_nostr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34313, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>49.0</td>\n",
       "      <td>5452.0</td>\n",
       "      <td>3.431300e+04</td>\n",
       "      <td>12.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>167.000000</td>\n",
       "      <td>52.000000</td>\n",
       "      <td>538.000000</td>\n",
       "      <td>3.431300e+04</td>\n",
       "      <td>1255.000000</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>454.000000</td>\n",
       "      <td>6.888000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.432479e+05</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>122.106381</td>\n",
       "      <td>26950.969231</td>\n",
       "      <td>323589.141264</td>\n",
       "      <td>8.303156e+06</td>\n",
       "      <td>984.823363</td>\n",
       "      <td>127.086957</td>\n",
       "      <td>43.035549</td>\n",
       "      <td>1.787051e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.966880e+05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>245.912228</td>\n",
       "      <td>21088.525193</td>\n",
       "      <td>83046.636894</td>\n",
       "      <td>3.707262e+06</td>\n",
       "      <td>557.667035</td>\n",
       "      <td>108.261012</td>\n",
       "      <td>39.858090</td>\n",
       "      <td>3.904299e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.890000e+02</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.453798</td>\n",
       "      <td>4900.000000</td>\n",
       "      <td>167137.000000</td>\n",
       "      <td>2.620600e+04</td>\n",
       "      <td>3.160000</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>7.859208</td>\n",
       "      <td>4.530000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.106580e+05</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.701232</td>\n",
       "      <td>12300.000000</td>\n",
       "      <td>316748.000000</td>\n",
       "      <td>5.602731e+06</td>\n",
       "      <td>870.000000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>31.872500</td>\n",
       "      <td>4.577000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.127550e+05</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>24.659574</td>\n",
       "      <td>23400.000000</td>\n",
       "      <td>316748.000000</td>\n",
       "      <td>6.807713e+06</td>\n",
       "      <td>975.000000</td>\n",
       "      <td>95.000000</td>\n",
       "      <td>36.945000</td>\n",
       "      <td>4.577000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.399280e+05</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>49.461738</td>\n",
       "      <td>30000.000000</td>\n",
       "      <td>316748.000000</td>\n",
       "      <td>9.908403e+06</td>\n",
       "      <td>975.000000</td>\n",
       "      <td>190.000000</td>\n",
       "      <td>43.175000</td>\n",
       "      <td>4.577000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.607662e+06</td>\n",
       "      <td>25818237.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1230.000000</td>\n",
       "      <td>85422.800000</td>\n",
       "      <td>602715.000000</td>\n",
       "      <td>1.573708e+07</td>\n",
       "      <td>6200.000000</td>\n",
       "      <td>332.000000</td>\n",
       "      <td>826.000000</td>\n",
       "      <td>1.671699e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       samp_salinity  altitude   taxonomy_id  ref_biomaterial    fire  \\\n",
       "count           49.0    5452.0  3.431300e+04             12.0   150.0   \n",
       "mean             8.0       0.0  6.432479e+05       25818237.0  2004.0   \n",
       "std              0.0       0.0  3.966880e+05              0.0     0.0   \n",
       "min              8.0       0.0  6.890000e+02       25818237.0  2004.0   \n",
       "25%              8.0       0.0  4.106580e+05       25818237.0  2004.0   \n",
       "50%              8.0       0.0  4.127550e+05       25818237.0  2004.0   \n",
       "75%              8.0       0.0  9.399280e+05       25818237.0  2004.0   \n",
       "max              8.0       0.0  2.607662e+06       25818237.0  2004.0   \n",
       "\n",
       "       pool_dna_extracts  microbial_biomass  light_intensity   entrez_value  \\\n",
       "count               12.0         167.000000        52.000000     538.000000   \n",
       "mean                 9.0         122.106381     26950.969231  323589.141264   \n",
       "std                  0.0         245.912228     21088.525193   83046.636894   \n",
       "min                  9.0           0.453798      4900.000000  167137.000000   \n",
       "25%                  9.0           5.701232     12300.000000  316748.000000   \n",
       "50%                  9.0          24.659574     23400.000000  316748.000000   \n",
       "75%                  9.0          49.461738     30000.000000  316748.000000   \n",
       "max                  9.0        1230.000000     85422.800000  602715.000000   \n",
       "\n",
       "       biosample_id  annual_season_precpt  extreme_salinity  water_content  \\\n",
       "count  3.431300e+04           1255.000000         23.000000     454.000000   \n",
       "mean   8.303156e+06            984.823363        127.086957      43.035549   \n",
       "std    3.707262e+06            557.667035        108.261012      39.858090   \n",
       "min    2.620600e+04              3.160000         30.000000       7.859208   \n",
       "25%    5.602731e+06            870.000000         40.000000      31.872500   \n",
       "50%    6.807713e+06            975.000000         95.000000      36.945000   \n",
       "75%    9.908403e+06            975.000000        190.000000      43.175000   \n",
       "max    1.573708e+07           6200.000000        332.000000     826.000000   \n",
       "\n",
       "         host_taxid  \n",
       "count  6.888000e+03  \n",
       "mean   1.787051e+05  \n",
       "std    3.904299e+05  \n",
       "min    4.530000e+03  \n",
       "25%    4.577000e+03  \n",
       "50%    4.577000e+03  \n",
       "75%    4.577000e+03  \n",
       "max    1.671699e+06  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_onlyenv_nona_num_nostr.shape)\n",
    "df_onlyenv_nona_num_nostr.describe(include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275236</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275237</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275238</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275239</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275240</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34313 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          samp_salinity  altitude  taxonomy_id  ref_biomaterial  fire  \\\n",
       "19559              True      True        False             True  True   \n",
       "19560              True      True        False             True  True   \n",
       "19561              True      True        False             True  True   \n",
       "19562              True      True        False             True  True   \n",
       "19563              True      True        False             True  True   \n",
       "...                 ...       ...          ...              ...   ...   \n",
       "14275236           True      True        False             True  True   \n",
       "14275237           True      True        False             True  True   \n",
       "14275238           True      True        False             True  True   \n",
       "14275239           True      True        False             True  True   \n",
       "14275240           True      True        False             True  True   \n",
       "\n",
       "          pool_dna_extracts  microbial_biomass  light_intensity  entrez_value  \\\n",
       "19559                  True               True             True          True   \n",
       "19560                  True               True             True          True   \n",
       "19561                  True               True             True          True   \n",
       "19562                  True               True             True          True   \n",
       "19563                  True               True             True          True   \n",
       "...                     ...                ...              ...           ...   \n",
       "14275236               True               True             True          True   \n",
       "14275237               True               True             True          True   \n",
       "14275238               True               True             True          True   \n",
       "14275239               True               True             True          True   \n",
       "14275240               True               True             True          True   \n",
       "\n",
       "          biosample_id  annual_season_precpt  extreme_salinity  water_content  \\\n",
       "19559            False                  True              True           True   \n",
       "19560            False                  True              True           True   \n",
       "19561            False                  True              True           True   \n",
       "19562            False                  True              True           True   \n",
       "19563            False                  True              True           True   \n",
       "...                ...                   ...               ...            ...   \n",
       "14275236         False                  True              True           True   \n",
       "14275237         False                  True              True           True   \n",
       "14275238         False                  True              True           True   \n",
       "14275239         False                  True              True           True   \n",
       "14275240         False                  True              True           True   \n",
       "\n",
       "          host_taxid  \n",
       "19559           True  \n",
       "19560           True  \n",
       "19561           True  \n",
       "19562           True  \n",
       "19563           True  \n",
       "...              ...  \n",
       "14275236        True  \n",
       "14275237        True  \n",
       "14275238        True  \n",
       "14275239        True  \n",
       "14275240        True  \n",
       "\n",
       "[34313 rows x 14 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_nostr.isnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19559       True\n",
      "19560       True\n",
      "19561       True\n",
      "19562       True\n",
      "19563       True\n",
      "            ... \n",
      "14275236    True\n",
      "14275237    True\n",
      "14275238    True\n",
      "14275239    True\n",
      "14275240    True\n",
      "Length: 34313, dtype: bool\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_salinity</th>\n",
       "      <th>altitude</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>fire</th>\n",
       "      <th>pool_dna_extracts</th>\n",
       "      <th>microbial_biomass</th>\n",
       "      <th>light_intensity</th>\n",
       "      <th>entrez_value</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>annual_season_precpt</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>water_content</th>\n",
       "      <th>host_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19559</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19560</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19561</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19562</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26209</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19563</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>449393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275236</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15737079</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275237</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15737080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275238</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15737081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275239</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15737082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14275240</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15737083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34313 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          samp_salinity  altitude  taxonomy_id  ref_biomaterial  fire  \\\n",
       "19559               NaN       NaN       449393              NaN   NaN   \n",
       "19560               NaN       NaN       449393              NaN   NaN   \n",
       "19561               NaN       NaN       449393              NaN   NaN   \n",
       "19562               NaN       NaN       449393              NaN   NaN   \n",
       "19563               NaN       NaN       449393              NaN   NaN   \n",
       "...                 ...       ...          ...              ...   ...   \n",
       "14275236            NaN       NaN       256318              NaN   NaN   \n",
       "14275237            NaN       NaN       256318              NaN   NaN   \n",
       "14275238            NaN       NaN       256318              NaN   NaN   \n",
       "14275239            NaN       NaN       256318              NaN   NaN   \n",
       "14275240            NaN       NaN       256318              NaN   NaN   \n",
       "\n",
       "          pool_dna_extracts  microbial_biomass  light_intensity  entrez_value  \\\n",
       "19559                   NaN                NaN              NaN           NaN   \n",
       "19560                   NaN                NaN              NaN           NaN   \n",
       "19561                   NaN                NaN              NaN           NaN   \n",
       "19562                   NaN                NaN              NaN           NaN   \n",
       "19563                   NaN                NaN              NaN           NaN   \n",
       "...                     ...                ...              ...           ...   \n",
       "14275236                NaN                NaN              NaN           NaN   \n",
       "14275237                NaN                NaN              NaN           NaN   \n",
       "14275238                NaN                NaN              NaN           NaN   \n",
       "14275239                NaN                NaN              NaN           NaN   \n",
       "14275240                NaN                NaN              NaN           NaN   \n",
       "\n",
       "          biosample_id  annual_season_precpt  extreme_salinity  water_content  \\\n",
       "19559            26206                   NaN               NaN            NaN   \n",
       "19560            26207                   NaN               NaN            NaN   \n",
       "19561            26208                   NaN               NaN            NaN   \n",
       "19562            26209                   NaN               NaN            NaN   \n",
       "19563            26210                   NaN               NaN            NaN   \n",
       "...                ...                   ...               ...            ...   \n",
       "14275236      15737079                   NaN               NaN            NaN   \n",
       "14275237      15737080                   NaN               NaN            NaN   \n",
       "14275238      15737081                   NaN               NaN            NaN   \n",
       "14275239      15737082                   NaN               NaN            NaN   \n",
       "14275240      15737083                   NaN               NaN            NaN   \n",
       "\n",
       "          host_taxid  \n",
       "19559            NaN  \n",
       "19560            NaN  \n",
       "19561            NaN  \n",
       "19562            NaN  \n",
       "19563            NaN  \n",
       "...              ...  \n",
       "14275236         NaN  \n",
       "14275237         NaN  \n",
       "14275238         NaN  \n",
       "14275239         NaN  \n",
       "14275240         NaN  \n",
       "\n",
       "[34313 rows x 14 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_nostr_nan_index = df_onlyenv_nona_num_nostr.isnull()\n",
    "df_onlyenv_nona_num_nostr_nan_rows_index = df_onlyenv_nona_num_nostr_nan_index.any(axis=1)\n",
    "print(df_onlyenv_nona_num_nostr_nan_rows_index)\n",
    "df_onlyenv_nona_num_nostr_nan_rows_index_rows = df_onlyenv_nona_num_nostr[df_onlyenv_nona_num_nostr_nan_rows_index]\n",
    "df_onlyenv_nona_num_nostr_nan_rows_index_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    34313.000000\n",
       "mean        11.561332\n",
       "std          0.608845\n",
       "min          9.000000\n",
       "25%         11.000000\n",
       "50%         12.000000\n",
       "75%         12.000000\n",
       "max         12.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyenv_nona_num_nostr_nans = df_onlyenv_nona_num_nostr.isnull().sum(axis=1)\n",
    "df_onlyenv_nona_num_nostr_nans.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Input contains NaN, infinity or a value too large for dtype('float32').",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-103-48f469f268f8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mdtree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mDecisionTreeClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmin_samples_split\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m9999\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m#df_trim = df.drop('env_medium', axis=1)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mdtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Decision Tree Classifer Created'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/tree/tree.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight, check_input, X_idx_sorted)\u001b[0m\n\u001b[1;32m    799\u001b[0m             \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_weight\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    800\u001b[0m             \u001b[0mcheck_input\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcheck_input\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 801\u001b[0;31m             X_idx_sorted=X_idx_sorted)\n\u001b[0m\u001b[1;32m    802\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    803\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/tree/tree.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight, check_input, X_idx_sorted)\u001b[0m\n\u001b[1;32m    114\u001b[0m         \u001b[0mrandom_state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_random_state\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom_state\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcheck_input\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 116\u001b[0;31m             \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mDTYPE\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccept_sparse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"csc\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    117\u001b[0m             \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensure_2d\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcheck_array\u001b[0;34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mforce_all_finite\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m             _assert_all_finite(array,\n\u001b[0;32m--> 573\u001b[0;31m                                allow_nan=force_all_finite == 'allow-nan')\n\u001b[0m\u001b[1;32m    574\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    575\u001b[0m     \u001b[0mshape_repr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_shape_repr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36m_assert_all_finite\u001b[0;34m(X, allow_nan)\u001b[0m\n\u001b[1;32m     54\u001b[0m                 not allow_nan and not np.isfinite(X).all()):\n\u001b[1;32m     55\u001b[0m             \u001b[0mtype_err\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'infinity'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mallow_nan\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'NaN, infinity'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg_err\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype_err\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Input contains NaN, infinity or a value too large for dtype('float32')."
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "features = []\n",
    "\n",
    "y= env_series#df['env_medium']\n",
    "X= df_onlyenv_nona_num_nostr #df[features]\n",
    "\n",
    "###TODO: increase splits\n",
    "dtree=DecisionTreeClassifier(min_samples_split=20, random_state=9999)\n",
    "#df_trim = df.drop('env_medium', axis=1) \n",
    "dtree.fit(X,y)\n",
    "print('Decision Tree Classifer Created')\n",
    "\n",
    "exec_time = (time.time() - start_time)\n",
    "print('seconds ' + str(exec_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_tree(tree, feature_names):\n",
    "    \"\"\"Create tree png using graphviz.\n",
    "\n",
    "    Args\n",
    "    ----\n",
    "    tree -- scikit-learn DecsisionTree.\n",
    "    feature_names -- list of feature names.\n",
    "    \"\"\"\n",
    "    with open(\"dt.dot\", 'w') as f:\n",
    "        export_graphviz(tree, out_file=f,\n",
    "                        feature_names=feature_names)\n",
    "\n",
    "    command = [\"dot\", \"-Tpng\", \"dt.dot\", \"-o\", \"dt.png\"]\n",
    "    try:\n",
    "        subprocess.check_call(command)\n",
    "    except:\n",
    "        exit(\"Could not run dot, ie graphviz, to \"\n",
    "             \"produce visualization\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "visualize_tree(dt, features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "dot_data = StringIO()\n",
    "export_graphviz(dtree, out_file=dot_data, feature_names=labels,  \n",
    "                filled=True, rounded=True,\n",
    "                special_characters=True)\n",
    "graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  \n",
    "Image(graph.create_png())\n",
    "\n",
    "exec_time = (time.time() - start_time)\n",
    "print('seconds ' + str(exec_time))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
